{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "086abf9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/blasks/Library/Caches/pypoetry/virtualenvs/barnacle-manuscript-4PSOCytX-py3.10/lib/python3.10/site-packages/rpy2/robjects/lib/ggplot2.py:70: UserWarning: This was designed againt ggplot2 versions starting with 3.3. but you have 3.4.2\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# imports\n",
    "\n",
    "# python packages\n",
    "import matplotlib as mpl\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.font_manager\n",
    "import rpy2\n",
    "from functools import reduce\n",
    "import xarray as xr\n",
    "\n",
    "# rpy2 imports\n",
    "from rpy2 import robjects as ro\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.ipython.ggplot import image_png\n",
    "from rpy2.robjects import pandas2ri\n",
    "\n",
    "# load rpy2 extension for ipython\n",
    "pandas2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "# stop showing SettingWithCopyWarning\n",
    "pd.options.mode.chained_assignment = None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7c54814d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.4.1\n"
     ]
    }
   ],
   "source": [
    "# install & import r package sctransform\n",
    "\n",
    "# check if sctransform is installed\n",
    "if not ro.packages.isinstalled('sctransform'):\n",
    "    # select CRAN mirror\n",
    "    utils = importr('utils')\n",
    "    utils.chooseCRANmirror(ind=1)\n",
    "    # install sctransform\n",
    "    utils.install_packages(ro.vectors.StrVector(['sctransform']))\n",
    "    \n",
    "# check if glmGamPoi is installed\n",
    "if not ro.packages.isinstalled('glmGamPoi'):\n",
    "    print('Please install glmGamPoi: https://github.com/const-ae/glmGamPoi')\n",
    "\n",
    "# import sctransform\n",
    "sctransform = importr('sctransform')\n",
    "rmatrix = importr('Matrix')\n",
    "\n",
    "# should be version 0.3.5 or higher                           \n",
    "print(sctransform.__version__)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f50fa06d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# helper functions\n",
    "\n",
    "# function to calculate 0-sensitive geometric mean\n",
    "def geometric_mean(vector, pseudocount=1):\n",
    "    return np.exp(np.mean(np.log(vector + pseudocount))) - pseudocount\n",
    "\n",
    "# function to convert pandas dataframe to r matrix\n",
    "def pandas_dataframe_to_r_matrix(df, dtype=float):\n",
    "    \"\"\"\n",
    "    Function to convert pandas DataFrame objects to R matrix objects.\n",
    "    \"\"\"\n",
    "    if dtype is float:\n",
    "        vector = ro.vectors.FloatVector(df.values.flatten().tolist())\n",
    "    elif dtype is str:\n",
    "        vector = ro.vectors.StrVector(df.values.flatten().tolist())\n",
    "    elif dtype is int:\n",
    "        vector = ro.vectors.FloatVector(df.values.flatten().tolist())\n",
    "    else:\n",
    "        raise ValueError('The dtype {} is not recognized'.format(dtype))\n",
    "    matrix = rmatrix.Matrix(\n",
    "        data=vector, \n",
    "        nrow=df.shape[0], \n",
    "        ncol=df.shape[1], \n",
    "        byrow=True, \n",
    "        dimnames=[df.index.to_list(), df.columns.to_list()], \n",
    "        sparse=True\n",
    "    )\n",
    "    return matrix\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3179d24a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "26c3bee8",
   "metadata": {},
   "source": [
    "# Read in collated mapping counts and metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "34e2f262",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>Cruise</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>Station</th>\n",
       "      <th>Cast</th>\n",
       "      <th>Datetime</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Depth</th>\n",
       "      <th>Replicate</th>\n",
       "      <th>SmallFraction</th>\n",
       "      <th>LargeFraction</th>\n",
       "      <th>Unfractionated</th>\n",
       "      <th>Batch</th>\n",
       "      <th>SampleName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.A</td>\n",
       "      <td>G1</td>\n",
       "      <td>Gradients 1 surface</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4/20/16 5:51</td>\n",
       "      <td>23.495833</td>\n",
       "      <td>-157.994333</td>\n",
       "      <td>15</td>\n",
       "      <td>A</td>\n",
       "      <td>G1.SURF.NS.S02C1.15m.0_2um.A</td>\n",
       "      <td>G1.SURF.NS.S02C1.15m.3um.A</td>\n",
       "      <td>False</td>\n",
       "      <td>G1_SURF</td>\n",
       "      <td>G1.SURF.NS.S02C1.15m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.B</td>\n",
       "      <td>G1</td>\n",
       "      <td>Gradients 1 surface</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4/20/16 5:51</td>\n",
       "      <td>23.495833</td>\n",
       "      <td>-157.994333</td>\n",
       "      <td>15</td>\n",
       "      <td>B</td>\n",
       "      <td>G1.SURF.NS.S02C1.15m.0_2um.B</td>\n",
       "      <td>G1.SURF.NS.S02C1.15m.3um.B</td>\n",
       "      <td>False</td>\n",
       "      <td>G1_SURF</td>\n",
       "      <td>G1.SURF.NS.S02C1.15m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.C</td>\n",
       "      <td>G1</td>\n",
       "      <td>Gradients 1 surface</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4/20/16 5:51</td>\n",
       "      <td>23.495833</td>\n",
       "      <td>-157.994333</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>G1.SURF.NS.S02C1.15m.0_2um.C</td>\n",
       "      <td>G1.SURF.NS.S02C1.15m.3um.C</td>\n",
       "      <td>False</td>\n",
       "      <td>G1_SURF</td>\n",
       "      <td>G1.SURF.NS.S02C1.15m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G1.SURF.NS.S04C1.15m.A</td>\n",
       "      <td>G1</td>\n",
       "      <td>Gradients 1 surface</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4/22/16 5:40</td>\n",
       "      <td>28.143167</td>\n",
       "      <td>-158.000667</td>\n",
       "      <td>15</td>\n",
       "      <td>A</td>\n",
       "      <td>G1.SURF.NS.S04C1.15m.0_2um.A</td>\n",
       "      <td>G1.SURF.NS.S04C1.15m.3um.A</td>\n",
       "      <td>False</td>\n",
       "      <td>G1_SURF</td>\n",
       "      <td>G1.SURF.NS.S04C1.15m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G1.SURF.NS.S04C1.15m.B</td>\n",
       "      <td>G1</td>\n",
       "      <td>Gradients 1 surface</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4/22/16 5:40</td>\n",
       "      <td>28.143167</td>\n",
       "      <td>-158.000667</td>\n",
       "      <td>15</td>\n",
       "      <td>B</td>\n",
       "      <td>G1.SURF.NS.S04C1.15m.0_2um.B</td>\n",
       "      <td>G1.SURF.NS.S04C1.15m.3um.B</td>\n",
       "      <td>False</td>\n",
       "      <td>G1_SURF</td>\n",
       "      <td>G1.SURF.NS.S04C1.15m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>G3.UW.NS.UW40_1.7m.B</td>\n",
       "      <td>G3</td>\n",
       "      <td>Gradients 3 underway</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4/24/19 5:57</td>\n",
       "      <td>40.880000</td>\n",
       "      <td>-158.000000</td>\n",
       "      <td>7</td>\n",
       "      <td>B</td>\n",
       "      <td>G3.UW.NS.UW40_1.7m.0_2um.B</td>\n",
       "      <td>G3.UW.NS.UW40_1.7m.3um.B</td>\n",
       "      <td>False</td>\n",
       "      <td>G3_SURF</td>\n",
       "      <td>G3.UW.NS.UW40_1.7m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>G3.UW.NS.UW40_1.7m.C</td>\n",
       "      <td>G3</td>\n",
       "      <td>Gradients 3 underway</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4/24/19 5:57</td>\n",
       "      <td>40.880000</td>\n",
       "      <td>-158.000000</td>\n",
       "      <td>7</td>\n",
       "      <td>C</td>\n",
       "      <td>G3.UW.NS.UW40_1.7m.0_2um.C</td>\n",
       "      <td>G3.UW.NS.UW40_1.7m.3um.C</td>\n",
       "      <td>False</td>\n",
       "      <td>G3_SURF</td>\n",
       "      <td>G3.UW.NS.UW40_1.7m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.A</td>\n",
       "      <td>G3</td>\n",
       "      <td>Gradients 3 underway</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4/25/19 6:01</td>\n",
       "      <td>40.090000</td>\n",
       "      <td>-158.000000</td>\n",
       "      <td>7</td>\n",
       "      <td>A</td>\n",
       "      <td>G3.UW.NS.UW40_2.7m.0_2um.A</td>\n",
       "      <td>G3.UW.NS.UW40_2.7m.3um.A</td>\n",
       "      <td>False</td>\n",
       "      <td>G3_SURF</td>\n",
       "      <td>G3.UW.NS.UW40_2.7m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.B</td>\n",
       "      <td>G3</td>\n",
       "      <td>Gradients 3 underway</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4/25/19 6:01</td>\n",
       "      <td>40.090000</td>\n",
       "      <td>-158.000000</td>\n",
       "      <td>7</td>\n",
       "      <td>B</td>\n",
       "      <td>G3.UW.NS.UW40_2.7m.0_2um.B</td>\n",
       "      <td>G3.UW.NS.UW40_2.7m.3um.B</td>\n",
       "      <td>False</td>\n",
       "      <td>G3_SURF</td>\n",
       "      <td>G3.UW.NS.UW40_2.7m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.C</td>\n",
       "      <td>G3</td>\n",
       "      <td>Gradients 3 underway</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4/25/19 6:01</td>\n",
       "      <td>40.090000</td>\n",
       "      <td>-158.000000</td>\n",
       "      <td>7</td>\n",
       "      <td>C</td>\n",
       "      <td>G3.UW.NS.UW40_2.7m.0_2um.C</td>\n",
       "      <td>G3.UW.NS.UW40_2.7m.3um.C</td>\n",
       "      <td>False</td>\n",
       "      <td>G3_SURF</td>\n",
       "      <td>G3.UW.NS.UW40_2.7m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>222 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SampleID Cruise               Dataset Experiment  Station  \\\n",
       "0    G1.SURF.NS.S02C1.15m.A     G1   Gradients 1 surface        NaN      2.0   \n",
       "1    G1.SURF.NS.S02C1.15m.B     G1   Gradients 1 surface        NaN      2.0   \n",
       "2    G1.SURF.NS.S02C1.15m.C     G1   Gradients 1 surface        NaN      2.0   \n",
       "3    G1.SURF.NS.S04C1.15m.A     G1   Gradients 1 surface        NaN      4.0   \n",
       "4    G1.SURF.NS.S04C1.15m.B     G1   Gradients 1 surface        NaN      4.0   \n",
       "..                      ...    ...                   ...        ...      ...   \n",
       "217    G3.UW.NS.UW40_1.7m.B     G3  Gradients 3 underway        NaN      NaN   \n",
       "218    G3.UW.NS.UW40_1.7m.C     G3  Gradients 3 underway        NaN      NaN   \n",
       "219    G3.UW.NS.UW40_2.7m.A     G3  Gradients 3 underway        NaN      NaN   \n",
       "220    G3.UW.NS.UW40_2.7m.B     G3  Gradients 3 underway        NaN      NaN   \n",
       "221    G3.UW.NS.UW40_2.7m.C     G3  Gradients 3 underway        NaN      NaN   \n",
       "\n",
       "     Cast      Datetime   Latitude   Longitude  Depth Replicate  \\\n",
       "0     1.0  4/20/16 5:51  23.495833 -157.994333     15         A   \n",
       "1     1.0  4/20/16 5:51  23.495833 -157.994333     15         B   \n",
       "2     1.0  4/20/16 5:51  23.495833 -157.994333     15         C   \n",
       "3     1.0  4/22/16 5:40  28.143167 -158.000667     15         A   \n",
       "4     1.0  4/22/16 5:40  28.143167 -158.000667     15         B   \n",
       "..    ...           ...        ...         ...    ...       ...   \n",
       "217   NaN  4/24/19 5:57  40.880000 -158.000000      7         B   \n",
       "218   NaN  4/24/19 5:57  40.880000 -158.000000      7         C   \n",
       "219   NaN  4/25/19 6:01  40.090000 -158.000000      7         A   \n",
       "220   NaN  4/25/19 6:01  40.090000 -158.000000      7         B   \n",
       "221   NaN  4/25/19 6:01  40.090000 -158.000000      7         C   \n",
       "\n",
       "                    SmallFraction               LargeFraction  Unfractionated  \\\n",
       "0    G1.SURF.NS.S02C1.15m.0_2um.A  G1.SURF.NS.S02C1.15m.3um.A           False   \n",
       "1    G1.SURF.NS.S02C1.15m.0_2um.B  G1.SURF.NS.S02C1.15m.3um.B           False   \n",
       "2    G1.SURF.NS.S02C1.15m.0_2um.C  G1.SURF.NS.S02C1.15m.3um.C           False   \n",
       "3    G1.SURF.NS.S04C1.15m.0_2um.A  G1.SURF.NS.S04C1.15m.3um.A           False   \n",
       "4    G1.SURF.NS.S04C1.15m.0_2um.B  G1.SURF.NS.S04C1.15m.3um.B           False   \n",
       "..                            ...                         ...             ...   \n",
       "217    G3.UW.NS.UW40_1.7m.0_2um.B    G3.UW.NS.UW40_1.7m.3um.B           False   \n",
       "218    G3.UW.NS.UW40_1.7m.0_2um.C    G3.UW.NS.UW40_1.7m.3um.C           False   \n",
       "219    G3.UW.NS.UW40_2.7m.0_2um.A    G3.UW.NS.UW40_2.7m.3um.A           False   \n",
       "220    G3.UW.NS.UW40_2.7m.0_2um.B    G3.UW.NS.UW40_2.7m.3um.B           False   \n",
       "221    G3.UW.NS.UW40_2.7m.0_2um.C    G3.UW.NS.UW40_2.7m.3um.C           False   \n",
       "\n",
       "       Batch            SampleName  \n",
       "0    G1_SURF  G1.SURF.NS.S02C1.15m  \n",
       "1    G1_SURF  G1.SURF.NS.S02C1.15m  \n",
       "2    G1_SURF  G1.SURF.NS.S02C1.15m  \n",
       "3    G1_SURF  G1.SURF.NS.S04C1.15m  \n",
       "4    G1_SURF  G1.SURF.NS.S04C1.15m  \n",
       "..       ...                   ...  \n",
       "217  G3_SURF    G3.UW.NS.UW40_1.7m  \n",
       "218  G3_SURF    G3.UW.NS.UW40_1.7m  \n",
       "219  G3_SURF    G3.UW.NS.UW40_2.7m  \n",
       "220  G3_SURF    G3.UW.NS.UW40_2.7m  \n",
       "221  G3_SURF    G3.UW.NS.UW40_2.7m  \n",
       "\n",
       "[222 rows x 16 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in metadata\n",
    "\n",
    "# filepaths\n",
    "filepath_genome_metadata = '../../data/metadata/genome-metadata.csv'\n",
    "filepath_ortholog_metadata = '../../data/metadata/ortholog-metadata.csv'\n",
    "filepath_sample_metadata = '../../data/metadata/sample-metadata.csv'\n",
    "\n",
    "# genome metadata\n",
    "genome_df = pd.read_csv(filepath_genome_metadata)\n",
    "\n",
    "#  ortholog metadata\n",
    "ortholog_df = pd.read_csv(filepath_ortholog_metadata)\n",
    "\n",
    "#  sample metadata\n",
    "sample_df = pd.read_csv(filepath_sample_metadata)\n",
    "sample_df['Cruise'] = sample_df['Cruise'].map({'kok1606': 'G1', \n",
    "                                               'mgl1704': 'G2', \n",
    "                                               'km1906': 'G3'})\n",
    "sample_df['SampleName'] = [name[:-2] for name in sample_df['SampleID']]\n",
    "\n",
    "# create annotations_df to map annotations back to orthologs\n",
    "annotations_df = ortholog_df[['CyCOGID', 'Annotation']].drop_duplicates()\n",
    "\n",
    "sample_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "639c9c8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MappingName</th>\n",
       "      <th>GeneLength</th>\n",
       "      <th>TPM</th>\n",
       "      <th>NumReads</th>\n",
       "      <th>SampleID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>GeneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-316-L16_2717627218</td>\n",
       "      <td>1116</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>G3.DIEL.NS.S4C21.15m.A</td>\n",
       "      <td>AG-316-L16</td>\n",
       "      <td>2717627218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AG-316-L16_2717627219</td>\n",
       "      <td>1290</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>G3.DIEL.NS.S4C21.15m.A</td>\n",
       "      <td>AG-316-L16</td>\n",
       "      <td>2717627219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-316-L16_2717627220</td>\n",
       "      <td>1578</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>G3.DIEL.NS.S4C21.15m.A</td>\n",
       "      <td>AG-316-L16</td>\n",
       "      <td>2717627220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AG-316-L16_2717627221</td>\n",
       "      <td>99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>G3.DIEL.NS.S4C21.15m.A</td>\n",
       "      <td>AG-316-L16</td>\n",
       "      <td>2717627221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AG-316-L16_2717627222</td>\n",
       "      <td>1050</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>G3.DIEL.NS.S4C21.15m.A</td>\n",
       "      <td>AG-316-L16</td>\n",
       "      <td>2717627222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219859471</th>\n",
       "      <td>Syn9_638323192</td>\n",
       "      <td>228</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>G3.DIEL.NS.S4C13.15m.B</td>\n",
       "      <td>Syn9</td>\n",
       "      <td>638323192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219859472</th>\n",
       "      <td>Syn9_638323193</td>\n",
       "      <td>711</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>G3.DIEL.NS.S4C13.15m.B</td>\n",
       "      <td>Syn9</td>\n",
       "      <td>638323193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219859473</th>\n",
       "      <td>Syn9_638323194</td>\n",
       "      <td>180</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>G3.DIEL.NS.S4C13.15m.B</td>\n",
       "      <td>Syn9</td>\n",
       "      <td>638323194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219859474</th>\n",
       "      <td>Syn9_638323195</td>\n",
       "      <td>231</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>G3.DIEL.NS.S4C13.15m.B</td>\n",
       "      <td>Syn9</td>\n",
       "      <td>638323195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219859475</th>\n",
       "      <td>Syn9_638323196</td>\n",
       "      <td>765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>G3.DIEL.NS.S4C13.15m.B</td>\n",
       "      <td>Syn9</td>\n",
       "      <td>638323196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>219859476 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     MappingName  GeneLength  TPM  NumReads  \\\n",
       "0          AG-316-L16_2717627218        1116  0.0       0.0   \n",
       "1          AG-316-L16_2717627219        1290  0.0       0.0   \n",
       "2          AG-316-L16_2717627220        1578  0.0       0.0   \n",
       "3          AG-316-L16_2717627221          99  0.0       0.0   \n",
       "4          AG-316-L16_2717627222        1050  0.0       0.0   \n",
       "...                          ...         ...  ...       ...   \n",
       "219859471         Syn9_638323192         228  0.0       0.0   \n",
       "219859472         Syn9_638323193         711  0.0       0.0   \n",
       "219859473         Syn9_638323194         180  0.0       0.0   \n",
       "219859474         Syn9_638323195         231  0.0       0.0   \n",
       "219859475         Syn9_638323196         765  0.0       0.0   \n",
       "\n",
       "                         SampleID  GenomeName      GeneID  \n",
       "0          G3.DIEL.NS.S4C21.15m.A  AG-316-L16  2717627218  \n",
       "1          G3.DIEL.NS.S4C21.15m.A  AG-316-L16  2717627219  \n",
       "2          G3.DIEL.NS.S4C21.15m.A  AG-316-L16  2717627220  \n",
       "3          G3.DIEL.NS.S4C21.15m.A  AG-316-L16  2717627221  \n",
       "4          G3.DIEL.NS.S4C21.15m.A  AG-316-L16  2717627222  \n",
       "...                           ...         ...         ...  \n",
       "219859471  G3.DIEL.NS.S4C13.15m.B        Syn9   638323192  \n",
       "219859472  G3.DIEL.NS.S4C13.15m.B        Syn9   638323193  \n",
       "219859473  G3.DIEL.NS.S4C13.15m.B        Syn9   638323194  \n",
       "219859474  G3.DIEL.NS.S4C13.15m.B        Syn9   638323195  \n",
       "219859475  G3.DIEL.NS.S4C13.15m.B        Syn9   638323196  \n",
       "\n",
       "[219859476 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in mapped transcript abundance data\n",
    "\n",
    "filepath_data = '../../data/2-mapping/collated_salmon_data.csv.gz'\n",
    "mappings_df = pd.read_csv(filepath_data)\n",
    "mappings_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c39bc258",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "af89f699",
   "metadata": {},
   "source": [
    "# Process mapped reads\n",
    "\n",
    "- Drop reference genes without ortholog mapping (hypothesized to mostly be rRNA & other highly conserved RNA)\n",
    "- Within each clade in each sample, aggregate reads mapping to the same ortholog\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "36cf3abf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# directory to store output\n",
    "\n",
    "output_dir = '../../data/3-normalization'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c9c6fc8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Group\n",
      "Prochlorococcus            0.745148\n",
      "Synechococcus              0.252534\n",
      "Uncultured-marine-virus    0.000285\n",
      "Virus                      0.002032\n",
      "Name: NumReads, dtype: float64\n",
      "Clade\n",
      "5.1A-CRD2            0.024086\n",
      "5.1A-II              0.007827\n",
      "5.1A-III             0.006876\n",
      "5.1A-IV              0.065266\n",
      "5.1A-UC-A-EnvC       0.005206\n",
      "5.1A-WPC1            0.000570\n",
      "5.1A-unclassified    0.014533\n",
      "5.1B-CRD1            0.025221\n",
      "5.1B-I               0.036755\n",
      "5.1B-IX              0.004742\n",
      "5.1B-V               0.000115\n",
      "5.1B-VI              0.000110\n",
      "5.1B-VIII            0.000217\n",
      "5.2                  0.022553\n",
      "5.3                  0.038389\n",
      "AMZ-II               0.000671\n",
      "HLI                  0.293027\n",
      "HLII                 0.181326\n",
      "HLII.HLVI            0.001085\n",
      "HLIII                0.015565\n",
      "HLIII.HLIV.HLV       0.000524\n",
      "HLIV                 0.011601\n",
      "HLVI                 0.002703\n",
      "LLI                  0.127157\n",
      "LLI.LLVIII           0.002676\n",
      "LLII.LLIII           0.024737\n",
      "LLIV                 0.002942\n",
      "LLVII                0.080787\n",
      "LLVIII               0.000345\n",
      "Unclassified         0.000067\n",
      "Name: NumReads, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "# look at read mappings per group and clade\n",
    "\n",
    "pangenome_df = mappings_df[['MappingName', 'SampleID', 'GenomeName', 'NumReads']]\n",
    "\n",
    "# add genus, clade, and ortholog info to dataframe\n",
    "pangenome_df['Group'] = pangenome_df['GenomeName'].map(genome_df.set_index('GenomeName')['Group'])\n",
    "pangenome_df['Clade'] = pangenome_df['GenomeName'].map(genome_df.set_index('GenomeName')['Clade'])\n",
    "pangenome_df['CyCOGID'] = pangenome_df['MappingName'].map(ortholog_df.set_index('MappingName')['CyCOGID'])\n",
    "\n",
    "# look at what proportion of reads mapped to each group\n",
    "total_reads = pangenome_df.NumReads.sum()\n",
    "print(pangenome_df.groupby('Group').NumReads.sum() / total_reads)\n",
    "print(pangenome_df.groupby('Clade').NumReads.sum() / total_reads)\n",
    "\n",
    "# make dataframe of all reads aggregated by clade within each sample\n",
    "clade_abun_df = pangenome_df.groupby(['SampleID', 'Group', 'Clade']).NumReads.sum().reset_index().rename(\n",
    "    columns={'NumReads': 'CladeReads'})\n",
    "clade_abun_df['SampleReads'] = clade_abun_df['SampleID'].map(clade_abun_df.groupby('SampleID').CladeReads.sum())\n",
    "clade_abun_df['RelativeAbundance'] = clade_abun_df['CladeReads'] / clade_abun_df['SampleReads']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ac08a0d6-0724-4de0-96d9-d06a37c9c3b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>Group</th>\n",
       "      <th>Clade</th>\n",
       "      <th>CyCOGID</th>\n",
       "      <th>Reads</th>\n",
       "      <th>Nonzero</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.A</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>60000002.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.A</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>60000003.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.A</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>60000004.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.A</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>60000005.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.A</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>60000006.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21899629</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.C</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>60039505.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21899630</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.C</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>60039506.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21899631</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.C</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>60039507.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21899632</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.C</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>60039508.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21899633</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.C</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>60039663.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21899634 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        SampleID            Group         Clade     CyCOGID  \\\n",
       "0         G1.SURF.NS.S02C1.15m.A  Prochlorococcus        AMZ-II  60000002.0   \n",
       "1         G1.SURF.NS.S02C1.15m.A  Prochlorococcus        AMZ-II  60000003.0   \n",
       "2         G1.SURF.NS.S02C1.15m.A  Prochlorococcus        AMZ-II  60000004.0   \n",
       "3         G1.SURF.NS.S02C1.15m.A  Prochlorococcus        AMZ-II  60000005.0   \n",
       "4         G1.SURF.NS.S02C1.15m.A  Prochlorococcus        AMZ-II  60000006.0   \n",
       "...                          ...              ...           ...         ...   \n",
       "21899629    G3.UW.NS.UW40_2.7m.C    Synechococcus  Unclassified  60039505.0   \n",
       "21899630    G3.UW.NS.UW40_2.7m.C    Synechococcus  Unclassified  60039506.0   \n",
       "21899631    G3.UW.NS.UW40_2.7m.C    Synechococcus  Unclassified  60039507.0   \n",
       "21899632    G3.UW.NS.UW40_2.7m.C    Synechococcus  Unclassified  60039508.0   \n",
       "21899633    G3.UW.NS.UW40_2.7m.C    Synechococcus  Unclassified  60039663.0   \n",
       "\n",
       "          Reads  Nonzero  \n",
       "0           0.0    False  \n",
       "1           0.0    False  \n",
       "2           0.0    False  \n",
       "3           0.0    False  \n",
       "4           0.0    False  \n",
       "...         ...      ...  \n",
       "21899629    0.0    False  \n",
       "21899630    1.0     True  \n",
       "21899631    0.0    False  \n",
       "21899632    0.0    False  \n",
       "21899633    0.0    False  \n",
       "\n",
       "[21899634 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# collect reads mapping to known orthologs only and aggregate within orthologs\n",
    "\n",
    "# down-select to only mappings that don't belong to a virus or virocell reference\n",
    "drop_references = genome_df[(genome_df['Group'] == 'Virus') | (genome_df['Virocell'])].GenomeName.to_list()\n",
    "pangenome_df = pangenome_df[~pangenome_df.GenomeName.isin(drop_references)]\n",
    "\n",
    "# down-select to only the genes with an ortholog mapping\n",
    "pangenome_df = pangenome_df[~pangenome_df['CyCOGID'].isna()]\n",
    "\n",
    "# aggregate all orthologs in same genus\n",
    "pangenome_df = pangenome_df.groupby(['SampleID', 'Group', 'Clade', 'CyCOGID']).agg(Reads=('NumReads', 'sum')).reset_index()\n",
    "\n",
    "# identify nonzero mappings (at least one read mapped to ortholog-sample-clade)\n",
    "pangenome_df['Nonzero'] = pangenome_df['Reads'].gt(0.0)\n",
    "\n",
    "pangenome_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "898b977d-c131-4685-b529-8456519da9f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>Group</th>\n",
       "      <th>Clade</th>\n",
       "      <th>CladeReads</th>\n",
       "      <th>SampleReads</th>\n",
       "      <th>RelativeAbundance</th>\n",
       "      <th>PangenomeSize</th>\n",
       "      <th>DetectedPangenomeSize</th>\n",
       "      <th>DetectedCyCOGs</th>\n",
       "      <th>PangenomeCoverage</th>\n",
       "      <th>DetectedPangenomeCoverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.A</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>507.205</td>\n",
       "      <td>5433304.056</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>1762.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.A</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>HLI</td>\n",
       "      <td>272443.296</td>\n",
       "      <td>5433304.056</td>\n",
       "      <td>0.050143</td>\n",
       "      <td>4906.0</td>\n",
       "      <td>3906.0</td>\n",
       "      <td>575.0</td>\n",
       "      <td>0.117203</td>\n",
       "      <td>0.147209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.A</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>HLII</td>\n",
       "      <td>4828459.042</td>\n",
       "      <td>5433304.056</td>\n",
       "      <td>0.888678</td>\n",
       "      <td>6997.0</td>\n",
       "      <td>4787.0</td>\n",
       "      <td>1740.0</td>\n",
       "      <td>0.248678</td>\n",
       "      <td>0.363484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.A</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>HLII.HLVI</td>\n",
       "      <td>34148.879</td>\n",
       "      <td>5433304.056</td>\n",
       "      <td>0.006285</td>\n",
       "      <td>1822.0</td>\n",
       "      <td>1144.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>0.080132</td>\n",
       "      <td>0.127622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G1.SURF.NS.S02C1.15m.A</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>HLIII</td>\n",
       "      <td>24885.537</td>\n",
       "      <td>5433304.056</td>\n",
       "      <td>0.004580</td>\n",
       "      <td>1652.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.006659</td>\n",
       "      <td>0.049774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6655</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.C</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>5.1B-VI</td>\n",
       "      <td>379.143</td>\n",
       "      <td>6469143.806</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>2564.0</td>\n",
       "      <td>226.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.003120</td>\n",
       "      <td>0.035398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6656</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.C</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>5.1B-VIII</td>\n",
       "      <td>2009.257</td>\n",
       "      <td>6469143.806</td>\n",
       "      <td>0.000311</td>\n",
       "      <td>2515.0</td>\n",
       "      <td>239.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.001193</td>\n",
       "      <td>0.012552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6657</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.C</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>5.2</td>\n",
       "      <td>267334.246</td>\n",
       "      <td>6469143.806</td>\n",
       "      <td>0.041325</td>\n",
       "      <td>5123.0</td>\n",
       "      <td>481.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.002538</td>\n",
       "      <td>0.027027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6658</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.C</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>5.3</td>\n",
       "      <td>1016616.381</td>\n",
       "      <td>6469143.806</td>\n",
       "      <td>0.157149</td>\n",
       "      <td>2783.0</td>\n",
       "      <td>526.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000719</td>\n",
       "      <td>0.003802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6659</th>\n",
       "      <td>G3.UW.NS.UW40_2.7m.C</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>12.500</td>\n",
       "      <td>6469143.806</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>440.0</td>\n",
       "      <td>283.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.018182</td>\n",
       "      <td>0.028269</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6660 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    SampleID            Group         Clade   CladeReads  \\\n",
       "0     G1.SURF.NS.S02C1.15m.A  Prochlorococcus        AMZ-II      507.205   \n",
       "1     G1.SURF.NS.S02C1.15m.A  Prochlorococcus           HLI   272443.296   \n",
       "2     G1.SURF.NS.S02C1.15m.A  Prochlorococcus          HLII  4828459.042   \n",
       "3     G1.SURF.NS.S02C1.15m.A  Prochlorococcus     HLII.HLVI    34148.879   \n",
       "4     G1.SURF.NS.S02C1.15m.A  Prochlorococcus         HLIII    24885.537   \n",
       "...                      ...              ...           ...          ...   \n",
       "6655    G3.UW.NS.UW40_2.7m.C    Synechococcus       5.1B-VI      379.143   \n",
       "6656    G3.UW.NS.UW40_2.7m.C    Synechococcus     5.1B-VIII     2009.257   \n",
       "6657    G3.UW.NS.UW40_2.7m.C    Synechococcus           5.2   267334.246   \n",
       "6658    G3.UW.NS.UW40_2.7m.C    Synechococcus           5.3  1016616.381   \n",
       "6659    G3.UW.NS.UW40_2.7m.C    Synechococcus  Unclassified       12.500   \n",
       "\n",
       "      SampleReads  RelativeAbundance  PangenomeSize  DetectedPangenomeSize  \\\n",
       "0     5433304.056           0.000093         1762.0                  127.0   \n",
       "1     5433304.056           0.050143         4906.0                 3906.0   \n",
       "2     5433304.056           0.888678         6997.0                 4787.0   \n",
       "3     5433304.056           0.006285         1822.0                 1144.0   \n",
       "4     5433304.056           0.004580         1652.0                  221.0   \n",
       "...           ...                ...            ...                    ...   \n",
       "6655  6469143.806           0.000059         2564.0                  226.0   \n",
       "6656  6469143.806           0.000311         2515.0                  239.0   \n",
       "6657  6469143.806           0.041325         5123.0                  481.0   \n",
       "6658  6469143.806           0.157149         2783.0                  526.0   \n",
       "6659  6469143.806           0.000002          440.0                  283.0   \n",
       "\n",
       "      DetectedCyCOGs  PangenomeCoverage  DetectedPangenomeCoverage  \n",
       "0                0.0           0.000000                   0.000000  \n",
       "1              575.0           0.117203                   0.147209  \n",
       "2             1740.0           0.248678                   0.363484  \n",
       "3              146.0           0.080132                   0.127622  \n",
       "4               11.0           0.006659                   0.049774  \n",
       "...              ...                ...                        ...  \n",
       "6655             8.0           0.003120                   0.035398  \n",
       "6656             3.0           0.001193                   0.012552  \n",
       "6657            13.0           0.002538                   0.027027  \n",
       "6658             2.0           0.000719                   0.003802  \n",
       "6659             8.0           0.018182                   0.028269  \n",
       "\n",
       "[6660 rows x 11 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate genome coverage in each sample\n",
    "\n",
    "clade_abun_df['PangenomeSize'] = clade_abun_df['Clade'].map(pangenome_df.groupby(['Clade']).CyCOGID.nunique())\n",
    "clade_abun_df['DetectedPangenomeSize'] = clade_abun_df['Clade'].map(\n",
    "    pangenome_df[pangenome_df.Nonzero].groupby(['Clade']).CyCOGID.nunique())\n",
    "clade_abun_df = pd.merge(\n",
    "    clade_abun_df, \n",
    "    pangenome_df.groupby(['SampleID', 'Clade']).Nonzero.sum().reset_index().rename(columns={'Nonzero':'DetectedCyCOGs'}), \n",
    "    on=['SampleID', 'Clade'], \n",
    "    how='left'\n",
    ")\n",
    "clade_abun_df['PangenomeCoverage'] = clade_abun_df['DetectedCyCOGs'] / clade_abun_df['PangenomeSize']\n",
    "clade_abun_df['DetectedPangenomeCoverage'] = clade_abun_df['DetectedCyCOGs'] / clade_abun_df['DetectedPangenomeSize']\n",
    "\n",
    "# save as csv\n",
    "clade_abun_df.to_csv('../../data/2-mapping/aggregate-clade-reads.csv', index=False)\n",
    "clade_abun_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a60fd502-7f4b-489d-a1a4-a25d8cb076d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# maxes\n",
    "\n",
    "clade_abun_df.groupby(['Group', 'Clade'])[['PangenomeSize', 'DetectedPangenomeSize', \n",
    "                                           'PangenomeCoverage', 'DetectedPangenomeCoverage']].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fe1324d-0d56-4992-91b6-6ffad63ab5f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "clade_abun_df.groupby(['Group', 'Clade']).DetectedPangenome.max() / clade_abun_df.groupby(['Group', 'Clade']).PangenomeSize.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b960da40-aa3f-4f30-bcc6-c3b04cde45dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_df['Clade'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8789e56-001f-4d81-b95c-bddda97d7da5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c4ec980",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e05049d7",
   "metadata": {},
   "source": [
    "# Separate out data by clade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef7922f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# separate out data for most highly represented clades in dataset\n",
    "\n",
    "# thresholds\n",
    "detection_min = 3\n",
    "coverage_threshold = 0.01\n",
    "clades = {\n",
    "    'pro': ['HLI', 'HLII', 'LLI', 'LLII.LLIII', 'LLIV', 'LLVII'], \n",
    "    'syn': ['5.1A-CRD2', '5.1A-II', '5.1A-III', '5.1A-IV', '5.1A-UC-A-EnvC', '5.1B-CRD1', '5.1B-I']\n",
    "}\n",
    "\n",
    "# storage variables\n",
    "ortho_list = []\n",
    "sample_list = []\n",
    "core_df_dict = {}\n",
    "\n",
    "for selected_clade in clades['pro'] + clades['syn']:\n",
    "    print(selected_clade)\n",
    "    clade_pangenome_df = pangenome_df[pangenome_df.Clade == selected_clade]\n",
    "    \n",
    "    # # make fresh directory\n",
    "    # dir_path = '{}/{}'.format(output_dir, selected_clade)\n",
    "    # if not os.path.isdir(dir_path):\n",
    "    #     os.makedirs(dir_path)\n",
    "    \n",
    "    # down-select data to detected orthologs\n",
    "    core_df = pd.DataFrame(clade_pangenome_df.groupby('CyCOGID').Nonzero.sum())\n",
    "    core_df = core_df[core_df.Nonzero.gt(0)]\n",
    "#     core_df = core_df[core_df.Nonzero.ge(detection_min)]\n",
    "    clade_pangenome_df = clade_pangenome_df[clade_pangenome_df.CyCOGID.isin(core_df.index)]\n",
    "\n",
    "    # down-select to only samples with coverage that is at least x% of the max coverage found in the dataset\n",
    "    coverage = clade_pangenome_df.groupby('SampleID').Nonzero.sum()\n",
    "    max_cov = coverage.max()\n",
    "    total_orthos = len(core_df.index)\n",
    "    print('Max coverage: {} / {} orthologs ({})'.format(max_cov, total_orthos, max_cov/total_orthos))\n",
    "    samples = coverage[coverage.gt(coverage.max() * coverage_threshold)].index\n",
    "    sample_list.append(samples)\n",
    "    min_cov = coverage[samples].min()\n",
    "    print('Min coverage: {} / {} orthologs ({})'.format(min_cov, total_orthos, min_cov/total_orthos))\n",
    "    print('Total samples: {}\\n'.format(len(samples)))\n",
    "    clade_pangenome_df = clade_pangenome_df[clade_pangenome_df['SampleID'].isin(samples)]\n",
    "    \n",
    "    # down-select data to only orthologs that show up in a minimum number of the remaining samples\n",
    "    core_df = pd.DataFrame(clade_pangenome_df.groupby('CyCOGID').Nonzero.sum())\n",
    "    core_df = core_df[core_df.Nonzero.ge(detection_min)]\n",
    "    clade_pangenome_df = clade_pangenome_df[clade_pangenome_df.CyCOGID.isin(core_df.index)]\n",
    "    ortho_list.append(core_df.index)\n",
    "\n",
    "    # pivot table\n",
    "    clade_core_df = clade_pangenome_df.pivot(index='CyCOGID', columns=['SampleID'], values='Reads')\n",
    "    core_df_dict[selected_clade] = clade_core_df\n",
    "    \n",
    "    # # save pre-normed data to csv\n",
    "    # clade_core_df.to_csv('{}/read_counts_{}.csv'.format(dir_path, selected_clade))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65f74da6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# look at intersections of samples and genes\n",
    "\n",
    "# universal\n",
    "print('Total cyanobacterial orthologs: {}'.format(len(reduce(np.union1d, ortho_list))))\n",
    "print('Total cyanobacterial samples: {}'.format(len(reduce(np.union1d, sample_list))))\n",
    "print('Shared cyanobacterial orthologs: {}'.format(len(reduce(np.intersect1d, ortho_list))))\n",
    "print('Shared cyanobacterial samples: {}\\n'.format(len(reduce(np.intersect1d, sample_list))))\n",
    "\n",
    "# pro\n",
    "print('Total Pro orthologs: {}'.format(len(reduce(np.union1d, ortho_list[:3]))))\n",
    "print('Total Pro samples: {}'.format(len(reduce(np.union1d, sample_list[:3]))))\n",
    "print('Shared Pro orthologs: {}'.format(len(reduce(np.intersect1d, ortho_list[:3]))))\n",
    "print('Shared Pro samples: {}\\n'.format(len(reduce(np.intersect1d, sample_list[:3]))))\n",
    "\n",
    "# syn\n",
    "print('Total Syn orthologs: {}'.format(len(reduce(np.union1d, ortho_list[3:]))))\n",
    "print('Total Syn samples: {}'.format(len(reduce(np.union1d, sample_list[3:]))))\n",
    "print('Shared Syn orthologs: {}'.format(len(reduce(np.intersect1d, ortho_list[3:]))))\n",
    "print('Shared Syn samples: {}\\n'.format(len(reduce(np.intersect1d, sample_list[3:]))))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58228a6d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "98d45800",
   "metadata": {},
   "source": [
    "# Normalize data using sctransform\n",
    "\n",
    "* Save diagnostic plots & csv files, as well as tensorized data in NetCDF4 format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d991bd7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# run the model on each clade, saving outputs and plots\n",
    "\n",
    "# store normalized data as slabs to arrange in xarray tensor\n",
    "read_slabs = {}\n",
    "residual_slabs = {}\n",
    "\n",
    "# iterate through pro & syn\n",
    "for genus in ['pro', 'syn']:\n",
    "    print(genus)\n",
    "    read_slabs[genus] = []\n",
    "    residual_slabs[genus] = []\n",
    "        \n",
    "    # iterate through clades\n",
    "    for clade in clades[genus]:\n",
    "        print(clade)\n",
    "\n",
    "        # make fresh directory\n",
    "        dir_path = '{}/{}'.format(output_dir, clade)\n",
    "        if not os.path.isdir(dir_path):\n",
    "            os.makedirs(dir_path)\n",
    "\n",
    "        # convert clade_core_df to r matrix\n",
    "        clade_core_df = core_df_dict[clade]\n",
    "        r_clade_core_df = pandas_dataframe_to_r_matrix(clade_core_df)\n",
    "\n",
    "        # fit vst normalization model\n",
    "        result = sctransform.vst(\n",
    "            r_clade_core_df, \n",
    "            n_genes='null', \n",
    "            min_cells=detection_min,\n",
    "            return_gene_attr=True, \n",
    "            return_cell_attr=True, \n",
    "            vst_flavor='v2', \n",
    "            verbosity=1\n",
    "        )\n",
    "\n",
    "        # save residuals as csv\n",
    "        residual_df = pd.DataFrame(\n",
    "            np.asarray(result[0]), \n",
    "            index=clade_core_df.index, \n",
    "            columns=clade_core_df.columns\n",
    "        )\n",
    "        residual_df.to_csv('{}/residuals_{}.csv'.format(dir_path, clade))\n",
    "        \n",
    "        # save residuals as slab \n",
    "        slab = (\n",
    "            residual_df\n",
    "            .reset_index()\n",
    "            .rename(columns={'OrthologID': 'ortholog'})\n",
    "            .melt(id_vars='ortholog', var_name='sample', value_name='residual')\n",
    "        )\n",
    "        slab['ortholog'] = slab['ortholog'].astype('Int64')\n",
    "        slab['clade'] = clade\n",
    "        residual_slabs[genus].append(slab.set_index(['ortholog', 'clade', 'sample']))\n",
    "        \n",
    "        # save read counts as slab \n",
    "        slab = (\n",
    "            clade_core_df\n",
    "            .reset_index()\n",
    "            .rename(columns={'OrthologID': 'ortholog'})\n",
    "            .melt(id_vars='ortholog', var_name='sample', value_name='readcount')\n",
    "        )\n",
    "        slab['ortholog'] = slab['ortholog'].astype('Int64')\n",
    "        slab['clade'] = clade\n",
    "        read_slabs[genus].append(slab.set_index(['ortholog', 'clade', 'sample']))\n",
    "\n",
    "        # save plot of model parameters\n",
    "        plots = sctransform.plot_model_pars(result, show_theta=True)\n",
    "        img = image_png(plots)\n",
    "        with open('{}/parameters_{}.png'.format(dir_path, clade), 'wb') as png:\n",
    "            png.write(img.data)\n",
    "\n",
    "        # plot high variance genes\n",
    "        residual_var = residual_df.var(axis=1)\n",
    "        three_sigma = 1 + residual_var.var() * 3\n",
    "        expression = clade_core_df.apply(geometric_mean, axis=1)\n",
    "        plt.figure(figsize=(10, 4))\n",
    "        sns.scatterplot(x=expression, y=residual_var, alpha=0.1);\n",
    "        plt.hlines(three_sigma, expression.min(), expression.max(), \n",
    "                   colors=['orange'], linestyles=[':'], label='3*sigma');\n",
    "        plt.xlabel('mean gene abundance (reads/sample)')\n",
    "        plt.xscale('log')\n",
    "        plt.ylabel('residual variance')\n",
    "        plt.title('normalized residual abundance of {} genes'.format(clade))\n",
    "        plt.savefig('{}/residual_variance_{}.png'.format(dir_path, clade))\n",
    "        plt.show()\n",
    "\n",
    "        # save csv of residual variances with annotations\n",
    "        res_var_df = residual_var.reset_index().rename(columns={0:'ResidualVariance'})\n",
    "        res_var_df['Annotation'] = res_var_df.OrthologID.map(annotations_df.set_index('ortholog')['annotation'])\n",
    "        res_var_df = res_var_df.sort_values('ResidualVariance', ascending=False).reset_index()\n",
    "        res_var_df.to_csv('{}/high_variance_orthologs_{}.csv'.format(dir_path, clade))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bc4d8f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make slabs into xarray tensors and save\n",
    "\n",
    "# make DataArrays for metadata\n",
    "replicate_da = xr.DataArray.from_series(sample_df.set_index('sample')['replicate'])\n",
    "samplename_da = xr.DataArray.from_series(sample_df.set_index('sample')['samplename'])\n",
    "annotation_da = xr.DataArray.from_series(annotations_df.set_index('ortholog')['annotation'])\n",
    "\n",
    "for genus in ['pro', 'syn']:\n",
    "    # make read count DataArray\n",
    "    read_data = read_slabs[genus]\n",
    "    read_all_slabs = pd.concat(read_data)\n",
    "    read_da = xr.DataArray.from_series(read_all_slabs['readcount']).fillna(0)\n",
    "    # make residual DataArray\n",
    "    residual_data = residual_slabs[genus]\n",
    "    residual_all_slabs = pd.concat(residual_data)\n",
    "    residual_da = xr.DataArray.from_series(residual_all_slabs['residual']).fillna(0)\n",
    "    # combine residuals with metadata into xarray Dataset\n",
    "    ds = xr.Dataset(\n",
    "        dict(\n",
    "            readcount=read_da, \n",
    "            residual=residual_da, \n",
    "            replicate=replicate_da.loc[residual_da.sample],\n",
    "            samplename=samplename_da.loc[residual_da.sample], \n",
    "            annotation=annotation_da.loc[residual_da.ortholog]\n",
    "        )\n",
    "    )\n",
    "    # save as netCDF4 file in data directory\n",
    "    ds.to_netcdf('{}/{}-tensor-dataset.nc'.format(output_dir, genus))\n",
    "\n",
    "ds\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e5fde11",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a2e8370b",
   "metadata": {},
   "source": [
    "## Batch corrected normalization\n",
    "\n",
    "As of February, 2023 this is not possible with the sctransform.vst function when using the \"v2\" version of the model (see [this issue](https://github.com/satijalab/sctransform/issues/126) for details). As the v2 version is the most appropriate for this data, batch correction is not possible at this time. Additionally, \n",
    "PCA and UMAP analysis of the data suggests that it may be difficult to disentangle batch effects from signal, since the batches are correlated with different sampling conditions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "292e0fd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # run the model on each clade, saving outputs and plots\n",
    "\n",
    "# # store normalized data as slabs to arrange in xarray tensor\n",
    "# slabs = {}\n",
    "# # directory to store output\n",
    "# output_dir = '../../data/3-normalization/batch-corrected'\n",
    "\n",
    "# # iterate through pro & syn\n",
    "# for genus in ['pro', 'syn']:\n",
    "#     print(genus)\n",
    "#     slabs[genus] = []\n",
    "        \n",
    "#     # iterate through clades\n",
    "#     for clade in clades[genus]:\n",
    "#         print(clade)\n",
    "\n",
    "#         # make fresh directory\n",
    "#         dir_path = '{}/{}'.format(output_dir, clade)\n",
    "#         if not os.path.isdir(dir_path):\n",
    "#             os.makedirs(dir_path)\n",
    "\n",
    "#         # convert clade_core_df to r matrix\n",
    "#         clade_core_df = core_df_dict[clade]\n",
    "#         r_clade_core_df = pandas_dataframe_to_r_matrix(clade_core_df)\n",
    "        \n",
    "#         # pull out sample attributes that match samples in clade_core_df\n",
    "#         sample_attr_df = sample_df.set_index('sample').loc[clade_core_df.columns, ['Dataset', 'Cruise']]\n",
    "        \n",
    "#         # convert pandas df to r dataframe\n",
    "#         r_sample_attr_df = pandas2ri.py2rpy(sample_attr_df)\n",
    "\n",
    "#         # fit vst normalization model\n",
    "#         result = sctransform.vst(\n",
    "#             r_clade_core_df, \n",
    "#             cell_attr=r_sample_attr_df, \n",
    "#             batch_var=ro.vectors.StrVector([\"Dataset\"]), \n",
    "#             n_genes='null', \n",
    "#             min_cells=detection_min,\n",
    "#             return_gene_attr=True, \n",
    "#             return_cell_attr=True, \n",
    "# #             vst_flavor='v2', \n",
    "#             verbosity=10\n",
    "#         )\n",
    "\n",
    "#         # save residuals as csv\n",
    "#         residual_df = pd.DataFrame(\n",
    "#             np.asarray(result[0]), \n",
    "#             index=clade_core_df.index, \n",
    "#             columns=clade_core_df.columns\n",
    "#         )\n",
    "#         residual_df.to_csv('{}/residuals_{}.csv'.format(dir_path, clade))\n",
    "        \n",
    "#         # save residuals as slab \n",
    "#         slab = (\n",
    "#             residual_df\n",
    "#             .reset_index()\n",
    "#             .rename(columns={'OrthologID': 'ortholog'})\n",
    "#             .melt(id_vars='ortholog', var_name='sample', value_name='residual')\n",
    "#         )\n",
    "#         slab['ortholog'] = slab['ortholog'].astype('Int64')\n",
    "#         slab['clade'] = clade\n",
    "#         slabs[genus].append(slab.set_index(['ortholog', 'clade', 'sample']))\n",
    "\n",
    "#         # save plot of model parameters\n",
    "#         plots = sctransform.plot_model_pars(result, show_theta=True)\n",
    "#         img = image_png(plots)\n",
    "#         with open('{}/parameters_{}.png'.format(dir_path, clade), 'wb') as png:\n",
    "#             png.write(img.data)\n",
    "\n",
    "#         # plot high variance genes\n",
    "#         residual_var = residual_df.var(axis=1)\n",
    "#         three_sigma = 1 + residual_var.var() * 3\n",
    "#         expression = clade_core_df.apply(geometric_mean, axis=1)\n",
    "#         plt.figure(figsize=(10, 4))\n",
    "#         sns.scatterplot(x=expression, y=residual_var, alpha=0.1);\n",
    "#         plt.hlines(three_sigma, expression.min(), expression.max(), \n",
    "#                    colors=['orange'], linestyles=[':'], label='3*sigma');\n",
    "#         plt.xlabel('mean gene abundance (reads/sample)')\n",
    "#         plt.xscale('log')\n",
    "#         plt.ylabel('residual variance')\n",
    "#         plt.title('normalized residual abundance of {} genes'.format(clade))\n",
    "#         plt.savefig('{}/residual_variance_{}.png'.format(dir_path, clade))\n",
    "#         plt.show()\n",
    "\n",
    "#         # save csv of residual variances with annotations\n",
    "#         res_var_df = residual_var.reset_index().rename(columns={0:'ResidualVariance'})\n",
    "#         res_var_df['Annotation'] = res_var_df.OrthologID.map(annotations_df.set_index('ortholog')['annotation'])\n",
    "#         res_var_df = res_var_df.sort_values('ResidualVariance', ascending=False).reset_index()\n",
    "#         res_var_df.to_csv('{}/high_variance_orthologs_{}.csv'.format(dir_path, clade))\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fe91913",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # make slabs into xarray tensors and save\n",
    "\n",
    "# # make DataArrays for metadata\n",
    "# replicate_da = xr.DataArray.from_series(sample_df.set_index('sample')['replicate'])\n",
    "# samplename_da = xr.DataArray.from_series(sample_df.set_index('sample')['samplename'])\n",
    "# annotation_da = xr.DataArray.from_series(annotations_df.set_index('ortholog')['annotation'])\n",
    "\n",
    "# for genus in ['pro', 'syn']:\n",
    "#     # make residual DataArray\n",
    "#     data = slabs[genus]\n",
    "#     all_slabs = pd.concat(data)\n",
    "#     residual_da = xr.DataArray.from_series(all_slabs['residual']).fillna(0)\n",
    "#     # combine residuals with metadata into xarray Dataset\n",
    "#     ds = xr.Dataset(\n",
    "#         dict(\n",
    "#             residual=residual_da, \n",
    "#             replicate=replicate_da,\n",
    "#             samplename=samplename_da, \n",
    "#             annotation=annotation_da.loc[residual_da.ortholog]\n",
    "#         )\n",
    "#     )\n",
    "#     # save as netCDF4 file in data directory\n",
    "#     ds.to_netcdf('{}/{}-res-abun.nc'.format(output_dir, genus))\n",
    "\n",
    "# ds\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22532611",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "902be528",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
