{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e6d7aab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import requests\n",
    "import json\n",
    "import xarray as xr\n",
    "import scipy.stats as stats\n",
    "\n",
    "from pathlib import Path\n",
    "from scipy.stats import mannwhitneyu\n",
    "from statsmodels.stats import multitest\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e628c05-b47f-4169-9695-153444df4e97",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "172b975b-c86b-4454-83b4-5d260b81da4d",
   "metadata": {},
   "source": [
    "# arrange metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "848c020a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenomeID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>Type</th>\n",
       "      <th>Group</th>\n",
       "      <th>Clade</th>\n",
       "      <th>Virocell</th>\n",
       "      <th>Completeness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2716884698</td>\n",
       "      <td>AG-316-L16</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>20.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2716884700</td>\n",
       "      <td>AG-316-N23</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>20.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2716884701</td>\n",
       "      <td>AG-316-P23</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>12.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2716884699</td>\n",
       "      <td>AG-316-L21</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>10.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2716884642</td>\n",
       "      <td>AG-316-A05</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>6.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>765</th>\n",
       "      <td>651703106</td>\n",
       "      <td>S-ShM2</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>N/A Virus</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>651703107</td>\n",
       "      <td>Syn19</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>N/A Virus</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>767</th>\n",
       "      <td>641201056</td>\n",
       "      <td>Syn5</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>N/A Virus</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>768</th>\n",
       "      <td>2595698410</td>\n",
       "      <td>metaG-MbCM1</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>N/A Virus</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>769</th>\n",
       "      <td>638275565</td>\n",
       "      <td>Syn9</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>N/A Virus</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>770 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       GenomeID   GenomeName     Type            Group      Clade  Virocell  \\\n",
       "0    2716884698   AG-316-L16      SAG  Prochlorococcus     AMZ-II     False   \n",
       "1    2716884700   AG-316-N23      SAG  Prochlorococcus     AMZ-II     False   \n",
       "2    2716884701   AG-316-P23      SAG  Prochlorococcus     AMZ-II     False   \n",
       "3    2716884699   AG-316-L21      SAG  Prochlorococcus     AMZ-II     False   \n",
       "4    2716884642   AG-316-A05      SAG  Prochlorococcus     AMZ-II     False   \n",
       "..          ...          ...      ...              ...        ...       ...   \n",
       "765   651703106       S-ShM2  ISOLATE            Virus  N/A Virus     False   \n",
       "766   651703107        Syn19  ISOLATE            Virus  N/A Virus     False   \n",
       "767   641201056         Syn5  ISOLATE            Virus  N/A Virus     False   \n",
       "768  2595698410  metaG-MbCM1  ISOLATE            Virus  N/A Virus     False   \n",
       "769   638275565         Syn9  ISOLATE            Virus  N/A Virus     False   \n",
       "\n",
       "     Completeness  \n",
       "0           20.69  \n",
       "1           20.69  \n",
       "2           12.07  \n",
       "3           10.34  \n",
       "4            6.90  \n",
       "..            ...  \n",
       "765          0.00  \n",
       "766          0.00  \n",
       "767          0.00  \n",
       "768          0.00  \n",
       "769          0.00  \n",
       "\n",
       "[770 rows x 7 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# genome metadata\n",
    "\n",
    "# filepaths\n",
    "filepath_genome_metadata = '../../data/metadata/genome-metadata.csv'\n",
    "filepath_ortholog_metadata = '../../data/metadata/ortholog-metadata.csv'\n",
    "\n",
    "# ortholog metadata\n",
    "ortho_df = pd.read_csv(filepath_ortholog_metadata)\n",
    "\n",
    "# genome metadata\n",
    "genome_df = pd.read_csv(filepath_genome_metadata)\n",
    "\n",
    "# edit the genome metadata for the purposes of this analysis\n",
    "genome_df.loc[genome_df['Group'].eq('Uncultured-marine-virus'), 'Group'] = 'Virus'\n",
    "genome_df.loc[genome_df['Group'].eq('Virus'), 'Clade'] = 'N/A Virus'\n",
    "genome_df.loc[genome_df['Virocell'], 'Group'] = 'Virocell'\n",
    "genome_df.loc[genome_df['Virocell'], 'Clade'] = 'N/A Virocell'\n",
    "\n",
    "genome_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c8d65dfc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MappingName</th>\n",
       "      <th>CyCOGID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>GenomeID</th>\n",
       "      <th>Group</th>\n",
       "      <th>Clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WH8102_2607658325</td>\n",
       "      <td>60000001</td>\n",
       "      <td>WH8102</td>\n",
       "      <td>2607658325</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "      <td>2606217514</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>5.1A-III</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIT0917_2681971350</td>\n",
       "      <td>60000001</td>\n",
       "      <td>MIT0917</td>\n",
       "      <td>2681971350</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "      <td>2681812859</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-424-P18_2717338506</td>\n",
       "      <td>60000001</td>\n",
       "      <td>AG-424-P18</td>\n",
       "      <td>2717338506</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "      <td>2716884419</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>HLII</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scB245a_521A19_2655604637</td>\n",
       "      <td>60000001</td>\n",
       "      <td>scB245a_521A19</td>\n",
       "      <td>2655604637</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "      <td>2654587735</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>HLII</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GFB01_2638208352</td>\n",
       "      <td>60000001</td>\n",
       "      <td>GFB01</td>\n",
       "      <td>2638208352</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "      <td>2636415834</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>5.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964917</th>\n",
       "      <td>AG-363-C02_2667889608</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667889608</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>2667527365</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLII.LLIII</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964918</th>\n",
       "      <td>AG-363-C02_2667889615</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667889615</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>2667527365</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLII.LLIII</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964919</th>\n",
       "      <td>AG-363-C02_2667890048</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667890048</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>2667527365</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLII.LLIII</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964920</th>\n",
       "      <td>AG-363-C02_2667890054</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667890054</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>2667527365</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLII.LLIII</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964921</th>\n",
       "      <td>AG-363-C02_2667890060</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667890060</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>2667527365</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>LLII.LLIII</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>964922 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      MappingName   CyCOGID      GenomeName      GeneID  \\\n",
       "0               WH8102_2607658325  60000001          WH8102  2607658325   \n",
       "1              MIT0917_2681971350  60000001         MIT0917  2681971350   \n",
       "2           AG-424-P18_2717338506  60000001      AG-424-P18  2717338506   \n",
       "3       scB245a_521A19_2655604637  60000001  scB245a_521A19  2655604637   \n",
       "4                GFB01_2638208352  60000001           GFB01  2638208352   \n",
       "...                           ...       ...             ...         ...   \n",
       "964917      AG-363-C02_2667889608  60040295      AG-363-C02  2667889608   \n",
       "964918      AG-363-C02_2667889615  60040295      AG-363-C02  2667889615   \n",
       "964919      AG-363-C02_2667890048  60040295      AG-363-C02  2667890048   \n",
       "964920      AG-363-C02_2667890054  60040295      AG-363-C02  2667890054   \n",
       "964921      AG-363-C02_2667890060  60040295      AG-363-C02  2667890060   \n",
       "\n",
       "                                      Annotation    GenomeID            Group  \\\n",
       "0       membrane protease FtsH catalytic subunit  2606217514    Synechococcus   \n",
       "1       membrane protease FtsH catalytic subunit  2681812859  Prochlorococcus   \n",
       "2       membrane protease FtsH catalytic subunit  2716884419  Prochlorococcus   \n",
       "3       membrane protease FtsH catalytic subunit  2654587735  Prochlorococcus   \n",
       "4       membrane protease FtsH catalytic subunit  2636415834    Synechococcus   \n",
       "...                                          ...         ...              ...   \n",
       "964917                      hypothetical protein  2667527365  Prochlorococcus   \n",
       "964918                      hypothetical protein  2667527365  Prochlorococcus   \n",
       "964919                      hypothetical protein  2667527365  Prochlorococcus   \n",
       "964920                      hypothetical protein  2667527365  Prochlorococcus   \n",
       "964921                      hypothetical protein  2667527365  Prochlorococcus   \n",
       "\n",
       "             Clade  \n",
       "0         5.1A-III  \n",
       "1              LLI  \n",
       "2             HLII  \n",
       "3             HLII  \n",
       "4              5.2  \n",
       "...            ...  \n",
       "964917  LLII.LLIII  \n",
       "964918  LLII.LLIII  \n",
       "964919  LLII.LLIII  \n",
       "964920  LLII.LLIII  \n",
       "964921  LLII.LLIII  \n",
       "\n",
       "[964922 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add genome id and clade info into ortholog df\n",
    "\n",
    "ortho_df['GenomeID'] = ortho_df['GenomeName'].map(genome_df.set_index('GenomeName')['GenomeID'])\n",
    "ortho_df['Group'] = ortho_df['GenomeName'].map(genome_df.set_index('GenomeName')['Group'])\n",
    "ortho_df['Clade'] = ortho_df['GenomeName'].map(genome_df.set_index('GenomeName')['Clade'])\n",
    "\n",
    "ortho_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a7add132",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CyCOGID</th>\n",
       "      <th>TotalRefs</th>\n",
       "      <th>Prochlorococcus</th>\n",
       "      <th>Synechococcus</th>\n",
       "      <th>Virocell</th>\n",
       "      <th>Virus</th>\n",
       "      <th>5.1A-CRD2</th>\n",
       "      <th>5.1A-II</th>\n",
       "      <th>5.1A-III</th>\n",
       "      <th>5.1A-IV</th>\n",
       "      <th>5.1A-UC-A-EnvC</th>\n",
       "      <th>5.1B-CRD1</th>\n",
       "      <th>5.1B-I</th>\n",
       "      <th>HLI</th>\n",
       "      <th>HLII</th>\n",
       "      <th>LLI</th>\n",
       "      <th>DescriptionCyCOG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>60000001</td>\n",
       "      <td>1376</td>\n",
       "      <td>1211</td>\n",
       "      <td>158</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>8</td>\n",
       "      <td>18</td>\n",
       "      <td>9</td>\n",
       "      <td>37</td>\n",
       "      <td>19</td>\n",
       "      <td>196</td>\n",
       "      <td>532</td>\n",
       "      <td>184</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>60000002</td>\n",
       "      <td>1453</td>\n",
       "      <td>1287</td>\n",
       "      <td>157</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>19</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "      <td>40</td>\n",
       "      <td>18</td>\n",
       "      <td>198</td>\n",
       "      <td>552</td>\n",
       "      <td>198</td>\n",
       "      <td>ATP-dependent Clp protease ATP-binding subunit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>60000003</td>\n",
       "      <td>1387</td>\n",
       "      <td>1222</td>\n",
       "      <td>156</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "      <td>21</td>\n",
       "      <td>8</td>\n",
       "      <td>41</td>\n",
       "      <td>14</td>\n",
       "      <td>203</td>\n",
       "      <td>523</td>\n",
       "      <td>192</td>\n",
       "      <td>ATP-dependent Clp protease proteolytic subunit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>60000004</td>\n",
       "      <td>1631</td>\n",
       "      <td>1369</td>\n",
       "      <td>255</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>24</td>\n",
       "      <td>16</td>\n",
       "      <td>27</td>\n",
       "      <td>26</td>\n",
       "      <td>62</td>\n",
       "      <td>29</td>\n",
       "      <td>173</td>\n",
       "      <td>542</td>\n",
       "      <td>265</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>60000005</td>\n",
       "      <td>990</td>\n",
       "      <td>877</td>\n",
       "      <td>108</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>25</td>\n",
       "      <td>13</td>\n",
       "      <td>128</td>\n",
       "      <td>383</td>\n",
       "      <td>131</td>\n",
       "      <td>chaperonin GroEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40290</th>\n",
       "      <td>60040291</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40291</th>\n",
       "      <td>60040292</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>Tryptophan-rich Synechocystis species C-termin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40292</th>\n",
       "      <td>60040293</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Putative transposase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40293</th>\n",
       "      <td>60040294</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>sulfate transport system substrate-binding pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40294</th>\n",
       "      <td>60040295</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40295 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CyCOGID  TotalRefs  Prochlorococcus  Synechococcus  Virocell  Virus  \\\n",
       "0      60000001       1376             1211            158         7      0   \n",
       "1      60000002       1453             1287            157         9      0   \n",
       "2      60000003       1387             1222            156         9      0   \n",
       "3      60000004       1631             1369            255         7      0   \n",
       "4      60000005        990              877            108         5      0   \n",
       "...         ...        ...              ...            ...       ...    ...   \n",
       "40290  60040291          4                4              0         0      0   \n",
       "40291  60040292          4                4              0         0      0   \n",
       "40292  60040293          5                0              5         0      0   \n",
       "40293  60040294          6                0              6         0      0   \n",
       "40294  60040295          6                6              0         0      0   \n",
       "\n",
       "       5.1A-CRD2  5.1A-II  5.1A-III  5.1A-IV  5.1A-UC-A-EnvC  5.1B-CRD1  \\\n",
       "0             17       17         8       18               9         37   \n",
       "1             10       19         9       16               9         40   \n",
       "2             12       16         9       21               8         41   \n",
       "3             22       24        16       27              26         62   \n",
       "4              8       11         5       14               7         25   \n",
       "...          ...      ...       ...      ...             ...        ...   \n",
       "40290          0        0         0        0               0          0   \n",
       "40291          0        0         0        0               0          0   \n",
       "40292          0        0         0        0               0          0   \n",
       "40293          0        0         0        0               0          0   \n",
       "40294          0        0         0        0               0          0   \n",
       "\n",
       "       5.1B-I  HLI  HLII  LLI  \\\n",
       "0          19  196   532  184   \n",
       "1          18  198   552  198   \n",
       "2          14  203   523  192   \n",
       "3          29  173   542  265   \n",
       "4          13  128   383  131   \n",
       "...       ...  ...   ...  ...   \n",
       "40290       0    0     0    4   \n",
       "40291       0    0     0    4   \n",
       "40292       0    0     0    0   \n",
       "40293       0    0     0    0   \n",
       "40294       0    0     0    0   \n",
       "\n",
       "                                        DescriptionCyCOG  \n",
       "0               membrane protease FtsH catalytic subunit  \n",
       "1      ATP-dependent Clp protease ATP-binding subunit...  \n",
       "2      ATP-dependent Clp protease proteolytic subunit...  \n",
       "3                                   hypothetical protein  \n",
       "4                                       chaperonin GroEL  \n",
       "...                                                  ...  \n",
       "40290                               hypothetical protein  \n",
       "40291  Tryptophan-rich Synechocystis species C-termin...  \n",
       "40292                               Putative transposase  \n",
       "40293  sulfate transport system substrate-binding pro...  \n",
       "40294                               hypothetical protein  \n",
       "\n",
       "[40295 rows x 17 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# start ko mapping file\n",
    "\n",
    "ko_map_df = ortho_df.groupby('CyCOGID').GeneID.count().reset_index().rename(columns={'GeneID': 'TotalRefs'})\n",
    "# add in genus counts\n",
    "ko_map_df = ko_map_df.join(\n",
    "    pd.DataFrame(ortho_df.groupby('CyCOGID')['Group'].value_counts()).rename(\n",
    "        columns={'Group': 'count'}).reset_index().pivot(\n",
    "        columns='Group', index='CyCOGID', values='count').fillna(0), \n",
    "    on='CyCOGID', \n",
    "    how='left'\n",
    ")\n",
    "# add in clade counts (only of select clades we deal with in this study)\n",
    "clades = ['HLI', 'HLII', 'LLI', '5.1B-I', '5.1B-CRD1', '5.1A-II', '5.1A-III', '5.1A-IV', '5.1A-CRD2', '5.1A-UC-A-EnvC']\n",
    "ko_map_df = ko_map_df.join(\n",
    "    pd.DataFrame(ortho_df[ortho_df.Clade.isin(clades)].groupby('CyCOGID')['Clade'].value_counts()).rename(\n",
    "        columns={'Clade': 'count'}).reset_index().pivot(\n",
    "        columns='Clade', index='CyCOGID', values='count').fillna(0), \n",
    "    on='CyCOGID', \n",
    "    how='left'\n",
    ").fillna(0)\n",
    "# convert dataframe back to ints\n",
    "ko_map_df = ko_map_df.astype(int)\n",
    "# add back CyCOG annotation\n",
    "ko_map_df['DescriptionCyCOG'] = ko_map_df['CyCOGID'].map(\n",
    "    ortho_df[['CyCOGID', 'Annotation']].drop_duplicates().set_index('CyCOGID')['Annotation']\n",
    ")\n",
    "\n",
    "ko_map_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41737ae1-fdf0-48c2-8db1-5715ee3f940b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "04b4fd2b-2e15-4033-bd9d-d4482b2d2b30",
   "metadata": {},
   "source": [
    "# import kegg annotations from genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "01a27cd8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 total missing directories\n"
     ]
    }
   ],
   "source": [
    "# check that there is a directory for each genome\n",
    "\n",
    "data_path = Path('../../data/refseqs/genomes/')\n",
    "\n",
    "count = 0\n",
    "for g_id in ortho_df['GenomeID'].unique():\n",
    "    if not os.path.isdir(data_path / str(g_id)):\n",
    "        print(f'{g_id} directory not found: {data_path / str(g_id)}')\n",
    "        count += 1\n",
    "        \n",
    "print(f'{count} total missing directories')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b6d3c6dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_oid</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>percent_identity</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subj_start</th>\n",
       "      <th>subj_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>ko_id</th>\n",
       "      <th>ko_name</th>\n",
       "      <th>EC</th>\n",
       "      <th>img_ko_flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2607658051</td>\n",
       "      <td>465</td>\n",
       "      <td>99.8</td>\n",
       "      <td>1</td>\n",
       "      <td>465</td>\n",
       "      <td>1</td>\n",
       "      <td>465</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>906.4</td>\n",
       "      <td>KO:K02313</td>\n",
       "      <td>chromosomal replication initiator protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2607658053</td>\n",
       "      <td>410</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>410</td>\n",
       "      <td>11</td>\n",
       "      <td>420</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>856.7</td>\n",
       "      <td>KO:K00799</td>\n",
       "      <td>glutathione S-transferase [EC:2.5.1.18]</td>\n",
       "      <td>EC:2.5.1.18</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2607658055</td>\n",
       "      <td>455</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>455</td>\n",
       "      <td>1</td>\n",
       "      <td>455</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>895.6</td>\n",
       "      <td>KO:K00383</td>\n",
       "      <td>glutathione reductase (NADPH) [EC:1.8.1.7]</td>\n",
       "      <td>EC:1.8.1.7</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2607658058</td>\n",
       "      <td>198</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>198</td>\n",
       "      <td>1</td>\n",
       "      <td>198</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>409.1</td>\n",
       "      <td>KO:K02276</td>\n",
       "      <td>cytochrome c oxidase subunit III [EC:1.9.3.1]</td>\n",
       "      <td>EC:1.9.3.1</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2607658059</td>\n",
       "      <td>562</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>562</td>\n",
       "      <td>1</td>\n",
       "      <td>562</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1173.7</td>\n",
       "      <td>KO:K02274</td>\n",
       "      <td>cytochrome c oxidase subunit I [EC:1.9.3.1]</td>\n",
       "      <td>EC:1.9.3.1</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492987</th>\n",
       "      <td>2610145142</td>\n",
       "      <td>243</td>\n",
       "      <td>35.7</td>\n",
       "      <td>19</td>\n",
       "      <td>239</td>\n",
       "      <td>48</td>\n",
       "      <td>257</td>\n",
       "      <td>4.300000e-27</td>\n",
       "      <td>123.6</td>\n",
       "      <td>KO:K03791</td>\n",
       "      <td>putative chitinase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492988</th>\n",
       "      <td>2610145144</td>\n",
       "      <td>309</td>\n",
       "      <td>37.5</td>\n",
       "      <td>4</td>\n",
       "      <td>292</td>\n",
       "      <td>17</td>\n",
       "      <td>311</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>198.4</td>\n",
       "      <td>KO:K03086</td>\n",
       "      <td>RNA polymerase primary sigma factor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492989</th>\n",
       "      <td>2610145149</td>\n",
       "      <td>109</td>\n",
       "      <td>41.2</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "      <td>97</td>\n",
       "      <td>1.500000e-13</td>\n",
       "      <td>77.4</td>\n",
       "      <td>KO:K03111</td>\n",
       "      <td>single-strand DNA-binding protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492990</th>\n",
       "      <td>2610145152</td>\n",
       "      <td>261</td>\n",
       "      <td>40.5</td>\n",
       "      <td>5</td>\n",
       "      <td>254</td>\n",
       "      <td>10</td>\n",
       "      <td>269</td>\n",
       "      <td>2.000000e-43</td>\n",
       "      <td>177.9</td>\n",
       "      <td>KO:K10906</td>\n",
       "      <td>exodeoxyribonuclease VIII [EC:3.1.11.-]</td>\n",
       "      <td>EC:3.1.11.-</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492991</th>\n",
       "      <td>2610145160</td>\n",
       "      <td>139</td>\n",
       "      <td>35.6</td>\n",
       "      <td>6</td>\n",
       "      <td>135</td>\n",
       "      <td>7</td>\n",
       "      <td>123</td>\n",
       "      <td>5.100000e-09</td>\n",
       "      <td>62.8</td>\n",
       "      <td>KO:K01160</td>\n",
       "      <td>crossover junction endodeoxyribonuclease RusA ...</td>\n",
       "      <td>EC:3.1.22.4</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>492992 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_oid  gene_length  percent_identity  query_start  query_end  \\\n",
       "0       2607658051          465              99.8            1        465   \n",
       "1       2607658053          410             100.0            1        410   \n",
       "2       2607658055          455             100.0            1        455   \n",
       "3       2607658058          198             100.0            1        198   \n",
       "4       2607658059          562             100.0            1        562   \n",
       "...            ...          ...               ...          ...        ...   \n",
       "492987  2610145142          243              35.7           19        239   \n",
       "492988  2610145144          309              37.5            4        292   \n",
       "492989  2610145149          109              41.2            1         95   \n",
       "492990  2610145152          261              40.5            5        254   \n",
       "492991  2610145160          139              35.6            6        135   \n",
       "\n",
       "        subj_start  subj_end        evalue  bit_score      ko_id  \\\n",
       "0                1       465  0.000000e+00      906.4  KO:K02313   \n",
       "1               11       420  0.000000e+00      856.7  KO:K00799   \n",
       "2                1       455  0.000000e+00      895.6  KO:K00383   \n",
       "3                1       198  0.000000e+00      409.1  KO:K02276   \n",
       "4                1       562  0.000000e+00     1173.7  KO:K02274   \n",
       "...            ...       ...           ...        ...        ...   \n",
       "492987          48       257  4.300000e-27      123.6  KO:K03791   \n",
       "492988          17       311  0.000000e+00      198.4  KO:K03086   \n",
       "492989           1        97  1.500000e-13       77.4  KO:K03111   \n",
       "492990          10       269  2.000000e-43      177.9  KO:K10906   \n",
       "492991           7       123  5.100000e-09       62.8  KO:K01160   \n",
       "\n",
       "                                                  ko_name           EC  \\\n",
       "0               chromosomal replication initiator protein          NaN   \n",
       "1                 glutathione S-transferase [EC:2.5.1.18]  EC:2.5.1.18   \n",
       "2              glutathione reductase (NADPH) [EC:1.8.1.7]   EC:1.8.1.7   \n",
       "3           cytochrome c oxidase subunit III [EC:1.9.3.1]   EC:1.9.3.1   \n",
       "4             cytochrome c oxidase subunit I [EC:1.9.3.1]   EC:1.9.3.1   \n",
       "...                                                   ...          ...   \n",
       "492987                                 putative chitinase          NaN   \n",
       "492988                RNA polymerase primary sigma factor          NaN   \n",
       "492989                  single-strand DNA-binding protein          NaN   \n",
       "492990            exodeoxyribonuclease VIII [EC:3.1.11.-]  EC:3.1.11.-   \n",
       "492991  crossover junction endodeoxyribonuclease RusA ...  EC:3.1.22.4   \n",
       "\n",
       "       img_ko_flag  \n",
       "0              Yes  \n",
       "1              Yes  \n",
       "2              Yes  \n",
       "3              Yes  \n",
       "4              Yes  \n",
       "...            ...  \n",
       "492987         Yes  \n",
       "492988         Yes  \n",
       "492989         Yes  \n",
       "492990         Yes  \n",
       "492991         Yes  \n",
       "\n",
       "[492992 rows x 13 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import all kegg annotations as a df\n",
    "\n",
    "ko_df = pd.DataFrame()\n",
    "for g_id in ortho_df['GenomeID'].unique():\n",
    "    df = pd.read_csv(data_path / f'{g_id}/{g_id}.ko.tab.txt', sep='\\t')\n",
    "    if len(ko_df) == 0:\n",
    "        ko_df = df\n",
    "    else:\n",
    "        ko_df = pd.concat([ko_df, df])\n",
    "        \n",
    "ko_df = ko_df.reset_index(drop=True)\n",
    "\n",
    "ko_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38574ab5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "03506f9f",
   "metadata": {},
   "source": [
    "# Deduplicate gene annotations\n",
    "\n",
    "**Problem:** Some genes (specific nucleotide sequence from particular reference genome) have more than one KO annotation hit. \n",
    "- Based on the counts below, this comes out to 25,218 (23227 + 1973 + 15 + 2 + 1) out of 492,992 annotated genes, or ~5.12%. \n",
    "- Most of these are annotations that actually have the same KO number, but just a different EC number or something. Only 3,375 genes (0.68%) have multiple distinct KO annotations\n",
    "- Of these, qualitatively it seems like the annotations are usually quite similar to one another (e.g. same pathway, or different subunits of the same protein complex)\n",
    "\n",
    "**Implemented solution:** Since the KO annotations are based on an HMM search, first select the annotation with the lowest E-value. Then if the E-values are exactly the same, randomly select one annotaiton or the other\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "65788b15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count\n",
      "1    440542\n",
      "2     23227\n",
      "3      1973\n",
      "4        15\n",
      "6         2\n",
      "5         1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# how many genes have more than one ko number assigned?\n",
    "\n",
    "print(ko_df['gene_oid'].value_counts().value_counts())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e0ba6310",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ko_id\n",
      "1    21836\n",
      "2     3333\n",
      "3       48\n",
      "4        1\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_oid</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>percent_identity</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subj_start</th>\n",
       "      <th>subj_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>ko_id</th>\n",
       "      <th>ko_name</th>\n",
       "      <th>EC</th>\n",
       "      <th>img_ko_flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1549</th>\n",
       "      <td>2681971726</td>\n",
       "      <td>240</td>\n",
       "      <td>93.75</td>\n",
       "      <td>1</td>\n",
       "      <td>240</td>\n",
       "      <td>1</td>\n",
       "      <td>240</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>513.0</td>\n",
       "      <td>KO:K06182</td>\n",
       "      <td>23S rRNA pseudouridine2604 synthase [EC:5.4.99...</td>\n",
       "      <td>EC:5.4.99.21</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1550</th>\n",
       "      <td>2681971726</td>\n",
       "      <td>240</td>\n",
       "      <td>93.75</td>\n",
       "      <td>1</td>\n",
       "      <td>240</td>\n",
       "      <td>1</td>\n",
       "      <td>240</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>513.0</td>\n",
       "      <td>KO:K06183</td>\n",
       "      <td>16S rRNA pseudouridine516 synthase [EC:5.4.99.19]</td>\n",
       "      <td>EC:5.4.99.19</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1561</th>\n",
       "      <td>2681971798</td>\n",
       "      <td>467</td>\n",
       "      <td>89.08</td>\n",
       "      <td>1</td>\n",
       "      <td>467</td>\n",
       "      <td>1</td>\n",
       "      <td>467</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>960.0</td>\n",
       "      <td>KO:K04094</td>\n",
       "      <td>methylenetetrahydrofolate--tRNA-(uracil-5-)-me...</td>\n",
       "      <td>EC:2.1.1.74</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1562</th>\n",
       "      <td>2681971798</td>\n",
       "      <td>467</td>\n",
       "      <td>90.15</td>\n",
       "      <td>1</td>\n",
       "      <td>467</td>\n",
       "      <td>1</td>\n",
       "      <td>467</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>969.0</td>\n",
       "      <td>KO:K03495</td>\n",
       "      <td>tRNA uridine 5-carboxymethylaminomethyl modifi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1620</th>\n",
       "      <td>2681971886</td>\n",
       "      <td>348</td>\n",
       "      <td>80.77</td>\n",
       "      <td>1</td>\n",
       "      <td>338</td>\n",
       "      <td>1</td>\n",
       "      <td>338</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>680.0</td>\n",
       "      <td>KO:K08919</td>\n",
       "      <td>chlorophyll a/b binding light-harvesting prote...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492225</th>\n",
       "      <td>2717343036</td>\n",
       "      <td>149</td>\n",
       "      <td>41.86</td>\n",
       "      <td>15</td>\n",
       "      <td>140</td>\n",
       "      <td>4</td>\n",
       "      <td>132</td>\n",
       "      <td>6.500000e-31</td>\n",
       "      <td>126.0</td>\n",
       "      <td>KO:K00980</td>\n",
       "      <td>glycerol-3-phosphate cytidylyltransferase [EC:...</td>\n",
       "      <td>EC:2.7.7.39</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492226</th>\n",
       "      <td>2717343036</td>\n",
       "      <td>149</td>\n",
       "      <td>44.78</td>\n",
       "      <td>15</td>\n",
       "      <td>145</td>\n",
       "      <td>4</td>\n",
       "      <td>137</td>\n",
       "      <td>3.000000e-36</td>\n",
       "      <td>141.0</td>\n",
       "      <td>KO:K03272</td>\n",
       "      <td>D-beta-D-heptose 7-phosphate kinase / D-beta-D...</td>\n",
       "      <td>EC:2.7.1.167</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492227</th>\n",
       "      <td>2717343036</td>\n",
       "      <td>149</td>\n",
       "      <td>44.78</td>\n",
       "      <td>15</td>\n",
       "      <td>145</td>\n",
       "      <td>4</td>\n",
       "      <td>137</td>\n",
       "      <td>3.000000e-36</td>\n",
       "      <td>141.0</td>\n",
       "      <td>KO:K03272</td>\n",
       "      <td>D-beta-D-heptose 7-phosphate kinase / D-beta-D...</td>\n",
       "      <td>EC:2.7.7.70</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492269</th>\n",
       "      <td>2717343185</td>\n",
       "      <td>112</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1</td>\n",
       "      <td>112</td>\n",
       "      <td>1</td>\n",
       "      <td>112</td>\n",
       "      <td>1.400000e-73</td>\n",
       "      <td>243.0</td>\n",
       "      <td>KO:K04751</td>\n",
       "      <td>nitrogen regulatory protein P-II 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492270</th>\n",
       "      <td>2717343185</td>\n",
       "      <td>112</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1</td>\n",
       "      <td>112</td>\n",
       "      <td>1</td>\n",
       "      <td>112</td>\n",
       "      <td>1.400000e-73</td>\n",
       "      <td>243.0</td>\n",
       "      <td>KO:K04752</td>\n",
       "      <td>nitrogen regulatory protein P-II 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6881 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_oid  gene_length  percent_identity  query_start  query_end  \\\n",
       "1549    2681971726          240             93.75            1        240   \n",
       "1550    2681971726          240             93.75            1        240   \n",
       "1561    2681971798          467             89.08            1        467   \n",
       "1562    2681971798          467             90.15            1        467   \n",
       "1620    2681971886          348             80.77            1        338   \n",
       "...            ...          ...               ...          ...        ...   \n",
       "492225  2717343036          149             41.86           15        140   \n",
       "492226  2717343036          149             44.78           15        145   \n",
       "492227  2717343036          149             44.78           15        145   \n",
       "492269  2717343185          112            100.00            1        112   \n",
       "492270  2717343185          112            100.00            1        112   \n",
       "\n",
       "        subj_start  subj_end        evalue  bit_score      ko_id  \\\n",
       "1549             1       240  0.000000e+00      513.0  KO:K06182   \n",
       "1550             1       240  0.000000e+00      513.0  KO:K06183   \n",
       "1561             1       467  0.000000e+00      960.0  KO:K04094   \n",
       "1562             1       467  0.000000e+00      969.0  KO:K03495   \n",
       "1620             1       338  0.000000e+00      680.0  KO:K08919   \n",
       "...            ...       ...           ...        ...        ...   \n",
       "492225           4       132  6.500000e-31      126.0  KO:K00980   \n",
       "492226           4       137  3.000000e-36      141.0  KO:K03272   \n",
       "492227           4       137  3.000000e-36      141.0  KO:K03272   \n",
       "492269           1       112  1.400000e-73      243.0  KO:K04751   \n",
       "492270           1       112  1.400000e-73      243.0  KO:K04752   \n",
       "\n",
       "                                                  ko_name            EC  \\\n",
       "1549    23S rRNA pseudouridine2604 synthase [EC:5.4.99...  EC:5.4.99.21   \n",
       "1550    16S rRNA pseudouridine516 synthase [EC:5.4.99.19]  EC:5.4.99.19   \n",
       "1561    methylenetetrahydrofolate--tRNA-(uracil-5-)-me...   EC:2.1.1.74   \n",
       "1562    tRNA uridine 5-carboxymethylaminomethyl modifi...           NaN   \n",
       "1620    chlorophyll a/b binding light-harvesting prote...           NaN   \n",
       "...                                                   ...           ...   \n",
       "492225  glycerol-3-phosphate cytidylyltransferase [EC:...   EC:2.7.7.39   \n",
       "492226  D-beta-D-heptose 7-phosphate kinase / D-beta-D...  EC:2.7.1.167   \n",
       "492227  D-beta-D-heptose 7-phosphate kinase / D-beta-D...   EC:2.7.7.70   \n",
       "492269                 nitrogen regulatory protein P-II 1           NaN   \n",
       "492270                 nitrogen regulatory protein P-II 2           NaN   \n",
       "\n",
       "       img_ko_flag  \n",
       "1549           Yes  \n",
       "1550           Yes  \n",
       "1561           Yes  \n",
       "1562           Yes  \n",
       "1620           Yes  \n",
       "...            ...  \n",
       "492225         Yes  \n",
       "492226         Yes  \n",
       "492227         Yes  \n",
       "492269         Yes  \n",
       "492270         Yes  \n",
       "\n",
       "[6881 rows x 13 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# deeper look: examine genes with more than one annotation\n",
    "\n",
    "counts = ko_df['gene_oid'].value_counts()\n",
    "ko_counts = ko_df[ko_df['gene_oid'].isin(counts[counts.gt(1)].index)].groupby('gene_oid')['ko_id'].nunique()\n",
    "print(ko_counts.value_counts())    # most have the same ko id, just different EC numbers or something\n",
    "\n",
    "ko_df[ko_df['gene_oid'].isin(ko_counts[ko_counts.gt(1)].index)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4759881e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# deduplicate genes with more than one annotation\n",
    "\n",
    "# for each gene_oid's set of annotations, select the one with the lowest e-value\n",
    "# this step will also randomly select one of the annotations to propogate in cases with the same KO \n",
    "# but different EC numbers, or different KOs but same e-value\n",
    "\n",
    "ko_df = ko_df.loc[ko_df.groupby('gene_oid')['evalue'].idxmin(), :].reset_index(drop=True)\n",
    "ko_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93c68d51",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b25a9231",
   "metadata": {},
   "source": [
    "# Deduplicate CyCOG annotations\n",
    "\n",
    "**Problem:** Some CyCOG (Clusters of Orthologous Genes) consist of genes with discordant KO annotations\n",
    "- Out of 40,295 orthologous gene families in CyCOGs v6, 36,693 no KO annotation, meaning just 3,602 (8.94%) contain at least one gene with a KO annotation.\n",
    "- In most of the annotated CyCOGs, all gene members had a concordant KO annotation. In 128 (3.55% of annotated CyCOGs), the KO annotations were discordant among gene members.\n",
    "- Of these, qualitatively it seems like the discordant annotations are usually quite similar to one another (e.g. same pathway, or different subunits of the same protein complex)\n",
    "\n",
    "**Implemented solution:** For each CyCOG, select the annotation applied to the majority of members as the representatitve CyCOG annotation. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98f5858a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# join kegg annotations onto reference gene set\n",
    "\n",
    "annot_df = pd.merge(ortho_df, ko_df, left_on='GeneID', right_on='gene_oid', how='left')\n",
    "\n",
    "# how many unique ko annotations per ortholog group?\n",
    "print(annot_df.groupby('CyCOGID').ko_id.nunique().value_counts())\n",
    "\n",
    "# drop the sequences without a ko annotation\n",
    "annot_df = annot_df[annot_df['ko_id'].notna()]\n",
    "\n",
    "annot_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70eeed9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# look at non-unique ko mappings\n",
    "\n",
    "counts = annot_df.groupby('CyCOGID').ko_id.nunique()\n",
    "collisions_df = annot_df[annot_df['CyCOGID'].isin(counts[counts.gt(1)].index)].groupby('CyCOGID')[\n",
    "    ['ko_id', 'ko_name']].value_counts().reset_index()\n",
    "\n",
    "pd.set_option('max_colwidth', None)\n",
    "collisions_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82c8468c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# deduplicate the discordant ortholog-ko mapping by simple majority vote\n",
    "\n",
    "ko_count_df = pd.DataFrame(annot_df.groupby('CyCOGID').ko_id.value_counts()).rename(\n",
    "    columns={'ko_id': 'count'}).reset_index()\n",
    "ko_count_df = ko_count_df.loc[ko_count_df.groupby('CyCOGID')['count'].idxmax(), :]\n",
    "\n",
    "# add in KO annotation\n",
    "name_map = annot_df[['ko_id', 'ko_name']].drop_duplicates()\n",
    "name_map['DescriptionKO'] = name_map['ko_name'].str.split(' \\[EC:').str[0]\n",
    "name_map = name_map.loc[name_map['ko_id'].drop_duplicates().index]    # pick one description from duplicates\n",
    "ko_count_df = pd.merge(left=ko_count_df, right=name_map[['ko_id', 'DescriptionKO']], on='ko_id', how='left')\n",
    "\n",
    "# remove prefix from KOID\n",
    "ko_count_df['KOID'] = ko_count_df['ko_id'].str[3:]\n",
    "# rename count column and drop old ko_id column\n",
    "ko_count_df = ko_count_df[['CyCOGID', 'KOID', 'DescriptionKO', 'count']].rename(\n",
    "    columns={'count': 'NRefsKO'}).set_index('CyCOGID')\n",
    "# calculate refs with KO mapping other than the one chosen\n",
    "ko_count_df['NRefsOtherKO'] = annot_df.groupby('CyCOGID').GeneID.count() - ko_count_df['NRefsKO']\n",
    "\n",
    "# add ko_count_df into ko_map_df\n",
    "ko_map_df = pd.merge(ko_map_df, ko_count_df, on='CyCOGID', how='left')\n",
    "\n",
    "pd.reset_option('max_colwidth')\n",
    "ko_map_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "392f86f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# there are many different CyCOG orthologies annotated with the same KO group -- look at some of these\n",
    "\n",
    "ko_map_counts = ko_map_df['KOID'].value_counts()\n",
    "print('There are {} of {} KOs with a non-unique CyCOGID-KOID mapping:'.format(\n",
    "    len(ko_map_counts[ko_map_counts.gt(1)]), ko_map_df['KOID'].nunique()))\n",
    "print(ko_map_counts[ko_map_counts.gt(1)].head(20))\n",
    "\n",
    "ko_map_df[ko_map_df['KOID'] == 'K06147'].head(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "640eb234",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7c6e9353",
   "metadata": {},
   "source": [
    "# Get KEGG pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "596172e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get full reference brite hierarchy of kegg orthologies \n",
    "\n",
    "url = f'http://rest.kegg.jp/get/br:ko00001/json'\n",
    "file = requests.get(url, allow_redirects=True)\n",
    "\n",
    "# url = f'https://rest.kegg.jp/list/ko/'\n",
    "# file = requests.get(url, allow_redirects=True)\n",
    "\n",
    "file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03225e7a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# parse pathways into dictionary\n",
    "\n",
    "pathway_dict = {}\n",
    "pathway_names = {}\n",
    "ko_dict = {}\n",
    "ko_names = {}\n",
    "\n",
    "# exclude pathways specific to eukaryotes, multicellular organisms, and humans\n",
    "excluded_systems = ['09144 Cellular community - eukaryotes', '09150 Organismal Systems', '09160 Human Diseases']\n",
    "\n",
    "# iterate through json\n",
    "for supersystem in json.loads(file.content)['children']:\n",
    "    print(f'{supersystem[\"name\"]}')\n",
    "    for system in supersystem['children']:\n",
    "        print(f'\\t{system[\"name\"]}')\n",
    "        # skip over excluded systems (pathways specific to eukaryotes, multicellular organisms, and humans)\n",
    "        if supersystem[\"name\"] in excluded_systems or system[\"name\"] in excluded_systems:\n",
    "            print('\\t\\tEXCLUDED')\n",
    "            continue\n",
    "        for pathway in system['children']:\n",
    "            path_id = f'ko{pathway[\"name\"].split(\" \")[0]}'\n",
    "            path_members = []\n",
    "            if 'children' in pathway.keys():\n",
    "                # parse ko numbers of each pathway\n",
    "                for ko in pathway['children']:\n",
    "                    ko_id = ko['name'].split(' ')[0]\n",
    "                    ko_names[ko_id] = ' '.join(ko['name'].split(' ')[1:]).strip(' ')\n",
    "                    path_members.append(ko_id)\n",
    "                    # add pathway to ko_dict\n",
    "                    if ko_id in ko_dict.keys():\n",
    "                        ko_dict[ko_id].append(path_id)\n",
    "                    else:\n",
    "                        ko_dict[ko_id] = [path_id]\n",
    "                pathway_dict[path_id] = path_members\n",
    "                pathway_names[path_id] = pathway[\"name\"][6:]\n",
    "            print(f'\\t\\t{path_id}: {pathway[\"name\"]} ({len(path_members)})')\n",
    "      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31741485",
   "metadata": {},
   "outputs": [],
   "source": [
    "# There are some KO descriptions from the most recent KEGG download that do not match the descriptions \n",
    "# of the old annotation. A rough screen shows 243/1752. I want to look at these in more detail\n",
    "\n",
    "name_df = ko_map_df[['KOID', 'DescriptionKO']].drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "names = []\n",
    "substrings = []\n",
    "for i, koid in enumerate(name_df['KOID']):\n",
    "    # append KEGG names\n",
    "    if koid in ko_names.keys():\n",
    "        name = ko_names[koid]\n",
    "    else:\n",
    "        name = np.nan\n",
    "    names.append(name)\n",
    "    # check if previous KO Description is a substring of current KEGG names\n",
    "    substrings.append(str(name_df.loc[i, 'DescriptionKO']) in str(name))\n",
    "# add to dataframe  \n",
    "name_df['KEGGNames'] = names\n",
    "name_df['SubstringMatch'] = substrings\n",
    "\n",
    "# # fix display options\n",
    "# pd.set_option('display.max_rows', None)\n",
    "# pd.set_option('max_colwidth', None)\n",
    "\n",
    "# # scren all those that don't have a match\n",
    "# name_df[~name_df['SubstringMatch']]\n",
    "\n",
    "# manual inspection of some incongruent (updated?) KOs:\n",
    "problem_kos = ['K02500', 'K02428', 'K02259', 'K01594', 'K00870', 'K02501', 'K05808', 'K03606', 'K01144', \n",
    "               'K03186', 'K01234', 'K03651', 'K07011', 'K03082', 'K01003', 'K05663', 'K03152']\n",
    "name_df[name_df['KOID'].isin(problem_kos)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83669a1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# match up pathway information to pangenome ko list\n",
    "\n",
    "pathway_df = ko_map_df[ko_map_df['KOID'].notna()].reset_index(drop=True)\n",
    "\n",
    "data_dict = {}\n",
    "for koid in pathway_df['KOID'].unique():\n",
    "    if koid in ko_dict.keys():\n",
    "        pathways = ko_dict[koid]\n",
    "    else:\n",
    "        continue\n",
    "    for pathway in pathways:\n",
    "        if pathway not in data_dict.keys():\n",
    "            data_dict[pathway] = pathway_df['KOID'].eq(koid).astype(int)\n",
    "        else:\n",
    "            data_dict[pathway] = data_dict[pathway] + pathway_df['KOID'].eq(koid).astype(int)\n",
    "\n",
    "pathway_df = pd.merge(pathway_df, pd.DataFrame(data_dict), left_index=True, right_index=True)\n",
    "pathway_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e86ba9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# look at pathway statistics and compile dictionary of CyCOGs associated with each pathway\n",
    "\n",
    "cycog_pathway_dict = {}\n",
    "\n",
    "for path, cycog_count in pd.DataFrame(data_dict).sum().sort_values(ascending=False).items():\n",
    "    ko_count = pathway_df.loc[pathway_df[path].gt(0), 'KOID'].nunique()\n",
    "    total_kos = len(pathway_dict[path])\n",
    "    print(f'{path}: {pathway_names[path]}\\n\\t{cycog_count} CyCOGs representing {ko_count}/{total_kos} unique KOs')\n",
    "    # append list of cycogs associated with each pathway to dictionary\n",
    "    cycog_pathway_dict[path] = pathway_df.loc[pathway_df[path].gt(0), 'CyCOGID'].to_list()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cd2f152",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "13e87350",
   "metadata": {},
   "source": [
    "# Calculate Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f2b5271",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up parameters and data\n",
    "\n",
    "# significance level\n",
    "alpha = 0.01\n",
    "enrichments_df = pd.DataFrame()\n",
    "\n",
    "# helper function for fdrcorrection\n",
    "def adjust_pvals(pvals):\n",
    "    return multitest.fdrcorrection(pvals, alpha=alpha)[1]\n",
    "\n",
    "# read in cluster data\n",
    "data_dir = Path('../../data')\n",
    "pro_ds = xr.open_dataset(data_dir / '5-models/pro-aligned-models.nc')\n",
    "syn_ds = xr.open_dataset(data_dir / '5-models/syn-aligned-models.nc')\n",
    "out_dir = data_dir / '7-interpretation'\n",
    "if not out_dir.is_dir():\n",
    "    out_dir.mkdir(parents=True)\n",
    "\n",
    "# save pathway_df to out_dir\n",
    "pathway_df.to_csv(out_dir / 'cycog-kegg-pathways.csv', index=False)\n",
    "\n",
    "# analyze only \"robust\" clusters\n",
    "pro_clusters = [\n",
    "    'pro1', 'pro2', 'pro3', 'pro4', 'pro5', 'pro6', 'pro7', 'pro8', 'pro9', 'pro10', 'pro11', 'pro12', 'pro13', 'pro14', 'pro15'\n",
    "]\n",
    "syn_clusters = [\n",
    "    'syn1', 'syn2', 'syn3', 'syn4', 'syn5', 'syn6', 'syn7', 'syn8', 'syn9', 'syn10', 'syn11', 'syn12', 'syn13', 'syn14', 'syn15'\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bce00e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull out median bootstrap weights\n",
    "\n",
    "# pull out pro gene weight data\n",
    "pro_gene_df = pro_ds.median(dim=['Bootstrap', 'Replicate']).GeneWeight.T.to_pandas()\n",
    "pro_gene_df = pro_gene_df.rename(columns=dict(zip(pro_gene_df.columns, [f\"pro{i}\" for i in pro_gene_df.columns])))\n",
    "# select only robust components\n",
    "pro_gene_df = pro_gene_df.loc[:, pro_clusters]\n",
    "\n",
    "# pull out syn gene weight data\n",
    "syn_gene_df = syn_ds.median(dim=['Bootstrap', 'Replicate']).GeneWeight.T.to_pandas()\n",
    "syn_gene_df = syn_gene_df.rename(columns=dict(zip(syn_gene_df.columns, [f\"syn{i}\" for i in syn_gene_df.columns])))\n",
    "# select only robust components\n",
    "syn_gene_df = syn_gene_df.loc[:, syn_clusters]\n",
    "\n",
    "syn_gene_df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65395397",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run Mann-Whitney U test for every pathway against every pro cluster\n",
    "\n",
    "input_df = pro_gene_df\n",
    "\n",
    "p_vals = []\n",
    "for pathway in cycog_pathway_dict.keys():\n",
    "    x = input_df.loc[input_df.index.isin(cycog_pathway_dict[pathway]), :]\n",
    "    y = input_df.loc[~input_df.index.isin(cycog_pathway_dict[pathway]), :]\n",
    "    result = mannwhitneyu(x, y, alternative='greater', axis=0, method='asymptotic')\n",
    "    p_vals.append(result.pvalue)\n",
    "pro_enrich_df = pd.DataFrame(p_vals, index=cycog_pathway_dict.keys(), columns=input_df.columns)\n",
    "# drop the pathways that are all NAs\n",
    "pro_enrich_df = pro_enrich_df[~pro_enrich_df.isna().all(axis=1)]\n",
    "# adjust p-values\n",
    "pro_enrich_df = pro_enrich_df.apply(adjust_pvals, raw=True)\n",
    "\n",
    "pro_enrich_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51e43897",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# list the most enriched pathways for each cluster\n",
    "\n",
    "for cluster in pro_enrich_df.columns:\n",
    "    pathways = pro_enrich_df[pro_enrich_df[cluster].lt(alpha)][cluster].sort_values(ascending=True)\n",
    "    print(f'\\n{cluster} ({len(pathways)} enriched pathways)')\n",
    "    for path, pval in pathways.items():\n",
    "        enrichments_df.loc[path, 'description'] = pathway_names[path]\n",
    "        enrichments_df.loc[path, cluster] = pval\n",
    "        print(f'\\t{path} (p={pval:.2e}): {pathway_names[path]}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c65b66c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run Mann-Whitney U test for every pathway against every syn cluster\n",
    "\n",
    "input_df = syn_gene_df\n",
    "\n",
    "p_vals = []\n",
    "for pathway in cycog_pathway_dict.keys():\n",
    "    x = input_df.loc[input_df.index.isin(cycog_pathway_dict[pathway]), :]\n",
    "    y = input_df.loc[~input_df.index.isin(cycog_pathway_dict[pathway]), :]\n",
    "    result = mannwhitneyu(x, y, alternative='greater', axis=0, method='asymptotic')\n",
    "    p_vals.append(result.pvalue)\n",
    "syn_enrich_df = pd.DataFrame(p_vals, index=cycog_pathway_dict.keys(), columns=input_df.columns)\n",
    "# drop the pathways that are all NAs\n",
    "syn_enrich_df = syn_enrich_df[~syn_enrich_df.isna().all(axis=1)]\n",
    "# adjust p-values\n",
    "syn_enrich_df = syn_enrich_df.apply(adjust_pvals, raw=True)\n",
    "\n",
    "syn_enrich_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d03ab60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# list the most enriched pathways for each cluster\n",
    "\n",
    "for cluster in syn_enrich_df.columns:\n",
    "    pathways = syn_enrich_df[syn_enrich_df[cluster].lt(alpha)][cluster].sort_values(ascending=True)\n",
    "    print(f'\\n{cluster} ({len(pathways)} enriched pathways)')\n",
    "    for path, pval in pathways.items():\n",
    "        enrichments_df.loc[path, 'description'] = pathway_names[path]\n",
    "        enrichments_df.loc[path, cluster] = pval\n",
    "        print(f'\\t{path} (p={pval:.2E}): {pathway_names[path]}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60d5316e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reformat and save enrichments dataframe\n",
    "\n",
    "enrichments_df = enrichments_df.sort_index().reset_index().rename(columns={'index': 'KEGG pathway'})\n",
    "enrichments_df.to_csv(out_dir / 'kegg-enrichments.csv', index=False)\n",
    "\n",
    "enrichments_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68729e93-8f0b-4695-8ba2-a7248d98b66b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
