{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af9a6f78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.colors as mcolors\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import pycmap\n",
    "import seaborn as sns\n",
    "from tlviz.factor_tools import factor_match_score\n",
    "import xarray as xr\n",
    "import warnings\n",
    "\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b574290-0e6c-4857-bca6-4d77ff44cbfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions to build continuous colormap\n",
    "# source: https://towardsdatascience.com/beautiful-custom-colormaps-with-matplotlib-5bab3d1f0e72\n",
    "\n",
    "def hex_to_rgb(value):\n",
    "    '''\n",
    "    Converts hex to rgb colours\n",
    "    value: string of 6 characters representing a hex colour.\n",
    "    Returns: list length 3 of RGB values'''\n",
    "    value = value.strip(\"#\") # removes hash symbol if present\n",
    "    lv = len(value)\n",
    "    return tuple(int(value[i:i + lv // 3], 16) for i in range(0, lv, lv // 3))\n",
    "\n",
    "\n",
    "def rgb_to_dec(value):\n",
    "    '''\n",
    "    Converts rgb to decimal colours (i.e. divides each value by 256)\n",
    "    value: list (length 3) of RGB values\n",
    "    Returns: list (length 3) of decimal values'''\n",
    "    return [v/256 for v in value]\n",
    "\n",
    "def get_continuous_cmap(hex_list, float_list=None):\n",
    "    ''' creates and returns a color map that can be used in heat map figures.\n",
    "        If float_list is not provided, colour map graduates linearly between each color in hex_list.\n",
    "        If float_list is provided, each color in hex_list is mapped to the respective location in float_list. \n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        hex_list: list of hex code strings\n",
    "        float_list: list of floats between 0 and 1, same length as hex_list. Must start with 0 and end with 1.\n",
    "        \n",
    "        Returns\n",
    "        ----------\n",
    "        colour map'''\n",
    "    rgb_list = [rgb_to_dec(hex_to_rgb(i)) for i in hex_list]\n",
    "    if float_list:\n",
    "        pass\n",
    "    else:\n",
    "        float_list = list(np.linspace(0,1,len(rgb_list)))\n",
    "        \n",
    "    cdict = dict()\n",
    "    for num, col in enumerate(['red', 'green', 'blue']):\n",
    "        col_list = [[float_list[i], rgb_list[i][num], rgb_list[i][num]] for i in range(len(float_list))]\n",
    "        cdict[col] = col_list\n",
    "    cmp = mcolors.LinearSegmentedColormap('my_cmp', segmentdata=cdict, N=256)\n",
    "    return cmp\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e0d4f0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set aesthetic parameters\n",
    "\n",
    "line_color = '#0F0A0A'    # dark\n",
    "# line_color = '#E5E4E2'    # light\n",
    "\n",
    "neutral_color = '#E0E0E0'\n",
    "\n",
    "# thursday\n",
    "accent_colors = ['#9B5DE5', '#FFAC69', '#00C9AE', '#FD3F92', '#0F0A0A', \n",
    "                 '#959AB1', '#FFDB66', '#63B9FF','#FFB1CA', '#4F1DD7']\n",
    "\n",
    "grays = ['#0F0A0A', '#52525E', '#747688', '#959AB1', '#959AB1', '#CECFD5', '#E0E0E0']\n",
    "\n",
    "style = {'axes.edgecolor': line_color,\n",
    "         'axes.labelcolor': line_color,\n",
    "         'text.color': line_color,\n",
    "         'xtick.color': line_color,\n",
    "         'ytick.color': line_color,\n",
    "         'font.family': 'Helvetica',\n",
    "         'font.Helvetica': ['Helvetica']}\n",
    "\n",
    "palette = sns.color_palette(accent_colors)\n",
    "\n",
    "sns.set_context('talk', rc={'lines.linewidth': 2})\n",
    "sns.set_palette(palette)\n",
    "# sns.set_palette('tab20')\n",
    "sns.set_style('ticks', style)\n",
    "\n",
    "plt.rcParams['legend.frameon'] = False\n",
    "plt.rcParams['axes.spines.left'] = True\n",
    "plt.rcParams['axes.spines.right'] = False\n",
    "plt.rcParams['axes.spines.top'] = False\n",
    "plt.rcParams['axes.spines.bottom'] = True\n",
    "plt.rcParams['axes.facecolor'] = 'none'\n",
    "plt.rcParams['figure.facecolor'] = 'none'\n",
    "plt.rcParams['patch.linewidth'] = 0\n",
    "plt.rcParams['patch.edgecolor'] = 'none'\n",
    "plt.rcParams['savefig.dpi'] = 300\n",
    "\n",
    "# bi_hexes = [neutral_color, '#7222D3']\n",
    "bi_hexes = [neutral_color, line_color]\n",
    "bi_cmap = get_continuous_cmap(bi_hexes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef3fca52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# helper functions\n",
    "\n",
    "def unique_bins(array, n_bins, bounds=None, reverse=False):\n",
    "    '''This is a function that maps an array of continuous values with potential duplicates to \n",
    "    a set of bins in such a way that each of the original continuous values is mapped to a unique bin, and the\n",
    "    distribution of binned values approximates the distribution of continuous values as closely as possible.\n",
    "    \n",
    "    My quick and dirty approach to this problem sorts the input array into the appropriate number of bins. Then,\n",
    "    starting in the middle and alternating outward left and right, the algorithm checks to see if each bin has \n",
    "    more than one resident. If so, it will again alternate outward left and right from that position and allocate\n",
    "    the extra residents to the nearest free bins.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    array : np.array\n",
    "        Sorted array of values to be placed into bins.\n",
    "    n_bins : int\n",
    "        Number of bins to use.\n",
    "    bounds : (float, float)\n",
    "        Upper and lower bounds to use for bin-mapping.\n",
    "    reverse : bool\n",
    "        Optional flag to reverse the index ordering of the output.\n",
    "    \n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    index : np.array\n",
    "        Indices indicating mapping of input values to bins, where index[i] indicates the index of the bin \n",
    "        mapping of input value array[i] (assuming input array is sorted).\n",
    "    '''\n",
    "    # check number of bins\n",
    "    if not n_bins >= len(array):\n",
    "        raise ValueError('The number of bins must be >= len(array).')\n",
    "    # add bounds if supplied\n",
    "    if bounds is not None:\n",
    "        array = np.concatenate((array, bounds))\n",
    "    # make histogram\n",
    "    counts, _ = np.histogram(array, bins=n_bins)\n",
    "    # remove bounds if added\n",
    "    if bounds is not None:\n",
    "        counts[0] -= 1\n",
    "        counts[-1] -= 1\n",
    "    # start from the middle\n",
    "    i = int(n_bins/2)\n",
    "    for j in range(n_bins):\n",
    "        # alternate left and right\n",
    "        if j % 2:\n",
    "            i -= j\n",
    "        else:\n",
    "            i += j\n",
    "        # resolve collisions\n",
    "        if counts[i] > 1:\n",
    "            # set new counter for exploring nearby positions\n",
    "            k = i\n",
    "            for l in range(2*n_bins):\n",
    "                # alternate left and right\n",
    "                if l % 2:\n",
    "                    k -= l\n",
    "                else:\n",
    "                    k += l\n",
    "                # check if we're still in range\n",
    "                if 0 <= k < n_bins:\n",
    "                    # check for an empty spot\n",
    "                    if counts[k] == 0:\n",
    "                        counts[k] = 1\n",
    "                        counts[i] -= 1\n",
    "                        # check if all collisions have been resolved yet\n",
    "                        if counts[i] == 1:\n",
    "                            break\n",
    "    # select indices of non-zero values\n",
    "    index = np.where(counts == 1)[0]\n",
    "    # reverse if requested\n",
    "    if reverse:\n",
    "        index = (n_bins - 1) - index\n",
    "    return index\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74c99328",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "973fd7d5",
   "metadata": {},
   "source": [
    "# import data\n",
    "### cluster data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99af7522",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in aligned model datasets\n",
    "\n",
    "datapath = Path('../../data/')\n",
    "outdir = Path('../../figures/7-interpretation/')\n",
    "\n",
    "pro_ds = xr.open_dataset(datapath / '5-models/pro-aligned-models.nc')\n",
    "syn_ds = xr.open_dataset(datapath / '5-models/syn-aligned-models.nc')\n",
    "\n",
    "fe_clusters = {'pro': 7, 'syn': 8}\n",
    "n_clusters = {'pro': 15, 'syn': 12}\n",
    "\n",
    "pro_ds\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1ed5190",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in metadata\n",
    "\n",
    "sample_df = pd.read_csv(datapath / 'metadata/sample_metadata_merged.csv')\n",
    "sample_df['Datetime'] = pd.to_datetime(sample_df['Datetime'], format='%m/%d/%y %H:%M')\n",
    "\n",
    "sample_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ee09366",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull out clusters hypothesized to relate to iron limitation\n",
    "\n",
    "pro_fe_df = pro_ds.sel(Component=fe_clusters['pro']).SampleWeight.to_series().reset_index()\n",
    "pro_fe_df['Genus'] = 'Prochlorococcus'\n",
    "syn_fe_df = syn_ds.sel(Component=fe_clusters['syn']).SampleWeight.to_series().reset_index()\n",
    "syn_fe_df['Genus'] = 'Synechococcus'\n",
    "\n",
    "fe_cluster_df = pd.concat([pro_fe_df, syn_fe_df]).reset_index(drop=True)\n",
    "fe_cluster_df = pd.merge(fe_cluster_df, sample_df, on='SampleName', how='left')\n",
    "\n",
    "fe_cluster_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27dbc2fb-a397-4262-a4cc-00a2371ebe2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull out clusters hypothesized to relate to nitrogen limitation\n",
    "\n",
    "pro_n_df = pro_ds.sel(Component=n_clusters['pro']).SampleWeight.to_series().reset_index()\n",
    "pro_n_df['Genus'] = 'Prochlorococcus'\n",
    "syn_n_df = syn_ds.sel(Component=n_clusters['syn']).SampleWeight.to_series().reset_index()\n",
    "syn_n_df['Genus'] = 'Synechococcus'\n",
    "\n",
    "n_cluster_df = pd.concat([pro_n_df, syn_n_df]).reset_index(drop=True)\n",
    "n_cluster_df = pd.merge(n_cluster_df, sample_df, on='SampleName', how='left')\n",
    "\n",
    "n_cluster_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e78b60bc-a1d2-4f2c-9cb7-c3bd37c90af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in gene annotations\n",
    "\n",
    "gene_names_df = pd.read_excel(\n",
    "    datapath / '7-interpretation/fe-n-cluster-annotations.xlsx', \n",
    "    sheet_name='Fe-N-CyCOGs', \n",
    "    usecols=['Ortholog', 'Gene Name', 'Protein Name', 'Expression Response']\n",
    ").rename(columns={'Gene Name': 'GeneLabel', 'Protein Name': 'ProteinLabel', 'Expression Response': 'ExpressionResponse'})\n",
    "\n",
    "gene_names_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f0c7262-49d3-45b7-85e7-e34367e7f029",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b059af34-8dab-4caa-9f02-4630b222175c",
   "metadata": {},
   "source": [
    "# import data\n",
    "### environmental data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a91ba8a-8c92-4b30-8dfb-d113742ecd0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define salinity and chlorophyll fronts\n",
    "\n",
    "# source: Juranek et al., 2020 https://doi.org/10.1029/2020GB006702\n",
    "sal_fronts = {2016: 32.15, 2017: 32.5, 2019: 32.34}\n",
    "chl_fronts = {2016: 33, 2017: 36.2, 2019: 35}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "992535be-e4df-44db-9ffd-3fda08a571cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import data from cmap\n",
    "\n",
    "api = pycmap.API(token='55bd80f9-8d5d-4c8f-8917-f5b0e924ed45')\n",
    "api.search_catalog('iron John gradients')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59e3204a-1030-4b2e-b970-0243fb19911f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import all iron data from G1, G2, G3\n",
    "\n",
    "fe_df = pd.concat([\n",
    "    api.get_dataset('tblKOK1606_Gradients1_Diss_Trace_Metal_Profile'), \n",
    "    api.get_dataset('tblMGL1704_Gradients2_Diss_Trace_Metal_Profile'), \n",
    "    api.get_dataset('tblMGL1704_Gradients2_Diss_Trace_Metal_Surface'), \n",
    "    api.get_dataset('tblKM1906_Gradients3_Diss_Trace_Metal_Profile')\n",
    "]).reset_index(drop=True)\n",
    "fe_df['year'] = pd.to_datetime(fe_df['time']).dt.year\n",
    "\n",
    "fe_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4cd55e6-52de-49de-a1c2-caba1654e868",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import nitrogen data from G1, G2, G3\n",
    "\n",
    "# dissolved nitrate+nitrite (Karl lab)\n",
    "n_df =  pd.concat([\n",
    "    api.get_dataset('tblKOK1606_Gradients1_Nutrients'), \n",
    "    api.get_dataset('tblMGL1704_Gradients2_Nutrients'), \n",
    "    api.get_dataset('tblKM1906_Gradients3_Organic_Inorganic_Nutrients').rename(columns={'NO3_plus_NO2': 'NO3_NO2'})\n",
    "]).reset_index(drop=True)\n",
    "n_df['year'] = pd.to_datetime(n_df['time']).dt.year\n",
    "\n",
    "# drop any rows that don't have nitrate+nitrite values\n",
    "n_df = n_df[n_df['NO3_NO2'].notna()]\n",
    "\n",
    "n_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "587ddcac-52c1-4317-abc7-0f0d8ebbfb6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge Fe data on dissolved N data\n",
    "\n",
    "# only select surface and depth samples for joining\n",
    "source_df = n_df[n_df.depth.le(15)].reset_index(drop=True)\n",
    "source_cols = source_df.columns.to_list()\n",
    "tolerances = [2, 0.25, 0.25, 15] # time, lat, lon, depth\n",
    "\n",
    "# define target datasets\n",
    "targets = {\n",
    "    # dissolved Fe\n",
    "    'tblKOK1606_Gradients1_Diss_Trace_Metal_Profile': {\n",
    "        'variables': ['Fe_dissolved'], \n",
    "        'tolerances': tolerances\n",
    "    },\n",
    "    'tblMGL1704_Gradients2_Diss_Trace_Metal_Profile': {\n",
    "        'variables': ['Fe_dissolved'], \n",
    "        'tolerances': tolerances\n",
    "    },\n",
    "    'tblMGL1704_Gradients2_Diss_Trace_Metal_Surface': {\n",
    "        'variables': ['Fe_dissolved'], \n",
    "        'tolerances': tolerances\n",
    "    },\n",
    "    'tblKM1906_Gradients3_Diss_Trace_Metal_Profile': {\n",
    "        'variables': ['Fe_dissolved'], \n",
    "        'tolerances': tolerances\n",
    "    }\n",
    "}\n",
    "\n",
    "# join target data on the source dataframe\n",
    "merged_df = pycmap.Sample(\n",
    "    source=source_df, \n",
    "    targets=targets, \n",
    "    replaceWithMonthlyClimatolog=False\n",
    ")\n",
    "\n",
    "# merge all of the iron data into one column\n",
    "fe_start = 'CMAP_Fe_dissolved_tblKOK1606_Gradients1_Diss_Trace_Metal_Profile'\n",
    "fe_end = 'CMAP_Fe_dissolved_tblKM1906_Gradients3_Diss_Trace_Metal_Profile'\n",
    "merged_df['Fe_dissolved'] = merged_df.loc[:, fe_start:fe_end].max(axis=1)\n",
    "\n",
    "# trim the dataframe\n",
    "merged_df = merged_df[source_cols + ['Fe_dissolved']]\n",
    "\n",
    "# calculate ratio\n",
    "merged_df['N:Fe'] = (merged_df['NO3_NO2'] * 1e-6) / (merged_df['Fe_dissolved'] * 1e-9)\n",
    "\n",
    "merged_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51df6d20-abe0-470f-a0c8-0e497dd68b40",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1630921c-73af-4737-98cc-fe3cd1f38d1c",
   "metadata": {},
   "source": [
    "# collect ecotype and gene weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d86b3676-9884-4c97-91a5-d6306136613a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull out ecotype weights\n",
    "\n",
    "with pd.option_context('mode.chained_assignment', None):\n",
    "    pro_fe_eco_df = pro_ds.sel(Component=fe_clusters['pro']).TaxonWeight.to_series().reset_index()\n",
    "    pro_fe_eco_df['Genus'] = 'Prochlorococcus'\n",
    "    pro_fe_eco_df['Cluster'] = f\"pro{fe_clusters['pro']}\"\n",
    "    pro_fe_eco_df['ClusterGroup'] = 'Fe'\n",
    "    pro_n_eco_df = pro_ds.sel(Component=n_clusters['pro']).TaxonWeight.to_series().reset_index()\n",
    "    pro_n_eco_df['Genus'] = 'Prochlorococcus'\n",
    "    pro_n_eco_df['Cluster'] = f\"pro{n_clusters['pro']}\"\n",
    "    pro_n_eco_df['ClusterGroup'] = 'N'\n",
    "    syn_fe_eco_df = syn_ds.sel(Component=fe_clusters['syn']).TaxonWeight.to_series().reset_index()\n",
    "    syn_fe_eco_df['Genus'] = 'Synechococcus'\n",
    "    syn_fe_eco_df['Cluster'] = f\"syn{fe_clusters['syn']}\"\n",
    "    syn_fe_eco_df['ClusterGroup'] = 'Fe'\n",
    "    syn_n_eco_df = syn_ds.sel(Component=n_clusters['syn']).TaxonWeight.to_series().reset_index()\n",
    "    syn_n_eco_df['Genus'] = 'Synechococcus'\n",
    "    syn_n_eco_df['Cluster'] = f\"syn{n_clusters['syn']}\"\n",
    "    syn_n_eco_df['ClusterGroup'] = 'N'\n",
    "ecotype_df = pd.concat([pro_fe_eco_df, pro_n_eco_df, syn_fe_eco_df, syn_n_eco_df])\n",
    "\n",
    "# swap out clade labels with display labels\n",
    "clade_order = ['HLI', 'HLII', 'LLI', '5.1B-I', '5.1B-CRD1', '5.1A-II', '5.1A-III', '5.1A-IV', '5.1A-CRD2', '5.1A-UC-A-EnvC']\n",
    "clade_labels = ['HLI', 'HLII', 'LLI', 'I', 'CRD1', 'II', 'III', 'IV', 'CRD2', 'UC-A']\n",
    "ecotype_df['Clade'] = ecotype_df['Clade'].map(dict(zip(clade_order, clade_labels)))\n",
    "\n",
    "ecotype_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0044e54d-1a99-422e-bbf8-2fbf9dd5dd9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull out gene weights\n",
    "\n",
    "geneweight_df = pd.DataFrame()\n",
    "# pull gene weights of all clusters\n",
    "with pd.option_context('mode.chained_assignment', None):\n",
    "    for genus in ['pro', 'syn']:\n",
    "        ds = {'pro': pro_ds, 'syn': syn_ds}[genus]\n",
    "        for clustergroup in ['Fe', 'N']:\n",
    "            cluster = {'Fe': fe_clusters, 'N': n_clusters}[clustergroup][genus]\n",
    "            cluster_geneweight_df = ds.sel(Component=cluster).GeneWeight.to_series().reset_index()\n",
    "            medians = cluster_geneweight_df.groupby('Ortholog').GeneWeight.median()\n",
    "            cluster_geneweight_df = cluster_geneweight_df[\n",
    "                cluster_geneweight_df.Ortholog.isin(medians[~medians.eq(0.)].index)].reset_index(drop=True)\n",
    "            cluster_geneweight_df['Nonzero'] = cluster_geneweight_df.GeneWeight != 0\n",
    "            cluster_geneweight_df['% Bootstraps'] = cluster_geneweight_df['Ortholog'].map(\n",
    "                cluster_geneweight_df.groupby(['Ortholog']).Nonzero.mean()) * 100\n",
    "            cluster_geneweight_df['Genus'] = {'pro': 'Prochlorococcus', 'syn': 'Synechococcus'}[genus]\n",
    "            cluster_geneweight_df['Cluster'] = f\"{genus}{cluster}\"\n",
    "            cluster_geneweight_df['ClusterGroup'] = clustergroup\n",
    "            cluster_geneweight_df['Annotation'] = cluster_geneweight_df['Ortholog'].map(ds['Annotation'].to_series())\n",
    "            if len(geneweight_df) == 0:\n",
    "                geneweight_df = cluster_geneweight_df\n",
    "            else:\n",
    "                geneweight_df = pd.concat([geneweight_df, cluster_geneweight_df])\n",
    "\n",
    "median_gene_df = geneweight_df.groupby(\n",
    "    ['Cluster', 'Genus', 'ClusterGroup', 'Ortholog', 'Annotation', '% Bootstraps']).GeneWeight.median().reset_index()\n",
    "median_gene_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09fd6fd8-5dc3-4d1f-8c32-196403fda1ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set gene labels to be displayed in plot\n",
    "\n",
    "median_gene_df['GeneLabel'] = median_gene_df.Ortholog.map(gene_names_df.set_index('Ortholog')['GeneLabel'])\n",
    "median_gene_df['ExpressionResponse'] = median_gene_df.Ortholog.map(gene_names_df.set_index('Ortholog')['ExpressionResponse'])\n",
    "# fill in unannotated genes with CyCOG label\n",
    "median_gene_df.loc[median_gene_df.GeneLabel.isna(), 'GeneLabel'] = median_gene_df.loc[median_gene_df.GeneLabel.isna(), 'Ortholog']\n",
    "# pull out top genes\n",
    "top_genes = median_gene_df[np.abs(median_gene_df['GeneWeight']).ge(0.04)].Ortholog.unique()\n",
    "# append the tauA taurine transporter of another CyCOG\n",
    "top_genes = np.append(top_genes, [60007639])\n",
    "\n",
    "# look at how included genes\n",
    "print(len(top_genes))\n",
    "median_gene_df[median_gene_df['Ortholog'].isin(top_genes)].sort_values(\n",
    "    ['ClusterGroup', 'Cluster', 'GeneWeight'], ascending=[True, True, False])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bed53d78-2d22-4d0e-941a-be3054da6129",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12e4231b-ea20-4624-89ef-f8308fe5692a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "87702799-5d97-4f15-b0fb-2b1ce4b017f6",
   "metadata": {},
   "source": [
    "# panels a & b\n",
    "\n",
    "- A: Ecotype Weights\n",
    "- B: Gene Weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6499af0-a22e-4b83-b20f-56378785d487",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make ecotype & gene weight figure\n",
    "\n",
    "# set up figure\n",
    "sample_weight_range = (-0.05, 1.05) # range of weights to show on sample plots\n",
    "fig = plt.figure(figsize=(20, 4))\n",
    "rows = 4\n",
    "cols = 20\n",
    "grid = plt.GridSpec(rows, cols, wspace=.2, hspace=.2)\n",
    "\n",
    "# labels and colors\n",
    "clusters = [f\"pro{fe_clusters['pro']}\", f\"syn{fe_clusters['syn']}\", f\"pro{n_clusters['pro']}\", f\"syn{n_clusters['syn']}\"]\n",
    "genus_palette = dict(zip(['Prochlorococcus', 'Synechococcus'], [sns.color_palette()[i] for i in [2, 3]]))\n",
    "clade_order = ['HLI', 'HLII', 'LLI', 'I', 'CRD1', 'II', 'III', 'IV', 'CRD2', 'UC-A']\n",
    "\n",
    "# plot taxon weights\n",
    "for i, cid in enumerate(clusters):\n",
    "    axis = plt.subplot(grid[i, 1:6])\n",
    "    plot_df = ecotype_df[ecotype_df['Cluster'] == cid]\n",
    "    sns.boxplot(data=plot_df, x='Clade', y='TaxonWeight', hue='Genus', order=clade_order, hue_order=['Prochlorococcus', 'Synechococcus'], \n",
    "                palette=genus_palette, fill=False, whis=(5, 95), fliersize=1, \n",
    "                # legend=(i == n_clusters-1), \n",
    "                legend=False, \n",
    "                ax=axis)\n",
    "    # labels and such\n",
    "    axis.annotate(cid, xy=(-0.3, 0.5), xycoords='axes fraction', va='center', ha='left', fontsize=24, \n",
    "                  annotation_clip=False, color=genus_palette[plot_df.Genus.unique()[0]])\n",
    "    axis.set(title=('Ecotype Weights' if not i else None), ylim=sample_weight_range, xlabel=None, ylabel=None)\n",
    "    if i != 3:\n",
    "        axis.tick_params(bottom=False, labelbottom=False)\n",
    "    else:\n",
    "        axis.set_xticks(clade_order, clade_order, rotation=90)\n",
    "\n",
    "# plot gene weights\n",
    "axis = plt.subplot(grid[:4, 6:])\n",
    "axis.spines[['left', 'bottom']].set_visible(False)\n",
    "plot_df = median_gene_df[median_gene_df['Ortholog'].isin(top_genes)].sort_values(\n",
    "    ['ClusterGroup', 'Cluster', 'GeneWeight'], ascending=[True, True, False]).reset_index(drop=True)\n",
    "n_genes = plot_df.GeneLabel.nunique()\n",
    "with pd.option_context('mode.chained_assignment', None):\n",
    "    plot_df['x'] = plot_df.GeneLabel.map(dict(zip(plot_df.GeneLabel.unique(), np.arange(n_genes))))\n",
    "    plot_df['y'] = plot_df.Cluster.map(dict(zip(clusters, reversed(np.arange(len(clusters))))))\n",
    "sns.scatterplot(x='x', y='y', hue='GeneWeight', \n",
    "    # size='% Bootstraps', \n",
    "    # sizes=(50, 500), \n",
    "    s=400, edgecolor=None, palette=bi_cmap, data=plot_df, legend=True, ax=axis\n",
    ");\n",
    "# plot marker gene status\n",
    "axis.plot('x', 'y', '.', data=plot_df, markerfacecolor='w', markeredgecolor='none', zorder=0, label=' ')\n",
    "axis.plot('x', 'y', 'ks', data=plot_df[plot_df['ExpressionResponse'] == 'Fe responsive'], markeredgewidth=3, \n",
    "          markerfacecolor='none', ms=26, markeredgecolor=sns.color_palette()[1], label='Fe responsive')\n",
    "axis.plot('x', 'y', '.', data=plot_df, markerfacecolor='w', markeredgecolor='none', zorder=0, label='  ')\n",
    "axis.plot('x', 'y', 'ks', data=plot_df[plot_df['ExpressionResponse'] == 'N responsive'], markeredgewidth=3, \n",
    "          markerfacecolor='none', ms=26, markeredgecolor=sns.color_palette()[7], label='N responsive')\n",
    "\n",
    "axis.set(title='Gene Weights', xlabel=None, ylabel=None, xlim=(-0.5, n_genes-0.5), ylim=(-0.5, 3.5))\n",
    "axis.set_xticks(np.arange(n_genes), plot_df.GeneLabel.unique(), rotation=90)\n",
    "axis.tick_params(left=False, labelleft=False, bottom=False)\n",
    "axis.legend(title='     Median\\nGene Weight', loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13034d0a-ff91-4aaa-ab2b-8ad19c71edd5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cdc4fc98-acb4-44e0-9056-5f7678e980e0",
   "metadata": {},
   "source": [
    "# panels c & d\n",
    "\n",
    "- C: Surface Samples\n",
    "- D: Incubations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9492298-c8e5-4ca9-abd6-4007e71715fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up incubation weights with both Fe and nitrogen clusters\n",
    "\n",
    "# set up data\n",
    "treatments = ['control 0h', 'control 96h', 'low Fe', 'high Fe', 'Fe', 'N+P+Fe', 'N+P', 'low N+P', 'high N+P']\n",
    "with pd.option_context('mode.chained_assignment', None):\n",
    "    fe_incubation_df = fe_cluster_df[fe_cluster_df.Treatment.isin(treatments)]\n",
    "    fe_incubation_df['ClusterGroup'] = 'Fe'\n",
    "    n_incubation_df = n_cluster_df[n_cluster_df.Treatment.isin(treatments)]\n",
    "    n_incubation_df['ClusterGroup'] = 'N'\n",
    "incubation_df = pd.concat([fe_incubation_df, n_incubation_df]).reset_index(drop=True).fillna(0)\n",
    "incubation_df = incubation_df.sort_values('Treatment', key=lambda s: s.apply(treatments.index))\n",
    "incubation_df['Condition'] = incubation_df['Condition'].map({'gyre': 'Gyre', 'TZ': 'TZ', 'HNLC': 'HNLC'})\n",
    "\n",
    "incubation_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86a873cb-5981-48f0-b157-d617e03236f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# build combined fe & n cluster figure\n",
    "\n",
    "# set up figure\n",
    "sample_weight_range = (-0.05, 0.7) # range of weights to show on sample plots\n",
    "fig = plt.figure(figsize=(24, 12))\n",
    "rows = 7\n",
    "cols = 27\n",
    "grid = plt.GridSpec(rows, cols, hspace=.3, wspace=.3)\n",
    "n_bins = 21\n",
    "bounds = (23, 43)\n",
    "depth_limit = 15 # meters\n",
    "# time_classes = ['2am', '6am', '10am']\n",
    "time_classes = ['2am', '6am', '10am', '2pm', '6pm', '10pm']\n",
    "fliersize = 1\n",
    "annot_kwgs = {\n",
    "    'xycoords': 'axes fraction', 'va': 'center', 'ha': 'left', 'fontsize': 26, 'annotation_clip': False\n",
    "}\n",
    "\n",
    "# panel c: surface transects + environmental covariates\n",
    "for i, year in enumerate([2016, 2017, 2019]):\n",
    "    # Fe limitation cluster\n",
    "    axis = plt.subplot(grid[:, 7*i:7*i+2])\n",
    "    # add annotation\n",
    "    if i == 0:\n",
    "        axis.annotate('C', xy=(-0.5, 1.1), **annot_kwgs)\n",
    "    # get data together\n",
    "    plot_df = fe_cluster_df[fe_cluster_df['SampleGroup'] == 'surface transect']\n",
    "    plot_df = plot_df[plot_df['TimeClass'].isin(time_classes)]\n",
    "    plot_df = plot_df[plot_df['Year'] == year]\n",
    "    latitudes = plot_df[\n",
    "        ['SampleName', 'Latitude']].drop_duplicates().set_index('SampleName').sort_values('Latitude')['Latitude']\n",
    "    lat_bins = unique_bins(latitudes, n_bins, bounds=bounds, reverse=False)\n",
    "    plot_df['lat_idx'] = plot_df['SampleName'].map(dict(zip(latitudes.index, lat_bins)))\n",
    "    # make figure\n",
    "    sns.boxplot(\n",
    "        data=plot_df, x='SampleWeight', y='lat_idx', hue='Genus', hue_order=['Prochlorococcus', 'Synechococcus'], \n",
    "        palette=[sns.color_palette()[i] for i in [2, 3]], orient='y', order=np.arange(n_bins), \n",
    "        fill=False, whis=(5, 95), fliersize=fliersize, legend=False, ax=axis\n",
    "    )\n",
    "    # set labels and such\n",
    "    axis.set_yticks(np.arange(1, n_bins, 2), [f'{int(l)}N' for l in np.arange(bounds[0]+1, bounds[1]+1, 2)])\n",
    "    axis.set(xlim=sample_weight_range, ylim=(-0.5, n_bins-0.5), xlabel='Weight', ylabel='', title='Fe Clusters')\n",
    "    axis.set_yticks(np.arange(n_bins), minor=True)\n",
    "    \n",
    "    # N acquisition cluster\n",
    "    axis = plt.subplot(grid[:, 7*i+2:7*i+4])\n",
    "    # add title\n",
    "    if i == 1:\n",
    "        axis.annotate('Surface Transects', xy=(.5, 1.1), xycoords='axes fraction', va='center', ha='center', fontsize=22)\n",
    "    # get data together\n",
    "    plot_df = n_cluster_df[n_cluster_df['SampleGroup'] == 'surface transect']\n",
    "    plot_df = plot_df[plot_df['TimeClass'].isin(time_classes)]\n",
    "    plot_df = plot_df[plot_df['Year'] == year]\n",
    "    plot_df['lat_idx'] = plot_df['SampleName'].map(dict(zip(latitudes.index, lat_bins)))\n",
    "    # make figure\n",
    "    sns.boxplot(\n",
    "        data=plot_df, x='SampleWeight', y='lat_idx', hue='Genus', hue_order=['Prochlorococcus', 'Synechococcus'], \n",
    "        palette=[sns.color_palette()[i] for i in [2, 3]], orient='y', order=np.arange(n_bins), fill=False, \n",
    "        whis=(5, 95), fliersize=fliersize, legend=False, ax=axis\n",
    "    )\n",
    "    # set labels and such\n",
    "    axis.set_yticks(np.arange(1, n_bins, 2))\n",
    "    axis.set(xlim=sample_weight_range, ylim=(-0.5, n_bins-0.5), xlabel='Weight', ylabel='', title=f\"{year}\\nN Clusters\")\n",
    "    axis.set_yticks(np.arange(n_bins), minor=True)\n",
    "    axis.tick_params(left=True, labelleft=False)\n",
    "\n",
    "    # environmental covariates\n",
    "    axis = plt.subplot(grid[:, 7*i+4:7*i+6])\n",
    "    # plot salinity front\n",
    "    axis.hlines(y=sal_fronts[year], xmin=0, xmax=2, linewidth=10, color=sns.color_palette()[5], alpha=0.2, label='Salinity Front')\n",
    "    # plot chlorophyll front\n",
    "    axis.hlines(y=chl_fronts[year], xmin=0, xmax=2, linewidth=10, color=sns.color_palette()[2], alpha=0.2, label='Chlorophyll Front')\n",
    "    \n",
    "    # dissolved iron\n",
    "    color = sns.color_palette()[1]\n",
    "    sns.lineplot(\n",
    "        data=fe_df[(fe_df['year'] == year) & (fe_df['depth'].le(depth_limit))], x='Fe_dissolved', \n",
    "        y='lat', orient='y', ax=axis, color=color, marker='.', label='[Fe] (nmol/L)', legend=(i==2)\n",
    "    )\n",
    "    axis.set(xlim=(-.05, 0.85), ylim=(bounds[0]-0.5, bounds[1]+.5), yticks=np.arange(1, n_bins, 2)+bounds[0], \n",
    "             xlabel='[Fe]', ylabel='')\n",
    "    axis.set_yticks(np.arange(n_bins)+bounds[0], minor=True)\n",
    "    axis.tick_params(left=True, labelleft=False)\n",
    "    axis.tick_params(axis='x', colors=color)\n",
    "    axis.xaxis.label.set_color(color)\n",
    "    axis.spines['bottom'].set_color(color)\n",
    "    if i == 2:\n",
    "        axis.legend(loc='upper left', bbox_to_anchor=(2.3, -0.06))\n",
    "\n",
    "    # dissolved nitrate + nitrite\n",
    "    color = sns.color_palette()[7]\n",
    "    axis2 = plt.twiny(axis)\n",
    "    sns.lineplot(\n",
    "        data=n_df[(n_df['year'] == year) & (n_df['depth'].le(depth_limit))], x='NO3_NO2', y='lat', orient='y', \n",
    "        ax=axis2, color=color, marker='.', label='[NO$_3^-$+NO$_2^-$] (umol/L)', legend=(i==2)\n",
    "    )\n",
    "    axis2.set(xlim=(-0.5, 7.5), ylim=(bounds[0]-0.5, bounds[1]+.5), yticks=np.arange(1, n_bins, 2)+bounds[0], \n",
    "             xlabel='[NO$_3^-$+NO$_2^-$]', ylabel='')\n",
    "    axis2.set_yticks(np.arange(n_bins)+bounds[0], minor=True)\n",
    "    axis2.tick_params(top=False, labeltop=False, bottom=True, labelbottom=True)\n",
    "    axis2.tick_params(axis='x', colors=color)\n",
    "    axis2.xaxis.set_label_position('bottom')\n",
    "    axis2.xaxis.label.set_color(color)\n",
    "    axis2.spines['bottom'].set_color(color)\n",
    "    axis2.spines['bottom'].set_position(('outward', 60))\n",
    "    if i == 2:\n",
    "        axis2.legend(loc='upper left', bbox_to_anchor=(2.3, -0.165))\n",
    "\n",
    "    # N:Fe ratio\n",
    "    color = sns.color_palette()[4]\n",
    "    axis3 = plt.twiny(axis2)\n",
    "    sns.lineplot(\n",
    "        data=merged_df[(merged_df['year'] == year) & (merged_df['depth'].le(depth_limit))], x='N:Fe', y='lat', orient='y', \n",
    "        ax=axis3, color=color, marker='.', label='[NO$_3^-$+NO$_2^-$]:[Fe] Ratio', legend=(i==2)\n",
    "    )\n",
    "    axis3.set(xlim=(1, 1000000), ylim=(bounds[0]-0.5, bounds[1]+.5), yticks=np.arange(1, n_bins, 2)+bounds[0], \n",
    "             xlabel='[NO$_3^-$+NO$_2^-$]:[Fe]', ylabel='', xscale='log')\n",
    "    axis3.set_yticks(np.arange(n_bins)+bounds[0], minor=True)\n",
    "    axis3.tick_params(top=False, labeltop=False, bottom=True, labelbottom=True)\n",
    "    axis3.tick_params(axis='x', colors=color)\n",
    "    axis3.xaxis.set_label_position('bottom')\n",
    "    axis3.xaxis.label.set_color(color)\n",
    "    axis3.spines['bottom'].set_color(color)\n",
    "    axis3.spines['bottom'].set_position(('outward', 120))\n",
    "    if i == 2:\n",
    "        axis3.legend(loc='upper left', bbox_to_anchor=(2.3, -0.2))\n",
    "\n",
    "# panel d: incubation weights\n",
    "for i, condition in enumerate(['HNLC', 'TZ', 'Gyre']):\n",
    "    for j, clustergroup in enumerate(['Fe', 'N']):\n",
    "        axis = plt.subplot(grid[i*2:i*2+2, 23+(2*j):25+(2*j)])\n",
    "        plot_df = incubation_df[(incubation_df['Condition'] == condition) & (incubation_df['ClusterGroup'] == clustergroup)]\n",
    "        sns.boxplot(\n",
    "            data=plot_df, x='SampleWeight', y='Treatment', hue='Genus', \n",
    "            hue_order=['Prochlorococcus', 'Synechococcus'], palette=[sns.color_palette()[i] for i in [2, 3]], \n",
    "            orient='y', fill=False, whis=(5, 95), fliersize=fliersize, legend=(i==2 and j==0), ax=axis\n",
    "        )\n",
    "        # labels and such\n",
    "        if i == 0:\n",
    "            axis.set(title='{} Clusters'.format(clustergroup))\n",
    "            if j == 0:\n",
    "                axis.annotate('D', xy=(-1.5, 1.35), **annot_kwgs)\n",
    "            if j == 1:\n",
    "                axis.annotate('Incubations (2017)', xy=(0, 1.35), xycoords='axes fraction', va='center', ha='center', fontsize=22)\n",
    "        if i == 2:\n",
    "            xlabel = 'Weight'\n",
    "            if j == 0:\n",
    "                axis.legend(loc='upper left', bbox_to_anchor=(-0.52, -0.4))\n",
    "        else:\n",
    "            xlabel = None\n",
    "        if j == 0:\n",
    "            ylabel = f\"{condition} ({plot_df.Latitude.mean():.0f}N)\"\n",
    "        else:\n",
    "            ylabel = None\n",
    "        axis.set(xlim=sample_weight_range, xlabel=xlabel, ylabel=ylabel)\n",
    "        axis.tick_params(left=True, labelleft=(j==0), bottom=True, labelbottom=(i==2))\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30024dad-6cf3-4581-a484-5998349d9acf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "eae5b104-7156-4642-9cda-183d0e8a8cf7",
   "metadata": {},
   "source": [
    "# full combined figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "685eb7bd-351d-47bb-9388-2cb15ed5c39a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# build combined fe & n cluster figure\n",
    "\n",
    "# set up figure\n",
    "fig = plt.figure(figsize=(24, 20))\n",
    "rows = 17\n",
    "cols = 27\n",
    "grid = plt.GridSpec(rows, cols, hspace=.3, wspace=.3)\n",
    "n_bins = 21\n",
    "bounds = (23, 43)\n",
    "depth_limit = 15 # meters\n",
    "# time_classes = ['2am', '6am', '10am']\n",
    "time_classes = ['2am', '6am', '10am', '2pm', '6pm', '10pm']\n",
    "fliersize = 1\n",
    "annot_kwgs = {\n",
    "    'xycoords': 'axes fraction', 'va': 'center', 'ha': 'left', 'fontsize': 26, 'annotation_clip': False\n",
    "}\n",
    "\n",
    "# labels and colors\n",
    "clusters = [f\"pro{fe_clusters['pro']}\", f\"syn{fe_clusters['syn']}\", f\"pro{n_clusters['pro']}\", f\"syn{n_clusters['syn']}\"]\n",
    "genus_palette = dict(zip(['Prochlorococcus', 'Synechococcus'], [sns.color_palette()[i] for i in [2, 3]]))\n",
    "clade_order = ['HLI', 'HLII', 'LLI', 'I', 'CRD1', 'II', 'III', 'IV', 'CRD2', 'UC-A']\n",
    "\n",
    "# panel a: taxon weights\n",
    "sample_weight_range = (-0.05, 1.05) # range of weights to show on sample plots\n",
    "for i, cid in enumerate(clusters):\n",
    "    axis = plt.subplot(grid[i, 2:8])\n",
    "    # add annotation\n",
    "    if i == 0:\n",
    "        axis.annotate('A', xy=(-0.5, 1.6), **annot_kwgs)\n",
    "    elif i == 1: \n",
    "        axis.annotate('Fe Clusters', xy=(-0.4, 1), xycoords='axes fraction', va='center', ha='center', rotation=90, fontsize=18)\n",
    "    elif i == 3:\n",
    "        axis.annotate('N Clusters', xy=(-0.4, 1), xycoords='axes fraction', va='center', ha='center', rotation=90, fontsize=18)\n",
    "    plot_df = ecotype_df[ecotype_df['Cluster'] == cid]\n",
    "    sns.boxplot(data=plot_df, x='Clade', y='TaxonWeight', hue='Genus', order=clade_order, hue_order=['Prochlorococcus', 'Synechococcus'], \n",
    "                palette=genus_palette, fill=False, whis=(5, 95), fliersize=1, \n",
    "                # legend=(i == n_clusters-1), \n",
    "                legend=False, \n",
    "                ax=axis)\n",
    "    # labels and such\n",
    "    axis.annotate(cid, xy=(-0.3, 0.5), xycoords='axes fraction', va='center', ha='left', fontsize=22, \n",
    "                  annotation_clip=False, color=genus_palette[plot_df.Genus.unique()[0]])\n",
    "    axis.set(title=('Ecotype Weights' if not i else None), ylim=sample_weight_range, xlabel=None, ylabel=None)\n",
    "    if i != 3:\n",
    "        axis.tick_params(bottom=False, labelbottom=False)\n",
    "    else:\n",
    "        axis.set_xticks(clade_order, clade_order, rotation=90)\n",
    "\n",
    "# panel b: gene weights\n",
    "axis = plt.subplot(grid[:4, 9:24])\n",
    "axis.annotate('B', xy=(-0.05, 1.15), **annot_kwgs)\n",
    "axis.spines[['left', 'bottom']].set_visible(False)\n",
    "plot_df = median_gene_df[median_gene_df['Ortholog'].isin(top_genes)].sort_values(\n",
    "    ['ClusterGroup', 'Cluster', 'GeneWeight'], ascending=[True, True, False]).reset_index(drop=True)\n",
    "n_genes = plot_df.GeneLabel.nunique()\n",
    "with pd.option_context('mode.chained_assignment', None):\n",
    "    plot_df['x'] = plot_df.GeneLabel.map(dict(zip(plot_df.GeneLabel.unique(), np.arange(n_genes))))\n",
    "    plot_df['y'] = plot_df.Cluster.map(dict(zip(clusters, reversed(np.arange(len(clusters))))))\n",
    "sns.scatterplot(x='x', y='y', hue='GeneWeight', \n",
    "    # size='% Bootstraps', \n",
    "    # sizes=(50, 500), \n",
    "    s=400, edgecolor=None, palette=bi_cmap, data=plot_df, legend=True, ax=axis\n",
    ");\n",
    "# plot marker gene status\n",
    "axis.plot('x', 'y', '.', data=plot_df, markerfacecolor='w', markeredgecolor='none', zorder=0, label=' ')\n",
    "axis.plot('x', 'y', 'ks', data=plot_df[plot_df['ExpressionResponse'] == 'Fe responsive'], markeredgewidth=3, \n",
    "          markerfacecolor='none', ms=26, markeredgecolor=sns.color_palette()[1], label='Fe responsive')\n",
    "axis.plot('x', 'y', '.', data=plot_df, markerfacecolor='w', markeredgecolor='none', zorder=0, label='  ')\n",
    "axis.plot('x', 'y', 'ks', data=plot_df[plot_df['ExpressionResponse'] == 'N responsive'], markeredgewidth=3, \n",
    "          markerfacecolor='none', ms=26, markeredgecolor=sns.color_palette()[7], label='N responsive')\n",
    "\n",
    "axis.set(title='Gene Weights', xlabel=None, ylabel=None, xlim=(-0.6, n_genes-0.4), ylim=(-0.5, 3.5))\n",
    "axis.set_xticks(np.arange(n_genes), plot_df.GeneLabel.unique(), rotation=90)\n",
    "axis.tick_params(left=False, labelleft=False, bottom=False)\n",
    "axis.legend(title='     Median\\nGene Weight', loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "\n",
    "\n",
    "# panel c: surface transects + environmental covariates\n",
    "sample_weight_range = (-0.05, 0.7) # range of weights to show on sample plots\n",
    "for i, year in enumerate([2016, 2017, 2019]):\n",
    "    # Fe limitation cluster\n",
    "    axis = plt.subplot(grid[7:, 7*i:7*i+2])\n",
    "    # add annotation\n",
    "    if i == 0:\n",
    "        axis.annotate('C', xy=(-0.5, 1.1), **annot_kwgs)\n",
    "    # get data together\n",
    "    plot_df = fe_cluster_df[fe_cluster_df['SampleGroup'] == 'surface transect']\n",
    "    plot_df = plot_df[plot_df['TimeClass'].isin(time_classes)]\n",
    "    plot_df = plot_df[plot_df['Year'] == year]\n",
    "    latitudes = plot_df[\n",
    "        ['SampleName', 'Latitude']].drop_duplicates().set_index('SampleName').sort_values('Latitude')['Latitude']\n",
    "    lat_bins = unique_bins(latitudes, n_bins, bounds=bounds, reverse=False)\n",
    "    plot_df['lat_idx'] = plot_df['SampleName'].map(dict(zip(latitudes.index, lat_bins)))\n",
    "    # make figure\n",
    "    sns.boxplot(\n",
    "        data=plot_df, x='SampleWeight', y='lat_idx', hue='Genus', hue_order=['Prochlorococcus', 'Synechococcus'], \n",
    "        palette=[sns.color_palette()[i] for i in [2, 3]], orient='y', order=np.arange(n_bins), \n",
    "        fill=False, whis=(5, 95), fliersize=fliersize, legend=False, ax=axis\n",
    "    )\n",
    "    # set labels and such\n",
    "    axis.set_yticks(np.arange(1, n_bins, 2), [f'{int(l)}N' for l in np.arange(bounds[0]+1, bounds[1]+1, 2)])\n",
    "    axis.set(xlim=sample_weight_range, ylim=(-0.5, n_bins-0.5), xlabel='Weight', ylabel='', title='Fe Clusters')\n",
    "    axis.set_yticks(np.arange(n_bins), minor=True)\n",
    "    \n",
    "    # N acquisition cluster\n",
    "    axis = plt.subplot(grid[7:, 7*i+2:7*i+4])\n",
    "    # add title\n",
    "    if i == 1:\n",
    "        axis.annotate('Surface Transects', xy=(.5, 1.1), xycoords='axes fraction', va='center', ha='center', fontsize=22)\n",
    "    # get data together\n",
    "    plot_df = n_cluster_df[n_cluster_df['SampleGroup'] == 'surface transect']\n",
    "    plot_df = plot_df[plot_df['TimeClass'].isin(time_classes)]\n",
    "    plot_df = plot_df[plot_df['Year'] == year]\n",
    "    plot_df['lat_idx'] = plot_df['SampleName'].map(dict(zip(latitudes.index, lat_bins)))\n",
    "    # make figure\n",
    "    sns.boxplot(\n",
    "        data=plot_df, x='SampleWeight', y='lat_idx', hue='Genus', hue_order=['Prochlorococcus', 'Synechococcus'], \n",
    "        palette=[sns.color_palette()[i] for i in [2, 3]], orient='y', order=np.arange(n_bins), fill=False, \n",
    "        whis=(5, 95), fliersize=fliersize, legend=False, ax=axis\n",
    "    )\n",
    "    # set labels and such\n",
    "    axis.set_yticks(np.arange(1, n_bins, 2))\n",
    "    axis.set(xlim=sample_weight_range, ylim=(-0.5, n_bins-0.5), xlabel='Weight', ylabel='', title=f\"{year}\\nN Clusters\")\n",
    "    axis.set_yticks(np.arange(n_bins), minor=True)\n",
    "    axis.tick_params(left=True, labelleft=False)\n",
    "\n",
    "    # environmental covariates\n",
    "    axis = plt.subplot(grid[7:, 7*i+4:7*i+6])\n",
    "    # plot salinity front\n",
    "    axis.hlines(y=sal_fronts[year], xmin=0, xmax=2, linewidth=10, color=sns.color_palette()[5], alpha=0.2, label='Salinity Front')\n",
    "    # plot chlorophyll front\n",
    "    axis.hlines(y=chl_fronts[year], xmin=0, xmax=2, linewidth=10, color=sns.color_palette()[2], alpha=0.2, label='Chlorophyll Front')\n",
    "    \n",
    "    # dissolved iron\n",
    "    color = sns.color_palette()[1]\n",
    "    sns.lineplot(\n",
    "        data=fe_df[(fe_df['year'] == year) & (fe_df['depth'].le(depth_limit))], x='Fe_dissolved', \n",
    "        y='lat', orient='y', ax=axis, color=color, marker='.', label='[Fe] (nmol/L)', legend=(i==2)\n",
    "    )\n",
    "    axis.set(xlim=(-.05, 0.85), ylim=(bounds[0]-0.5, bounds[1]+.5), yticks=np.arange(1, n_bins, 2)+bounds[0], \n",
    "             xlabel='[Fe]', ylabel='')\n",
    "    axis.set_yticks(np.arange(n_bins)+bounds[0], minor=True)\n",
    "    axis.tick_params(left=True, labelleft=False)\n",
    "    axis.tick_params(axis='x', colors=color)\n",
    "    axis.xaxis.label.set_color(color)\n",
    "    axis.spines['bottom'].set_color(color)\n",
    "    if i == 2:\n",
    "        axis.legend(loc='upper left', bbox_to_anchor=(2.3, -0.1))\n",
    "\n",
    "    # dissolved nitrate + nitrite\n",
    "    color = sns.color_palette()[7]\n",
    "    axis2 = plt.twiny(axis)\n",
    "    sns.lineplot(\n",
    "        data=n_df[(n_df['year'] == year) & (n_df['depth'].le(depth_limit))], x='NO3_NO2', y='lat', orient='y', \n",
    "        ax=axis2, color=color, marker='.', label='[NO$_3^-$+NO$_2^-$] (umol/L)', legend=(i==2)\n",
    "    )\n",
    "    axis2.set(xlim=(-0.5, 7.5), ylim=(bounds[0]-0.5, bounds[1]+.5), yticks=np.arange(1, n_bins, 2)+bounds[0], \n",
    "             xlabel='[NO$_3^-$+NO$_2^-$]', ylabel='')\n",
    "    axis2.set_yticks(np.arange(n_bins)+bounds[0], minor=True)\n",
    "    axis2.tick_params(top=False, labeltop=False, bottom=True, labelbottom=True)\n",
    "    axis2.tick_params(axis='x', colors=color)\n",
    "    axis2.xaxis.set_label_position('bottom')\n",
    "    axis2.xaxis.label.set_color(color)\n",
    "    axis2.spines['bottom'].set_color(color)\n",
    "    axis2.spines['bottom'].set_position(('outward', 60))\n",
    "    if i == 2:\n",
    "        axis2.legend(loc='upper left', bbox_to_anchor=(2.3, -0.205))\n",
    "\n",
    "    # N:Fe ratio\n",
    "    color = sns.color_palette()[4]\n",
    "    axis3 = plt.twiny(axis2)\n",
    "    sns.lineplot(\n",
    "        data=merged_df[(merged_df['year'] == year) & (merged_df['depth'].le(depth_limit))], x='N:Fe', y='lat', orient='y', \n",
    "        ax=axis3, color=color, marker='.', label='[NO$_3^-$+NO$_2^-$]:[Fe] Ratio', legend=(i==2)\n",
    "    )\n",
    "    axis3.set(xlim=(1, 1000000), ylim=(bounds[0]-0.5, bounds[1]+.5), yticks=np.arange(1, n_bins, 2)+bounds[0], \n",
    "             xlabel='[NO$_3^-$+NO$_2^-$]:[Fe]', ylabel='', xscale='log')\n",
    "    axis3.set_yticks(np.arange(n_bins)+bounds[0], minor=True)\n",
    "    axis3.tick_params(top=False, labeltop=False, bottom=True, labelbottom=True)\n",
    "    axis3.tick_params(axis='x', colors=color)\n",
    "    axis3.xaxis.set_label_position('bottom')\n",
    "    axis3.xaxis.label.set_color(color)\n",
    "    axis3.spines['bottom'].set_color(color)\n",
    "    axis3.spines['bottom'].set_position(('outward', 120))\n",
    "    if i == 2:\n",
    "        axis3.legend(loc='upper left', bbox_to_anchor=(2.3, -0.24))\n",
    "\n",
    "# panel d: incubation weights\n",
    "for i, condition in enumerate(['HNLC', 'TZ', 'Gyre']):\n",
    "    for j, clustergroup in enumerate(['Fe', 'N']):\n",
    "        axis = plt.subplot(grid[i*3+7:i*3+10, 23+(2*j):25+(2*j)])\n",
    "        plot_df = incubation_df[(incubation_df['Condition'] == condition) & (incubation_df['ClusterGroup'] == clustergroup)]\n",
    "        sns.boxplot(\n",
    "            data=plot_df, x='SampleWeight', y='Treatment', hue='Genus', \n",
    "            hue_order=['Prochlorococcus', 'Synechococcus'], palette=[sns.color_palette()[i] for i in [2, 3]], \n",
    "            orient='y', fill=False, whis=(5, 95), fliersize=fliersize, legend=(i==2 and j==0), ax=axis\n",
    "        )\n",
    "        # labels and such\n",
    "        if i == 0:\n",
    "            axis.set(title='{} Clusters'.format(clustergroup))\n",
    "            if j == 0:\n",
    "                axis.annotate('D', xy=(-1.5, 1.35), **annot_kwgs)\n",
    "            if j == 1:\n",
    "                axis.annotate('Incubations (2017)', xy=(0, 1.35), xycoords='axes fraction', va='center', ha='center', fontsize=22)\n",
    "        if i == 2:\n",
    "            xlabel = 'Weight'\n",
    "            if j == 0:\n",
    "                axis.legend(loc='upper left', bbox_to_anchor=(-0.52, -0.35))\n",
    "        else:\n",
    "            xlabel = None\n",
    "        if j == 0:\n",
    "            ylabel = f\"{condition} ({plot_df.Latitude.mean():.0f}N)\"\n",
    "        else:\n",
    "            ylabel = None\n",
    "        axis.set(xlim=sample_weight_range, xlabel=xlabel, ylabel=ylabel)\n",
    "        axis.tick_params(left=True, labelleft=(j==0), bottom=True, labelbottom=(i==2))\n",
    "\n",
    "fig.savefig(outdir / 'fe-n-clusters.png', bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a84004c8-bc54-43e8-ba22-3bdb1d10856f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25a2bb0d-ef94-4b34-855c-11b0a0b67a50",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
