{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Sources\n",
    "\n",
    "This notebook documents the compilation of metadata products used as input to the analyses included in the project.\n",
    "- All external data were obtained from publically available sources. URLs and download instructions are included to aid in reproducing the provenance of these data. \n",
    "- All data generated by the Armbrust lab were posted to publically available resources. URLs and download instructions are included for these data as well.\n",
    "- The included code combines and cleans data sets, documents the reconciliation of any discrepancies, and produces cleaned metadata files that are used as inputs for project analyses.\n",
    "\n",
    "## Reference Genomes & CyCOGs\n",
    "\n",
    "All reference genomes in this project were derived from the Scientific Data publication [Single cell genomes of Prochlorococcus, Synechococcus, and sympatric microbes from diverse marine environments](https://www.nature.com/articles/sdata2018154) (Berube et al., 2018). The genome sequences and annotations used in this publication are hosted by JGI on the [IMG/ProPortal](https://img.jgi.doe.gov/cgi-bin/proportal/main.cgi) and can be downloaded using the [IMG Genome Cart tool](https://img.jgi.doe.gov/cgi-bin/mer/main.cgi?section=GenomeCart&page=genomeCart)\n",
    "\n",
    "### Figshare Files\n",
    "\n",
    "Download the following files from the [figshare site associated with Berube et al. (2018)](https://doi.org/10.6084/m9.figshare.c.4037048.v1):\n",
    "- File 4: [cycogs-genomes.tsv](https://figshare.com/articles/dataset/File_4_CyCOG_taxa/6007166)\n",
    "- File 5: [cycogs.tsv](https://figshare.com/articles/dataset/File_5_CyCOG_definitions/6007169)\n",
    "- File 14: [genome_assembly_summary_20180718.tsv](https://figshare.com/articles/dataset/File_14_Genome_assembly_summary/6281519)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CACHE\n",
    "\n",
    "I don't think I need this stuff, but stashing here in case\n",
    "\n",
    "### Publication Files\n",
    "- [Table 1](https://www.nature.com/articles/sdata2018154/tables/2)\n",
    "- [Table 2](https://www.nature.com/articles/sdata2018154/tables/3)\n",
    "\n",
    "\n",
    "### JGI IMG/ProPortal Files\n",
    "\n",
    "Download the metadata associated with all ProPortal reference genomes from the [IMG/ProPortal Genome Browser](https://img.jgi.doe.gov/cgi-bin/proportal/main.cgi?section=ProPortal&page=genomeList&class=datamart)\n",
    "- Use the `Table Configuration` section (bottom of the page) to display the following fields: \n",
    "    - Domain\n",
    "    - Sequencing Status\n",
    "    - Study Name\n",
    "    - Genome Name / Sample Name\n",
    "    - Sequencing Center\n",
    "    - IMG Genome ID (IMG Taxon ID)\n",
    "    - Genus\n",
    "    - Ecosystem Subtype\n",
    "    - Culture Type\n",
    "    - Depth In Meters\n",
    "    - Latitude\n",
    "    - Longitude\n",
    "    - Proportal Clade\n",
    "    - Proportal Ecotype\n",
    "    - Genome Size (Number of total bases)\n",
    "    - Gene Count (Number of total genes)\n",
    "    - CheckM2 Completeness\n",
    "- Click `Select All` and `Export` to download the metadata\n",
    "    - Name the file `proportal-metadata.tsv` to make it recognizeable by the following scripts\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean Genome Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genomeid</th>\n",
       "      <th>genomename</th>\n",
       "      <th>type</th>\n",
       "      <th>group</th>\n",
       "      <th>clade</th>\n",
       "      <th>virocell</th>\n",
       "      <th>completeness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2716884698</td>\n",
       "      <td>AG-316-L16</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>20.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2716884700</td>\n",
       "      <td>AG-316-N23</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>20.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2716884701</td>\n",
       "      <td>AG-316-P23</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>12.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2716884699</td>\n",
       "      <td>AG-316-L21</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>10.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2716884642</td>\n",
       "      <td>AG-316-A05</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>AMZ-II</td>\n",
       "      <td>False</td>\n",
       "      <td>6.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>765</th>\n",
       "      <td>651703106</td>\n",
       "      <td>S-ShM2</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>651703107</td>\n",
       "      <td>Syn19</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>767</th>\n",
       "      <td>641201056</td>\n",
       "      <td>Syn5</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>768</th>\n",
       "      <td>2595698410</td>\n",
       "      <td>metaG-MbCM1</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>769</th>\n",
       "      <td>638275565</td>\n",
       "      <td>Syn9</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>770 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       genomeid   genomename     type            group   clade  virocell  \\\n",
       "0    2716884698   AG-316-L16      SAG  Prochlorococcus  AMZ-II     False   \n",
       "1    2716884700   AG-316-N23      SAG  Prochlorococcus  AMZ-II     False   \n",
       "2    2716884701   AG-316-P23      SAG  Prochlorococcus  AMZ-II     False   \n",
       "3    2716884699   AG-316-L21      SAG  Prochlorococcus  AMZ-II     False   \n",
       "4    2716884642   AG-316-A05      SAG  Prochlorococcus  AMZ-II     False   \n",
       "..          ...          ...      ...              ...     ...       ...   \n",
       "765   651703106       S-ShM2  ISOLATE            Virus     NaN     False   \n",
       "766   651703107        Syn19  ISOLATE            Virus     NaN     False   \n",
       "767   641201056         Syn5  ISOLATE            Virus     NaN     False   \n",
       "768  2595698410  metaG-MbCM1  ISOLATE            Virus     NaN     False   \n",
       "769   638275565         Syn9  ISOLATE            Virus     NaN     False   \n",
       "\n",
       "     completeness  \n",
       "0           20.69  \n",
       "1           20.69  \n",
       "2           12.07  \n",
       "3           10.34  \n",
       "4            6.90  \n",
       "..            ...  \n",
       "765          0.00  \n",
       "766          0.00  \n",
       "767          0.00  \n",
       "768          0.00  \n",
       "769          0.00  \n",
       "\n",
       "[770 rows x 7 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clean genome metadata\n",
    "\n",
    "# combine genome metadata from Berube et al. 2018 publication\n",
    "genome_df = pd.merge(\n",
    "    left=pd.read_csv('metadata/cycogsgenomes.tsv', sep='\\t'), \n",
    "    right=pd.read_csv('metadata/genome_assembly_summary_20180718.tsv', sep='\\t'), \n",
    "    left_on='IMG_ID',\n",
    "    right_on='img_genome_id', \n",
    "    how='left'\n",
    ")\n",
    "\n",
    "# add updated clade assignments\n",
    "genome_df['clade'] = genome_df['IMG_ID'].map(\n",
    "    pd.read_csv('metadata/updated-genome-clades.csv', sep=',').set_index('img_genome_id')['clade']\n",
    ")\n",
    "\n",
    "# standardize field names\n",
    "genome_df.rename(\n",
    "    columns={\n",
    "        'IMG_ID':'genomeid', \n",
    "        'IID':'genomename', \n",
    "        'GROUP':'group', \n",
    "        'TYPE':'type', \n",
    "        'Completeness':'completeness', \n",
    "    }, \n",
    "    inplace=True\n",
    ")\n",
    "\n",
    "# add virocell field\n",
    "genome_df['virocell'] = (genome_df['usage_notes'] == 'likely virocell')\n",
    "\n",
    "# down select only needed fields & sort\n",
    "genome_df = genome_df[['genomeid', 'genomename', 'type', 'group', 'clade', 'virocell', 'completeness']].sort_values(\n",
    "    ['group', 'clade', 'type', 'completeness'], ascending=[True, True, False, False]\n",
    ").reset_index(drop=True)\n",
    "\n",
    "# save cleaned up genome dataframe\n",
    "genome_df.to_csv('metadata/genome-metadata.csv', index=False)\n",
    "\n",
    "genome_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean Ortholog Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mappingname</th>\n",
       "      <th>cycogid</th>\n",
       "      <th>genomename</th>\n",
       "      <th>geneid</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WH8102_2607658325</td>\n",
       "      <td>60000001</td>\n",
       "      <td>WH8102</td>\n",
       "      <td>2607658325</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIT0917_2681971350</td>\n",
       "      <td>60000001</td>\n",
       "      <td>MIT0917</td>\n",
       "      <td>2681971350</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-424-P18_2717338506</td>\n",
       "      <td>60000001</td>\n",
       "      <td>AG-424-P18</td>\n",
       "      <td>2717338506</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scB245a_521A19_2655604637</td>\n",
       "      <td>60000001</td>\n",
       "      <td>scB245a_521A19</td>\n",
       "      <td>2655604637</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GFB01_2638208352</td>\n",
       "      <td>60000001</td>\n",
       "      <td>GFB01</td>\n",
       "      <td>2638208352</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964917</th>\n",
       "      <td>AG-363-C02_2667889608</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667889608</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964918</th>\n",
       "      <td>AG-363-C02_2667889615</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667889615</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964919</th>\n",
       "      <td>AG-363-C02_2667890048</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667890048</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964920</th>\n",
       "      <td>AG-363-C02_2667890054</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667890054</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964921</th>\n",
       "      <td>AG-363-C02_2667890060</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667890060</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>964922 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      mappingname   cycogid      genomename      geneid  \\\n",
       "0               WH8102_2607658325  60000001          WH8102  2607658325   \n",
       "1              MIT0917_2681971350  60000001         MIT0917  2681971350   \n",
       "2           AG-424-P18_2717338506  60000001      AG-424-P18  2717338506   \n",
       "3       scB245a_521A19_2655604637  60000001  scB245a_521A19  2655604637   \n",
       "4                GFB01_2638208352  60000001           GFB01  2638208352   \n",
       "...                           ...       ...             ...         ...   \n",
       "964917      AG-363-C02_2667889608  60040295      AG-363-C02  2667889608   \n",
       "964918      AG-363-C02_2667889615  60040295      AG-363-C02  2667889615   \n",
       "964919      AG-363-C02_2667890048  60040295      AG-363-C02  2667890048   \n",
       "964920      AG-363-C02_2667890054  60040295      AG-363-C02  2667890054   \n",
       "964921      AG-363-C02_2667890060  60040295      AG-363-C02  2667890060   \n",
       "\n",
       "                                      annotation  \n",
       "0       membrane protease FtsH catalytic subunit  \n",
       "1       membrane protease FtsH catalytic subunit  \n",
       "2       membrane protease FtsH catalytic subunit  \n",
       "3       membrane protease FtsH catalytic subunit  \n",
       "4       membrane protease FtsH catalytic subunit  \n",
       "...                                          ...  \n",
       "964917                      hypothetical protein  \n",
       "964918                      hypothetical protein  \n",
       "964919                      hypothetical protein  \n",
       "964920                      hypothetical protein  \n",
       "964921                      hypothetical protein  \n",
       "\n",
       "[964922 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clean and join\n",
    "\n",
    "# read in cycog data\n",
    "filepath_cycogs = 'metadata/cycogs.tsv'\n",
    "cycog_df = pd.read_csv(filepath_cycogs, sep='\\t')\n",
    "cycog_df['cycogid'] =[int(id_val[6:]) for id_val in cycog_df['cycog_iid']]\n",
    "\n",
    "# make dictionary of cycog genes\n",
    "cog_dict = {}\n",
    "for _, row in cycog_df.iterrows():\n",
    "    cog = int(row['cycog_iid'][6:])\n",
    "    for gene in row['cycog_genes'].split(','):\n",
    "        cog_dict[gene] = cog\n",
    "\n",
    "# make dataframe from cog dict       \n",
    "cogs_df = pd.DataFrame.from_dict(cog_dict, dtype=int, orient='index', columns=['cycog_iid'])\n",
    "\n",
    "# clean up data columns\n",
    "cogs_df.reset_index(inplace=True)\n",
    "cogs_df.rename(columns={'index':'mappingname', 'cycog_iid':'cycogid'}, inplace=True)\n",
    "cogs_df['genomename'], cogs_df['geneid'] = list(zip(*cogs_df['mappingname'].str.rsplit(pat='_', n=1)))\n",
    "\n",
    "# join in annotation information\n",
    "cogs_df['annotation'] = cogs_df.cycogid.map(cycog_df.set_index('cycogid')['cycog_cns_product'])\n",
    "\n",
    "# save cleaned ortholog metadata\n",
    "cogs_df.to_csv('metadata/ortholog-metadata.csv', index=False)\n",
    "\n",
    "cogs_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
