{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Sources\n",
    "\n",
    "This notebook documents the compilation of metadata products used as input to the analyses included in the project.\n",
    "- All external data were obtained from publically available sources. URLs and download instructions are included to aid in reproducing the provenance of these data. \n",
    "- All data generated by the Armbrust lab were posted to publically available resources. URLs and download instructions are included for these data as well.\n",
    "- The included code combines and cleans data sets, documents the reconciliation of any discrepancies, and produces cleaned metadata files that are used as inputs for project analyses.\n",
    "\n",
    "## Reference Genomes & CyCOGs\n",
    "\n",
    "All reference genomes in this project were derived from the Scientific Data publication [Single cell genomes of Prochlorococcus, Synechococcus, and sympatric microbes from diverse marine environments](https://www.nature.com/articles/sdata2018154) (Berube et al., 2018). The genome sequences and annotations used in this publication are hosted by JGI on the [IMG/ProPortal](https://img.jgi.doe.gov/cgi-bin/proportal/main.cgi) and can be downloaded using the [IMG Genome Cart tool](https://img.jgi.doe.gov/cgi-bin/mer/main.cgi?section=GenomeCart&page=genomeCart)\n",
    "\n",
    "### Figshare Files\n",
    "\n",
    "Download the following files from the [figshare site associated with Berube et al. (2018)](https://doi.org/10.6084/m9.figshare.c.4037048.v1):\n",
    "- File 4: [cycogs-genomes.tsv](https://figshare.com/articles/dataset/File_4_CyCOG_taxa/6007166)\n",
    "- File 5: [cycogs.tsv](https://figshare.com/articles/dataset/File_5_CyCOG_definitions/6007169)\n",
    "- File 14: [genome_assembly_summary_20180718.tsv](https://figshare.com/articles/dataset/File_14_Genome_assembly_summary/6281519)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CACHE\n",
    "\n",
    "I don't think I need this stuff, but stashing here in case\n",
    "\n",
    "### Publication Files\n",
    "- [Table 1](https://www.nature.com/articles/sdata2018154/tables/2)\n",
    "- [Table 2](https://www.nature.com/articles/sdata2018154/tables/3)\n",
    "\n",
    "\n",
    "### JGI IMG/ProPortal Files\n",
    "\n",
    "Download the metadata associated with all ProPortal reference genomes from the [IMG/ProPortal Genome Browser](https://img.jgi.doe.gov/cgi-bin/proportal/main.cgi?section=ProPortal&page=genomeList&class=datamart)\n",
    "- Use the `Table Configuration` section (bottom of the page) to display the following fields: \n",
    "    - Domain\n",
    "    - Sequencing Status\n",
    "    - Study Name\n",
    "    - Genome Name / Sample Name\n",
    "    - Sequencing Center\n",
    "    - IMG Genome ID (IMG Taxon ID)\n",
    "    - Genus\n",
    "    - Ecosystem Subtype\n",
    "    - Culture Type\n",
    "    - Depth In Meters\n",
    "    - Latitude\n",
    "    - Longitude\n",
    "    - Proportal Clade\n",
    "    - Proportal Ecotype\n",
    "    - Genome Size (Number of total bases)\n",
    "    - Gene Count (Number of total genes)\n",
    "    - CheckM2 Completeness\n",
    "- Click `Select All` and `Export` to download the metadata\n",
    "    - Name the file `proportal-metadata.tsv` to make it recognizeable by the following scripts\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'5.1A-1': '5.1A-1',\n",
       " '5.1A-I': '5.1A-I',\n",
       " '5.1A-II': '5.1A-II',\n",
       " '5.1A-III': '5.1A-III',\n",
       " '5.1A-III/XV': '5.1A-III/XV',\n",
       " '5.1A-IV': '5.1A-IV',\n",
       " '5.1B-VII': '5.1B-VII',\n",
       " 'CDR2': 'CDR2',\n",
       " 'HLI': 'HLI',\n",
       " 'HLII': 'HLII',\n",
       " 'HLIII': 'HLIII',\n",
       " 'HLVI': 'HLVI',\n",
       " 'LLI': 'LLI',\n",
       " 'LLII/III': 'LLII/III',\n",
       " 'LLIV': 'LLIV',\n",
       " 'unclassified': 'unclassified',\n",
       " nan: nan}"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lister = genome_df['clade'].sort_values().unique()\n",
    "\n",
    "dict(zip(lister, lister))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IID</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>IMG_ID</th>\n",
       "      <th>Completeness</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>usage_notes</th>\n",
       "      <th>ecotype</th>\n",
       "      <th>clade</th>\n",
       "      <th>selection_criteria</th>\n",
       "      <th>checkm_contamination</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>AG-341-K05</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>2716884261</td>\n",
       "      <td>8.62</td>\n",
       "      <td>8.62</td>\n",
       "      <td>likely virocell</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLI</td>\n",
       "      <td>wgs amplification</td>\n",
       "      <td>0.00</td>\n",
       "      <td>JGI Microbial Genome Annotation Pipeline (MGAP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>AG-363-L17</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>2667527373</td>\n",
       "      <td>57.20</td>\n",
       "      <td>57.20</td>\n",
       "      <td>likely virocell</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLVI</td>\n",
       "      <td>wgs amplification</td>\n",
       "      <td>0.54</td>\n",
       "      <td>JGI Microbial Genome Annotation Pipeline (MGAP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>AG-418-C09</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>2716884766</td>\n",
       "      <td>91.30</td>\n",
       "      <td>91.30</td>\n",
       "      <td>likely virocell</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>ITS phylogeny</td>\n",
       "      <td>0.54</td>\n",
       "      <td>JGI Microbial Genome Annotation Pipeline (MGAP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>AG-418-C17</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>2716884767</td>\n",
       "      <td>52.47</td>\n",
       "      <td>52.47</td>\n",
       "      <td>likely virocell</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLII</td>\n",
       "      <td>ITS phylogeny</td>\n",
       "      <td>0.82</td>\n",
       "      <td>JGI Microbial Genome Annotation Pipeline (MGAP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-311-D23</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>2716884681</td>\n",
       "      <td>72.96</td>\n",
       "      <td>72.96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLI</td>\n",
       "      <td>wgs amplification</td>\n",
       "      <td>0.00</td>\n",
       "      <td>JGI Microbial Genome Annotation Pipeline (MGAP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>765</th>\n",
       "      <td>S-ShM2</td>\n",
       "      <td>Virus</td>\n",
       "      <td>651703106</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>Syn19</td>\n",
       "      <td>Virus</td>\n",
       "      <td>651703107</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>767</th>\n",
       "      <td>Syn5</td>\n",
       "      <td>Virus</td>\n",
       "      <td>641201056</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>768</th>\n",
       "      <td>metaG-MbCM1</td>\n",
       "      <td>Virus</td>\n",
       "      <td>2595698410</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>769</th>\n",
       "      <td>Syn9</td>\n",
       "      <td>Virus</td>\n",
       "      <td>638275565</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>770 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             IID            GROUP      IMG_ID  Completeness  \\\n",
       "87    AG-341-K05  Prochlorococcus  2716884261          8.62   \n",
       "209   AG-363-L17  Prochlorococcus  2667527373         57.20   \n",
       "332   AG-418-C09  Prochlorococcus  2716884766         91.30   \n",
       "333   AG-418-C17  Prochlorococcus  2716884767         52.47   \n",
       "0     AG-311-D23  Prochlorococcus  2716884681         72.96   \n",
       "..           ...              ...         ...           ...   \n",
       "765       S-ShM2            Virus   651703106          0.00   \n",
       "766        Syn19            Virus   651703107          0.00   \n",
       "767         Syn5            Virus   641201056          0.00   \n",
       "768  metaG-MbCM1            Virus  2595698410          0.00   \n",
       "769         Syn9            Virus   638275565          0.00   \n",
       "\n",
       "     checkm_completeness      usage_notes                  ecotype  \\\n",
       "87                  8.62  likely virocell   Low light adapted (LL)   \n",
       "209                57.20  likely virocell  High light adapted (HL)   \n",
       "332                91.30  likely virocell  High light adapted (HL)   \n",
       "333                52.47  likely virocell  High light adapted (HL)   \n",
       "0                  72.96              NaN   Low light adapted (LL)   \n",
       "..                   ...              ...                      ...   \n",
       "765                  NaN              NaN                      NaN   \n",
       "766                  NaN              NaN                      NaN   \n",
       "767                  NaN              NaN                      NaN   \n",
       "768                  NaN              NaN                      NaN   \n",
       "769                  NaN              NaN                      NaN   \n",
       "\n",
       "            clade selection_criteria  checkm_contamination  \\\n",
       "87            LLI  wgs amplification                  0.00   \n",
       "209          HLVI  wgs amplification                  0.54   \n",
       "332  unclassified      ITS phylogeny                  0.54   \n",
       "333          HLII      ITS phylogeny                  0.82   \n",
       "0             LLI  wgs amplification                  0.00   \n",
       "..            ...                ...                   ...   \n",
       "765           NaN                NaN                   NaN   \n",
       "766           NaN                NaN                   NaN   \n",
       "767           NaN                NaN                   NaN   \n",
       "768           NaN                NaN                   NaN   \n",
       "769           NaN                NaN                   NaN   \n",
       "\n",
       "                                            annotation  \n",
       "87   JGI Microbial Genome Annotation Pipeline (MGAP...  \n",
       "209  JGI Microbial Genome Annotation Pipeline (MGAP...  \n",
       "332  JGI Microbial Genome Annotation Pipeline (MGAP...  \n",
       "333  JGI Microbial Genome Annotation Pipeline (MGAP...  \n",
       "0    JGI Microbial Genome Annotation Pipeline (MGAP...  \n",
       "..                                                 ...  \n",
       "765                                                NaN  \n",
       "766                                                NaN  \n",
       "767                                                NaN  \n",
       "768                                                NaN  \n",
       "769                                                NaN  \n",
       "\n",
       "[770 rows x 11 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_df[['IID', 'GROUP', 'IMG_ID', 'Completeness', 'checkm_completeness', 'usage_notes', \n",
    "           'ecotype', 'clade', \n",
    "           'selection_criteria', 'checkm_contamination', 'annotation']].sort_values(\n",
    "    'usage_notes', ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['IID', 'GROUP', 'IMG_ID', 'TYPE', 'JGI_GENOMEPORTAL_NAME',\n",
       "       'Completeness', 'sag_id', 'phylogeny', 'ecotype', 'clade',\n",
       "       'img_genome_id', 'usage_notes', 'selection_criteria',\n",
       "       'completeness_software', 'completeness_score', 'checkm_completeness',\n",
       "       'checkm_contamination', 'checkm_strain_heterogeneity',\n",
       "       'checkm_genome_size--bp', 'checkm_scaffold_number', 'checkm_percent_gc',\n",
       "       'checkm_predicted_genes', 'checkm_tree_placement_taxonomy',\n",
       "       'checkm_marker_set_used_for_stats', '5S_rRNA_count', '16S_rRNA_count',\n",
       "       '23S_rRNA_count', 'total_standard_tRNA_count',\n",
       "       'non-redundant_standard_tRNA_count', 'tRNA_extraction_software',\n",
       "       'analysis_project_type', 'taxa_id', 'assembly_software',\n",
       "       'assembly_method_version', 'annotation', 'assembly_quality',\n",
       "       'cell_isolation_approach', 'single_cell_lysis_approach',\n",
       "       'wga_amp_approach', 'wga_amp_protocol', 'ncbi_biosample_accession',\n",
       "       'sequencing_coverage', 'seq_meth', 'env_biome', 'env_feature',\n",
       "       'env_material', 'env_package'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean Genome Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>group</th>\n",
       "      <th>ecotype</th>\n",
       "      <th>clade</th>\n",
       "      <th>virocell</th>\n",
       "      <th>completeness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2716884503</td>\n",
       "      <td>AG-679-M23</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>False</td>\n",
       "      <td>99.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2716884367</td>\n",
       "      <td>AG-388-E21</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>False</td>\n",
       "      <td>98.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2716884409</td>\n",
       "      <td>AG-418-M08</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>False</td>\n",
       "      <td>98.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2716884731</td>\n",
       "      <td>AG-388-J02</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>False</td>\n",
       "      <td>98.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2716884783</td>\n",
       "      <td>AG-442-D10</td>\n",
       "      <td>SAG</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>False</td>\n",
       "      <td>96.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>765</th>\n",
       "      <td>651703106</td>\n",
       "      <td>S-ShM2</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>651703107</td>\n",
       "      <td>Syn19</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>767</th>\n",
       "      <td>641201056</td>\n",
       "      <td>Syn5</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>768</th>\n",
       "      <td>2595698410</td>\n",
       "      <td>metaG-MbCM1</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>769</th>\n",
       "      <td>638275565</td>\n",
       "      <td>Syn9</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>770 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             id         name     type            group  \\\n",
       "0    2716884503   AG-679-M23      SAG  Prochlorococcus   \n",
       "1    2716884367   AG-388-E21      SAG  Prochlorococcus   \n",
       "2    2716884409   AG-418-M08      SAG  Prochlorococcus   \n",
       "3    2716884731   AG-388-J02      SAG  Prochlorococcus   \n",
       "4    2716884783   AG-442-D10      SAG  Prochlorococcus   \n",
       "..          ...          ...      ...              ...   \n",
       "765   651703106       S-ShM2  ISOLATE            Virus   \n",
       "766   651703107        Syn19  ISOLATE            Virus   \n",
       "767   641201056         Syn5  ISOLATE            Virus   \n",
       "768  2595698410  metaG-MbCM1  ISOLATE            Virus   \n",
       "769   638275565         Syn9  ISOLATE            Virus   \n",
       "\n",
       "                     ecotype clade  virocell  completeness  \n",
       "0    High light adapted (HL)   HLI     False         99.73  \n",
       "1    High light adapted (HL)   HLI     False         98.91  \n",
       "2    High light adapted (HL)   HLI     False         98.51  \n",
       "3    High light adapted (HL)   HLI     False         98.37  \n",
       "4    High light adapted (HL)   HLI     False         96.47  \n",
       "..                       ...   ...       ...           ...  \n",
       "765                      NaN   NaN     False          0.00  \n",
       "766                      NaN   NaN     False          0.00  \n",
       "767                      NaN   NaN     False          0.00  \n",
       "768                      NaN   NaN     False          0.00  \n",
       "769                      NaN   NaN     False          0.00  \n",
       "\n",
       "[770 rows x 8 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clean genome metadata\n",
    "\n",
    "genome_df = pd.merge(\n",
    "    left=pd.read_csv('metadata/cycogsgenomes.tsv', sep='\\t'), \n",
    "    right=pd.read_csv('metadata/genome_assembly_summary_20180718.tsv', sep='\\t'), \n",
    "    left_on='IMG_ID',\n",
    "    right_on='img_genome_id', \n",
    "    how='left'\n",
    ")\n",
    "\n",
    "# standardize field names\n",
    "genome_df.rename(\n",
    "    columns={\n",
    "        'IMG_ID':'id', \n",
    "        'IID':'name', \n",
    "        'GROUP':'group', \n",
    "        'TYPE':'type', \n",
    "        'Completeness':'completeness', \n",
    "    }, \n",
    "    inplace=True\n",
    ")\n",
    "\n",
    "# add virocell field\n",
    "genome_df['virocell'] = (genome_df['usage_notes'] == 'likely virocell')\n",
    "\n",
    "# fix clades\n",
    "genome_df['clade'] = genome_df['clade'].map({\n",
    "    '5.1A-1': '5.1A-I',\n",
    "    '5.1A-I': '5.1A-I',\n",
    "    '5.1A-II': '5.1A-II',\n",
    "    '5.1A-III': '5.1A-III',\n",
    "    '5.1A-III/XV': '5.1A-III/XV',\n",
    "    '5.1A-IV': '5.1A-IV',\n",
    "    '5.1B-VII': '5.1B-VII',\n",
    "    'CDR2': 'CRD2',\n",
    "    'HLI': 'HLI',\n",
    "    'HLII': 'HLII',\n",
    "    'HLIII': 'HLIII',\n",
    "    'HLVI': 'HLVI',\n",
    "    'LLI': 'LLI',\n",
    "    'LLII/III': 'LLII/III',\n",
    "    'LLIV': 'LLIV',\n",
    "    'unclassified': np.nan\n",
    "})\n",
    "\n",
    "# down select only needed fields & sort\n",
    "genome_df = genome_df[['id', 'name', 'type', 'group', 'ecotype', 'clade', 'virocell', 'completeness']].sort_values(\n",
    "    ['group', 'clade', 'type', 'completeness'], ascending=[True, True, False, False]\n",
    ").reset_index(drop=True)\n",
    "\n",
    "# save cleaned up genome dataframe\n",
    "genome_df.to_csv('metadata/genome-metadata.csv', index=False)\n",
    "\n",
    "genome_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BerubeProportalID</th>\n",
       "      <th>UpdatedProportalID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Ecotype</th>\n",
       "      <th>Clade</th>\n",
       "      <th>ReferenceType</th>\n",
       "      <th>IsolationLocation</th>\n",
       "      <th>Ecosystem</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Depth(m)</th>\n",
       "      <th>GenomeSize(bp)</th>\n",
       "      <th>GeneCount</th>\n",
       "      <th>Completeness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2716884681</td>\n",
       "      <td>2716884681</td>\n",
       "      <td>AG-311-D23</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1466304</td>\n",
       "      <td>1796</td>\n",
       "      <td>72.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2716884682</td>\n",
       "      <td>2716884682</td>\n",
       "      <td>AG-311-I02</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>20.0</td>\n",
       "      <td>195290</td>\n",
       "      <td>271</td>\n",
       "      <td>11.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2716884683</td>\n",
       "      <td>2716884683</td>\n",
       "      <td>AG-311-I09</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>20.0</td>\n",
       "      <td>697970</td>\n",
       "      <td>812</td>\n",
       "      <td>47.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2716884684</td>\n",
       "      <td>2716884684</td>\n",
       "      <td>AG-311-J05</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>20.0</td>\n",
       "      <td>623148</td>\n",
       "      <td>755</td>\n",
       "      <td>34.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2716884685</td>\n",
       "      <td>2716884685</td>\n",
       "      <td>AG-311-J23</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1427538</td>\n",
       "      <td>1677</td>\n",
       "      <td>77.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>676</th>\n",
       "      <td>2623620330</td>\n",
       "      <td>640427149</td>\n",
       "      <td>WH7803</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.1B-V</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Sargasso Sea, North Atlantic Ocean</td>\n",
       "      <td>Epipelagic</td>\n",
       "      <td>33.742300</td>\n",
       "      <td>-67.491300</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2366980</td>\n",
       "      <td>2591</td>\n",
       "      <td>98.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677</th>\n",
       "      <td>2623620868</td>\n",
       "      <td>638341215</td>\n",
       "      <td>WH7805</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.1B-VI</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>North Atlantic Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>33.750000</td>\n",
       "      <td>-67.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2620367</td>\n",
       "      <td>2937</td>\n",
       "      <td>99.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>678</th>\n",
       "      <td>2606217514</td>\n",
       "      <td>637000314</td>\n",
       "      <td>WH8102</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.1A-III</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Tropical Atlantic Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>22.495000</td>\n",
       "      <td>-65.600000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2434428</td>\n",
       "      <td>2588</td>\n",
       "      <td>99.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>679</th>\n",
       "      <td>2606217259</td>\n",
       "      <td>637000214</td>\n",
       "      <td>MED4</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Mediterranean Sea</td>\n",
       "      <td>Marginal Sea</td>\n",
       "      <td>43.200000</td>\n",
       "      <td>6.866667</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1657990</td>\n",
       "      <td>2069</td>\n",
       "      <td>99.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680</th>\n",
       "      <td>2623620733</td>\n",
       "      <td>637000213</td>\n",
       "      <td>SS120</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLII/III</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Sargasso Sea, North Atlantic Ocean</td>\n",
       "      <td>Epipelagic</td>\n",
       "      <td>28.983333</td>\n",
       "      <td>-64.350000</td>\n",
       "      <td>120.0</td>\n",
       "      <td>1751080</td>\n",
       "      <td>1933</td>\n",
       "      <td>99.86</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>681 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     BerubeProportalID  UpdatedProportalID  GenomeName            Genus  \\\n",
       "0           2716884681          2716884681  AG-311-D23  Prochlorococcus   \n",
       "1           2716884682          2716884682  AG-311-I02  Prochlorococcus   \n",
       "2           2716884683          2716884683  AG-311-I09  Prochlorococcus   \n",
       "3           2716884684          2716884684  AG-311-J05  Prochlorococcus   \n",
       "4           2716884685          2716884685  AG-311-J23  Prochlorococcus   \n",
       "..                 ...                 ...         ...              ...   \n",
       "676         2623620330           640427149      WH7803    Synechococcus   \n",
       "677         2623620868           638341215      WH7805    Synechococcus   \n",
       "678         2606217514           637000314      WH8102    Synechococcus   \n",
       "679         2606217259           637000214        MED4  Prochlorococcus   \n",
       "680         2623620733           637000213       SS120  Prochlorococcus   \n",
       "\n",
       "                     Ecotype     Clade ReferenceType  \\\n",
       "0     Low light adapted (LL)       LLI           SAG   \n",
       "1     Low light adapted (LL)       LLI           SAG   \n",
       "2    High light adapted (HL)       HLI           SAG   \n",
       "3    High light adapted (HL)       HLI           SAG   \n",
       "4     Low light adapted (LL)       LLI           SAG   \n",
       "..                       ...       ...           ...   \n",
       "676                      NaN    5.1B-V       ISOLATE   \n",
       "677                      NaN   5.1B-VI       ISOLATE   \n",
       "678                      NaN  5.1A-III       ISOLATE   \n",
       "679  High light adapted (HL)       HLI       ISOLATE   \n",
       "680   Low light adapted (LL)  LLII/III       ISOLATE   \n",
       "\n",
       "                      IsolationLocation     Ecosystem   Latitude  Longitude  \\\n",
       "0                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "1                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "2                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "3                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "4                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "..                                  ...           ...        ...        ...   \n",
       "676  Sargasso Sea, North Atlantic Ocean    Epipelagic  33.742300 -67.491300   \n",
       "677                North Atlantic Ocean       Pelagic  33.750000 -67.500000   \n",
       "678             Tropical Atlantic Ocean       Pelagic  22.495000 -65.600000   \n",
       "679                   Mediterranean Sea  Marginal Sea  43.200000   6.866667   \n",
       "680  Sargasso Sea, North Atlantic Ocean    Epipelagic  28.983333 -64.350000   \n",
       "\n",
       "     Depth(m)  GenomeSize(bp)  GeneCount  Completeness  \n",
       "0        20.0         1466304       1796         72.96  \n",
       "1        20.0          195290        271         11.16  \n",
       "2        20.0          697970        812         47.41  \n",
       "3        20.0          623148        755         34.64  \n",
       "4        20.0         1427538       1677         77.90  \n",
       "..        ...             ...        ...           ...  \n",
       "676      25.0         2366980       2591         98.91  \n",
       "677       NaN         2620367       2937         99.73  \n",
       "678       NaN         2434428       2588         99.46  \n",
       "679       5.0         1657990       2069         99.46  \n",
       "680     120.0         1751080       1933         99.86  \n",
       "\n",
       "[681 rows x 15 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# save cleaned up genome dataframe\n",
    "\n",
    "genome_df = genome_df.drop(columns='Prefix')\n",
    "genome_df.to_csv('../data/berube_pro_syn_set/genome_metadata.csv', index=False)\n",
    "genome_df.to_csv('../../../data/mappings/2021-02-18/collated/berube_pro_syn_set/genome_metadata.csv', index=False)\n",
    "genome_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean Ortholog Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MappingName</th>\n",
       "      <th>OrthologID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WH8102_2607658325</td>\n",
       "      <td>60000001</td>\n",
       "      <td>WH8102</td>\n",
       "      <td>2607658325</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIT0917_2681971350</td>\n",
       "      <td>60000001</td>\n",
       "      <td>MIT0917</td>\n",
       "      <td>2681971350</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-424-P18_2717338506</td>\n",
       "      <td>60000001</td>\n",
       "      <td>AG-424-P18</td>\n",
       "      <td>2717338506</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scB245a_521A19_2655604637</td>\n",
       "      <td>60000001</td>\n",
       "      <td>scB245a_521A19</td>\n",
       "      <td>2655604637</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GFB01_2638208352</td>\n",
       "      <td>60000001</td>\n",
       "      <td>GFB01</td>\n",
       "      <td>2638208352</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964917</th>\n",
       "      <td>AG-363-C02_2667889608</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667889608</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964918</th>\n",
       "      <td>AG-363-C02_2667889615</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667889615</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964919</th>\n",
       "      <td>AG-363-C02_2667890048</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667890048</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964920</th>\n",
       "      <td>AG-363-C02_2667890054</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667890054</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964921</th>\n",
       "      <td>AG-363-C02_2667890060</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667890060</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>964922 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      MappingName  OrthologID      GenomeName      GeneID  \\\n",
       "0               WH8102_2607658325    60000001          WH8102  2607658325   \n",
       "1              MIT0917_2681971350    60000001         MIT0917  2681971350   \n",
       "2           AG-424-P18_2717338506    60000001      AG-424-P18  2717338506   \n",
       "3       scB245a_521A19_2655604637    60000001  scB245a_521A19  2655604637   \n",
       "4                GFB01_2638208352    60000001           GFB01  2638208352   \n",
       "...                           ...         ...             ...         ...   \n",
       "964917      AG-363-C02_2667889608    60040295      AG-363-C02  2667889608   \n",
       "964918      AG-363-C02_2667889615    60040295      AG-363-C02  2667889615   \n",
       "964919      AG-363-C02_2667890048    60040295      AG-363-C02  2667890048   \n",
       "964920      AG-363-C02_2667890054    60040295      AG-363-C02  2667890054   \n",
       "964921      AG-363-C02_2667890060    60040295      AG-363-C02  2667890060   \n",
       "\n",
       "                                      Annotation  \n",
       "0       membrane protease FtsH catalytic subunit  \n",
       "1       membrane protease FtsH catalytic subunit  \n",
       "2       membrane protease FtsH catalytic subunit  \n",
       "3       membrane protease FtsH catalytic subunit  \n",
       "4       membrane protease FtsH catalytic subunit  \n",
       "...                                          ...  \n",
       "964917                      hypothetical protein  \n",
       "964918                      hypothetical protein  \n",
       "964919                      hypothetical protein  \n",
       "964920                      hypothetical protein  \n",
       "964921                      hypothetical protein  \n",
       "\n",
       "[964922 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clean and join\n",
    "\n",
    "# read in cycog data\n",
    "filepath_cycogs = '../../../data/genomes/berube_et_al_2018/cycogs.tsv'\n",
    "cycog_df = pd.read_csv(filepath_cycogs, sep='\\t')\n",
    "cycog_df['OrthologID'] =[int(id_val[6:]) for id_val in cycog_df['cycog_iid']]\n",
    "# make dictionary of cycog genes\n",
    "cog_dict = {}\n",
    "for _, row in cycog_df.iterrows():\n",
    "    cog = int(row['cycog_iid'][6:])\n",
    "    for gene in row['cycog_genes'].split(','):\n",
    "        cog_dict[gene] = cog\n",
    "# make dataframe from cog dict       \n",
    "cogs_df = pd.DataFrame.from_dict(cog_dict, dtype=int, orient='index', columns=['cycog_iid'])\n",
    "# clean up data columns\n",
    "cogs_df.reset_index(inplace=True)\n",
    "cogs_df.rename(columns={'index':'MappingName', 'cycog_iid':'OrthologID'}, inplace=True)\n",
    "cogs_df['GenomeName'], cogs_df['GeneID'] = list(zip(*cogs_df['MappingName'].str.rsplit('_', 1)))\n",
    "# join in annotation information\n",
    "cogs_df['Annotation'] = cogs_df.OrthologID.map(cycog_df.set_index('OrthologID')['cycog_cns_product'])\n",
    "cogs_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save cleaned ortholog metadata\n",
    "\n",
    "cogs_df.to_csv('../data/berube_pro_syn_set/ortholog_metadata.csv', index=False)\n",
    "cogs_df.to_csv('../../../data/mappings/2021-02-18/collated/berube_pro_syn_set/ortholog_metadata.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
