{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Sources\n",
    "\n",
    "This notebook documents the compilation of metadata products used as input to the analyses included in the project.\n",
    "- All external data were obtained from publically available sources. URLs and download instructions are included to aid in reproducing the provenance of these data. \n",
    "- All data generated by the Armbrust lab were posted to publically available resources. URLs and download instructions are included for these data as well.\n",
    "- The included code combines and cleans data sets, documents the reconciliation of any discrepancies, and produces cleaned metadata files that are used as inputs for project analyses.\n",
    "\n",
    "## Reference Genomes & CyCOGs\n",
    "\n",
    "All reference genomes in this project were derived from the Scientific Data publication [Single cell genomes of Prochlorococcus, Synechococcus, and sympatric microbes from diverse marine environments](https://www.nature.com/articles/sdata2018154) (Berube et al., 2018). The genome sequences and annotations used in this publication are hosted by JGI on the [IMG/ProPortal](https://img.jgi.doe.gov/cgi-bin/proportal/main.cgi) and can be downloaded using the [IMG Genome Cart tool](https://img.jgi.doe.gov/cgi-bin/mer/main.cgi?section=GenomeCart&page=genomeCart)\n",
    "\n",
    "### Publication Files\n",
    "- [Table 1](https://www.nature.com/articles/sdata2018154/tables/2)\n",
    "- [Table 2](https://www.nature.com/articles/sdata2018154/tables/3)\n",
    "\n",
    "### Figshare Files\n",
    "\n",
    "Download the following files from the [figshare site associated with Berube et al. (2018)](https://doi.org/10.6084/m9.figshare.c.4037048.v1):\n",
    "- File 4: [cycogs-genomes.tsv](https://figshare.com/articles/dataset/File_4_CyCOG_taxa/6007166)\n",
    "- File 5: [cycogs.tsv](https://figshare.com/articles/dataset/File_5_CyCOG_definitions/6007169)\n",
    "- File 6: [cyanobacteria_phylogeny_taxa.tsv](https://figshare.com/articles/dataset/File_6_Cyanobacteria_taxa/6007175)\n",
    "- File 14: [genome_assembly_summary_20180718.tsv](https://figshare.com/articles/dataset/File_14_Genome_assembly_summary/6281519)\n",
    "\n",
    "### JGI IMG/ProPortal Files\n",
    "\n",
    "Download the metadata associated with all ProPortal reference genomes from the [IMG/ProPortal Genome Browser](https://img.jgi.doe.gov/cgi-bin/proportal/main.cgi?section=ProPortal&page=genomeList&class=datamart)\n",
    "- Use the `Table Configuration` section (bottom of the page) to display the following fields: \n",
    "    - Domain\n",
    "    - Sequencing Status\n",
    "    - Study Name\n",
    "    - Genome Name / Sample Name\n",
    "    - Sequencing Center\n",
    "    - IMG Genome ID (IMG Taxon ID)\n",
    "    - Genus\n",
    "    - Ecosystem Subtype\n",
    "    - Culture Type\n",
    "    - Depth In Meters\n",
    "    - Latitude\n",
    "    - Longitude\n",
    "    - Proportal Clade\n",
    "    - Proportal Ecotype\n",
    "    - Genome Size (Number of total bases)\n",
    "    - Gene Count (Number of total genes)\n",
    "    - CheckM2 Completeness\n",
    "- Click `Select All` and `Export` to download the metadata\n",
    "    - Name the file `proportal-metadata.tsv` to make it recognizeable by the following scripts\n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon_oid</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2716884681</td>\n",
       "      <td>Uncultured_Prochlorococcus_sp._AG-311-D23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2716884682</td>\n",
       "      <td>Uncultured_Prochlorococcus_sp._AG-311-I02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2716884683</td>\n",
       "      <td>Uncultured_Prochlorococcus_sp._AG-311-I09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2716884684</td>\n",
       "      <td>Uncultured_Prochlorococcus_sp._AG-311-J05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2716884685</td>\n",
       "      <td>Uncultured_Prochlorococcus_sp._AG-311-J23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>765</th>\n",
       "      <td>651703106</td>\n",
       "      <td>Synechococcus_phage_S-ShM2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>651703107</td>\n",
       "      <td>Synechococcus_phage_Syn19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>767</th>\n",
       "      <td>641201056</td>\n",
       "      <td>Synechococcus_phage_Syn5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>768</th>\n",
       "      <td>2595698410</td>\n",
       "      <td>Synechococcus_phage_metaG-MbCM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>769</th>\n",
       "      <td>638275565</td>\n",
       "      <td>Synechococcus_phage_syn9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>770 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      taxon_oid                                       name\n",
       "0    2716884681  Uncultured_Prochlorococcus_sp._AG-311-D23\n",
       "1    2716884682  Uncultured_Prochlorococcus_sp._AG-311-I02\n",
       "2    2716884683  Uncultured_Prochlorococcus_sp._AG-311-I09\n",
       "3    2716884684  Uncultured_Prochlorococcus_sp._AG-311-J05\n",
       "4    2716884685  Uncultured_Prochlorococcus_sp._AG-311-J23\n",
       "..          ...                                        ...\n",
       "765   651703106                 Synechococcus_phage_S-ShM2\n",
       "766   651703107                  Synechococcus_phage_Syn19\n",
       "767   641201056                   Synechococcus_phage_Syn5\n",
       "768  2595698410            Synechococcus_phage_metaG-MbCM1\n",
       "769   638275565                   Synechococcus_phage_syn9\n",
       "\n",
       "[770 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make file for bulk download of genome & annotation data from JGI IMG/ProPortal\n",
    "\n",
    "cycog_genomes_df = pd.read_csv('cycogsgenomes.tsv', sep='\\t')\n",
    "\n",
    "proportal_genomes_df = cycog_genomes_df[['IMG_ID', 'JGI_GENOMEPORTAL_NAME']].rename(columns={\n",
    "    'IMG_ID': 'taxon_oid', \n",
    "    'JGI_GENOMEPORTAL_NAME': 'name'\n",
    "})\n",
    "\n",
    "proportal_genomes_df.to_csv('proportal-genomes.tsv', sep='\\t', index=False)\n",
    "\n",
    "proportal_genomes_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean Genome Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BerubeProportalID</th>\n",
       "      <th>UpdatedProportalID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Ecotype</th>\n",
       "      <th>Clade</th>\n",
       "      <th>ReferenceType</th>\n",
       "      <th>IsolationLocation</th>\n",
       "      <th>Ecosystem</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Depth(m)</th>\n",
       "      <th>GenomeSize(bp)</th>\n",
       "      <th>GeneCount</th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Prefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2716884681</td>\n",
       "      <td>2716884681</td>\n",
       "      <td>AG-311-D23</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1466304</td>\n",
       "      <td>1796</td>\n",
       "      <td>72.96</td>\n",
       "      <td>AG-311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2716884682</td>\n",
       "      <td>2716884682</td>\n",
       "      <td>AG-311-I02</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>195290</td>\n",
       "      <td>271</td>\n",
       "      <td>11.16</td>\n",
       "      <td>AG-311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2716884683</td>\n",
       "      <td>2716884683</td>\n",
       "      <td>AG-311-I09</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>697970</td>\n",
       "      <td>812</td>\n",
       "      <td>47.41</td>\n",
       "      <td>AG-311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2716884684</td>\n",
       "      <td>2716884684</td>\n",
       "      <td>AG-311-J05</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>623148</td>\n",
       "      <td>755</td>\n",
       "      <td>34.64</td>\n",
       "      <td>AG-311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2716884685</td>\n",
       "      <td>2716884685</td>\n",
       "      <td>AG-311-J23</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1427538</td>\n",
       "      <td>1677</td>\n",
       "      <td>77.90</td>\n",
       "      <td>AG-311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>676</th>\n",
       "      <td>2623620330</td>\n",
       "      <td>640427149</td>\n",
       "      <td>WH7803</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.1B-V</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Sargasso Sea, North Atlantic Ocean</td>\n",
       "      <td>Epipelagic</td>\n",
       "      <td>33.742300</td>\n",
       "      <td>-67.491300</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2366980</td>\n",
       "      <td>2591</td>\n",
       "      <td>98.91</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677</th>\n",
       "      <td>2623620868</td>\n",
       "      <td>638341215</td>\n",
       "      <td>WH7805</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.1B-VI</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>North Atlantic Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>33.750000</td>\n",
       "      <td>-67.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2620367</td>\n",
       "      <td>2937</td>\n",
       "      <td>99.73</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>678</th>\n",
       "      <td>2606217514</td>\n",
       "      <td>637000314</td>\n",
       "      <td>WH8102</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.1A-III</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Tropical Atlantic Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>22.495000</td>\n",
       "      <td>-65.600000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2434428</td>\n",
       "      <td>2588</td>\n",
       "      <td>99.46</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>679</th>\n",
       "      <td>2606217259</td>\n",
       "      <td>637000214</td>\n",
       "      <td>MED4</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Mediterranean Sea</td>\n",
       "      <td>Marginal Sea</td>\n",
       "      <td>43.200000</td>\n",
       "      <td>6.866667</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1657990</td>\n",
       "      <td>2069</td>\n",
       "      <td>99.46</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680</th>\n",
       "      <td>2623620733</td>\n",
       "      <td>637000213</td>\n",
       "      <td>SS120</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLII/III</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Sargasso Sea, North Atlantic Ocean</td>\n",
       "      <td>Epipelagic</td>\n",
       "      <td>28.983333</td>\n",
       "      <td>-64.350000</td>\n",
       "      <td>120.0</td>\n",
       "      <td>1751080</td>\n",
       "      <td>1933</td>\n",
       "      <td>99.86</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>681 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     BerubeProportalID  UpdatedProportalID  GenomeName            Genus  \\\n",
       "0           2716884681          2716884681  AG-311-D23  Prochlorococcus   \n",
       "1           2716884682          2716884682  AG-311-I02  Prochlorococcus   \n",
       "2           2716884683          2716884683  AG-311-I09  Prochlorococcus   \n",
       "3           2716884684          2716884684  AG-311-J05  Prochlorococcus   \n",
       "4           2716884685          2716884685  AG-311-J23  Prochlorococcus   \n",
       "..                 ...                 ...         ...              ...   \n",
       "676         2623620330           640427149      WH7803    Synechococcus   \n",
       "677         2623620868           638341215      WH7805    Synechococcus   \n",
       "678         2606217514           637000314      WH8102    Synechococcus   \n",
       "679         2606217259           637000214        MED4  Prochlorococcus   \n",
       "680         2623620733           637000213       SS120  Prochlorococcus   \n",
       "\n",
       "                     Ecotype     Clade ReferenceType  \\\n",
       "0     Low light adapted (LL)       LLI           SAG   \n",
       "1     Low light adapted (LL)       LLI           SAG   \n",
       "2    High light adapted (HL)       HLI           SAG   \n",
       "3    High light adapted (HL)       HLI           SAG   \n",
       "4     Low light adapted (LL)       LLI           SAG   \n",
       "..                       ...       ...           ...   \n",
       "676                      NaN    5.1B-V       ISOLATE   \n",
       "677                      NaN   5.1B-VI       ISOLATE   \n",
       "678                      NaN  5.1A-III       ISOLATE   \n",
       "679  High light adapted (HL)       HLI       ISOLATE   \n",
       "680   Low light adapted (LL)  LLII/III       ISOLATE   \n",
       "\n",
       "                      IsolationLocation     Ecosystem   Latitude  Longitude  \\\n",
       "0                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "1                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "2                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "3                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "4                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "..                                  ...           ...        ...        ...   \n",
       "676  Sargasso Sea, North Atlantic Ocean    Epipelagic  33.742300 -67.491300   \n",
       "677                North Atlantic Ocean       Pelagic  33.750000 -67.500000   \n",
       "678             Tropical Atlantic Ocean       Pelagic  22.495000 -65.600000   \n",
       "679                   Mediterranean Sea  Marginal Sea  43.200000   6.866667   \n",
       "680  Sargasso Sea, North Atlantic Ocean    Epipelagic  28.983333 -64.350000   \n",
       "\n",
       "     Depth(m)  GenomeSize(bp)  GeneCount  Completeness  Prefix  \n",
       "0         NaN         1466304       1796         72.96  AG-311  \n",
       "1         NaN          195290        271         11.16  AG-311  \n",
       "2         NaN          697970        812         47.41  AG-311  \n",
       "3         NaN          623148        755         34.64  AG-311  \n",
       "4         NaN         1427538       1677         77.90  AG-311  \n",
       "..        ...             ...        ...           ...     ...  \n",
       "676      25.0         2366980       2591         98.91     NaN  \n",
       "677       NaN         2620367       2937         99.73     NaN  \n",
       "678       NaN         2434428       2588         99.46     NaN  \n",
       "679       5.0         1657990       2069         99.46     NaN  \n",
       "680     120.0         1751080       1933         99.86     NaN  \n",
       "\n",
       "[681 rows x 16 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clean genome metadata\n",
    "\n",
    "filepath_genome_1 = '../data/berube_pro_syn_set/berube_pro_syn_set_compiled_crossref.tab'\n",
    "filepath_genome_2 = '../data/berube_pro_syn_set/cycogsgenomes.tsv'\n",
    "filepath_genome_3 = '../data/berube_pro_syn_set/taxontable2216_25-feb-2021.tab'\n",
    "genome_df = pd.read_csv(filepath_genome_1, sep='\\t')\n",
    "genome_df = pd.merge(left=pd.read_csv(filepath_genome_1, sep='\\t'), \n",
    "                     right=pd.read_csv(filepath_genome_2, sep='\\t'), left_on='berube_taxon_oid', \n",
    "                     right_on='IMG_ID', how='inner')\n",
    "genome_df = pd.merge(left=genome_df, right=pd.read_csv(filepath_genome_3, sep='\\t', \n",
    "                                                       usecols=['taxon_oid', 'Proportal Clade', 'Proportal Ecotype']), \n",
    "                     left_on='updated_taxon_oid', right_on='taxon_oid')\n",
    "genome_df.rename(columns={'berube_taxon_oid':'BerubeProportalID', 'updated_taxon_oid':'UpdatedProportalID', \n",
    "                          'IID':'GenomeName', 'GROUP':'Genus', 'TYPE':'ReferenceType', \n",
    "                          'Genome Size   * assembled':'GenomeSize(bp)', 'Gene Count   * assembled':'GeneCount', \n",
    "                          'Geographic Location':'IsolationLocation', 'GOLD Ecosystem Subtype':'Ecosystem', \n",
    "                          'Depth In Meters':'Depth(m)', 'Proportal Clade':'Clade', 'Proportal Ecotype':'Ecotype'}, \n",
    "                 inplace=True)\n",
    "genome_df = genome_df[['BerubeProportalID', 'UpdatedProportalID', 'GenomeName', 'Genus', 'Ecotype', 'Clade', \n",
    "                       'ReferenceType', 'IsolationLocation', 'Ecosystem', 'Latitude', 'Longitude', 'Depth(m)', \n",
    "                       'GenomeSize(bp)', 'GeneCount', 'Completeness']]\n",
    "# find Berube et al. 2018 prefixes to clean up depth, lat, & lon values based on published data\n",
    "genome_df['Prefix'] = [name[:6] for name in genome_df['GenomeName']]\n",
    "genome_df.loc[[name[:3] != 'AG-' for name in genome_df['GenomeName']], 'Prefix'] = np.nan\n",
    "genome_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Prefix</th>\n",
       "      <th>Biosample</th>\n",
       "      <th>Depth</th>\n",
       "      <th>Latitude_x</th>\n",
       "      <th>Longitude_x</th>\n",
       "      <th>Latitude_y</th>\n",
       "      <th>Longitude_y</th>\n",
       "      <th>Depth(m)</th>\n",
       "      <th>CheckLat</th>\n",
       "      <th>CheckLon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-311</td>\n",
       "      <td>SWC-01</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.80</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AG-315</td>\n",
       "      <td>SWC-02</td>\n",
       "      <td>55.0</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.80</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-321</td>\n",
       "      <td>SWC-03</td>\n",
       "      <td>14.0</td>\n",
       "      <td>-23.46</td>\n",
       "      <td>-88.77</td>\n",
       "      <td>-23.46</td>\n",
       "      <td>-88.77</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AG-331</td>\n",
       "      <td>SWC-04</td>\n",
       "      <td>112.0</td>\n",
       "      <td>-23.46</td>\n",
       "      <td>-88.77</td>\n",
       "      <td>-23.46</td>\n",
       "      <td>-88.77</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AG-335</td>\n",
       "      <td>SWC-05</td>\n",
       "      <td>14.0</td>\n",
       "      <td>-26.25</td>\n",
       "      <td>-103.96</td>\n",
       "      <td>-26.25</td>\n",
       "      <td>-103.96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AG-341</td>\n",
       "      <td>SWC-06</td>\n",
       "      <td>180.0</td>\n",
       "      <td>-26.25</td>\n",
       "      <td>-103.96</td>\n",
       "      <td>-26.25</td>\n",
       "      <td>-103.96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AG-347</td>\n",
       "      <td>SWC-07</td>\n",
       "      <td>5.0</td>\n",
       "      <td>23.75</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>22.75</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AG-402</td>\n",
       "      <td>SWC-08</td>\n",
       "      <td>100.0</td>\n",
       "      <td>23.75</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>22.75</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AG-355</td>\n",
       "      <td>SWC-09</td>\n",
       "      <td>10.0</td>\n",
       "      <td>31.07</td>\n",
       "      <td>-64.17</td>\n",
       "      <td>31.67</td>\n",
       "      <td>-64.17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AG-363</td>\n",
       "      <td>SWC-10</td>\n",
       "      <td>100.0</td>\n",
       "      <td>31.07</td>\n",
       "      <td>-64.17</td>\n",
       "      <td>31.67</td>\n",
       "      <td>-64.17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AG-388</td>\n",
       "      <td>SWC-11</td>\n",
       "      <td>8.0</td>\n",
       "      <td>36.20</td>\n",
       "      <td>-53.31</td>\n",
       "      <td>36.20</td>\n",
       "      <td>-53.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AG-412</td>\n",
       "      <td>SWC-12</td>\n",
       "      <td>119.0</td>\n",
       "      <td>24.71</td>\n",
       "      <td>-67.07</td>\n",
       "      <td>24.71</td>\n",
       "      <td>-67.07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AG-409</td>\n",
       "      <td>SWC-13</td>\n",
       "      <td>100.0</td>\n",
       "      <td>9.55</td>\n",
       "      <td>-50.47</td>\n",
       "      <td>9.55</td>\n",
       "      <td>-50.47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AG-418</td>\n",
       "      <td>SWC-14</td>\n",
       "      <td>57.7</td>\n",
       "      <td>24.00</td>\n",
       "      <td>-22.00</td>\n",
       "      <td>24.00</td>\n",
       "      <td>-22.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AG-424</td>\n",
       "      <td>SWC-15</td>\n",
       "      <td>90.8</td>\n",
       "      <td>38.32</td>\n",
       "      <td>-68.87</td>\n",
       "      <td>38.32</td>\n",
       "      <td>-68.87</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AG-429</td>\n",
       "      <td>SWC-16</td>\n",
       "      <td>89.9</td>\n",
       "      <td>26.14</td>\n",
       "      <td>-44.83</td>\n",
       "      <td>26.14</td>\n",
       "      <td>-44.83</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AG-432</td>\n",
       "      <td>SWC-17</td>\n",
       "      <td>99.7</td>\n",
       "      <td>22.33</td>\n",
       "      <td>-35.87</td>\n",
       "      <td>22.33</td>\n",
       "      <td>-35.87</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AG-450</td>\n",
       "      <td>SWC-20</td>\n",
       "      <td>50.6</td>\n",
       "      <td>-30.00</td>\n",
       "      <td>156.00</td>\n",
       "      <td>-30.00</td>\n",
       "      <td>156.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AG-436</td>\n",
       "      <td>SWC-18</td>\n",
       "      <td>71.6</td>\n",
       "      <td>17.40</td>\n",
       "      <td>-24.50</td>\n",
       "      <td>17.40</td>\n",
       "      <td>-24.50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AG-459</td>\n",
       "      <td>SWC-22</td>\n",
       "      <td>76.0</td>\n",
       "      <td>-32.50</td>\n",
       "      <td>-170.00</td>\n",
       "      <td>-32.50</td>\n",
       "      <td>-170.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AG-463</td>\n",
       "      <td>SWC-23</td>\n",
       "      <td>203.0</td>\n",
       "      <td>-32.50</td>\n",
       "      <td>-170.00</td>\n",
       "      <td>-32.50</td>\n",
       "      <td>-170.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AG-469</td>\n",
       "      <td>SWC-24</td>\n",
       "      <td>100.0</td>\n",
       "      <td>-32.50</td>\n",
       "      <td>-154.00</td>\n",
       "      <td>-32.50</td>\n",
       "      <td>-154.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AG-449</td>\n",
       "      <td>SWC-20</td>\n",
       "      <td>50.6</td>\n",
       "      <td>-30.00</td>\n",
       "      <td>156.00</td>\n",
       "      <td>-30.00</td>\n",
       "      <td>156.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AG-316</td>\n",
       "      <td>SWC-02</td>\n",
       "      <td>55.0</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.80</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>AG-323</td>\n",
       "      <td>SWC-03</td>\n",
       "      <td>14.0</td>\n",
       "      <td>-23.46</td>\n",
       "      <td>-88.77</td>\n",
       "      <td>-23.46</td>\n",
       "      <td>-88.77</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>AG-420</td>\n",
       "      <td>SWC-14</td>\n",
       "      <td>57.7</td>\n",
       "      <td>24.00</td>\n",
       "      <td>-22.00</td>\n",
       "      <td>24.00</td>\n",
       "      <td>-22.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>AG-442</td>\n",
       "      <td>SWC-19</td>\n",
       "      <td>21.2</td>\n",
       "      <td>-34.98</td>\n",
       "      <td>16.02</td>\n",
       "      <td>-34.98</td>\n",
       "      <td>16.02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>AG-455</td>\n",
       "      <td>SWC-21</td>\n",
       "      <td>50.4</td>\n",
       "      <td>-30.00</td>\n",
       "      <td>174.00</td>\n",
       "      <td>-30.00</td>\n",
       "      <td>174.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>AG-670</td>\n",
       "      <td>SWC-26</td>\n",
       "      <td>5.0</td>\n",
       "      <td>28.14</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>28.14</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>AG-673</td>\n",
       "      <td>SWC-27</td>\n",
       "      <td>90.0</td>\n",
       "      <td>28.14</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>28.14</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>AG-676</td>\n",
       "      <td>SWC-28</td>\n",
       "      <td>5.0</td>\n",
       "      <td>32.70</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>34.70</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>AG-679</td>\n",
       "      <td>SWC-29</td>\n",
       "      <td>60.0</td>\n",
       "      <td>32.70</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>34.70</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>AG-683</td>\n",
       "      <td>SWC-30</td>\n",
       "      <td>5.0</td>\n",
       "      <td>36.57</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>36.57</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>AG-686</td>\n",
       "      <td>SWC-31</td>\n",
       "      <td>65.0</td>\n",
       "      <td>36.57</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>36.57</td>\n",
       "      <td>-158.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>AG-345</td>\n",
       "      <td>SWC-06</td>\n",
       "      <td>180.0</td>\n",
       "      <td>-26.25</td>\n",
       "      <td>-103.96</td>\n",
       "      <td>-26.25</td>\n",
       "      <td>-103.96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Prefix Biosample  Depth  Latitude_x  Longitude_x  Latitude_y  Longitude_y  \\\n",
       "0   AG-311    SWC-01   20.0      -20.08       -70.80      -20.08       -70.80   \n",
       "1   AG-315    SWC-02   55.0      -20.08       -70.80      -20.08       -70.80   \n",
       "2   AG-321    SWC-03   14.0      -23.46       -88.77      -23.46       -88.77   \n",
       "3   AG-331    SWC-04  112.0      -23.46       -88.77      -23.46       -88.77   \n",
       "4   AG-335    SWC-05   14.0      -26.25      -103.96      -26.25      -103.96   \n",
       "5   AG-341    SWC-06  180.0      -26.25      -103.96      -26.25      -103.96   \n",
       "6   AG-347    SWC-07    5.0       23.75      -158.00       22.75      -158.00   \n",
       "7   AG-402    SWC-08  100.0       23.75      -158.00       22.75      -158.00   \n",
       "8   AG-355    SWC-09   10.0       31.07       -64.17       31.67       -64.17   \n",
       "9   AG-363    SWC-10  100.0       31.07       -64.17       31.67       -64.17   \n",
       "10  AG-388    SWC-11    8.0       36.20       -53.31       36.20       -53.31   \n",
       "11  AG-412    SWC-12  119.0       24.71       -67.07       24.71       -67.07   \n",
       "12  AG-409    SWC-13  100.0        9.55       -50.47        9.55       -50.47   \n",
       "13  AG-418    SWC-14   57.7       24.00       -22.00       24.00       -22.00   \n",
       "14  AG-424    SWC-15   90.8       38.32       -68.87       38.32       -68.87   \n",
       "15  AG-429    SWC-16   89.9       26.14       -44.83       26.14       -44.83   \n",
       "16  AG-432    SWC-17   99.7       22.33       -35.87       22.33       -35.87   \n",
       "17  AG-450    SWC-20   50.6      -30.00       156.00      -30.00       156.00   \n",
       "18  AG-436    SWC-18   71.6       17.40       -24.50       17.40       -24.50   \n",
       "19  AG-459    SWC-22   76.0      -32.50      -170.00      -32.50      -170.00   \n",
       "20  AG-463    SWC-23  203.0      -32.50      -170.00      -32.50      -170.00   \n",
       "21  AG-469    SWC-24  100.0      -32.50      -154.00      -32.50      -154.00   \n",
       "22  AG-449    SWC-20   50.6      -30.00       156.00      -30.00       156.00   \n",
       "23  AG-316    SWC-02   55.0      -20.08       -70.80      -20.08       -70.80   \n",
       "24  AG-323    SWC-03   14.0      -23.46       -88.77      -23.46       -88.77   \n",
       "25  AG-420    SWC-14   57.7       24.00       -22.00       24.00       -22.00   \n",
       "26  AG-442    SWC-19   21.2      -34.98        16.02      -34.98        16.02   \n",
       "27  AG-455    SWC-21   50.4      -30.00       174.00      -30.00       174.00   \n",
       "28  AG-670    SWC-26    5.0       28.14      -158.00       28.14      -158.00   \n",
       "29  AG-673    SWC-27   90.0       28.14      -158.00       28.14      -158.00   \n",
       "30  AG-676    SWC-28    5.0       32.70      -158.00       34.70      -158.00   \n",
       "31  AG-679    SWC-29   60.0       32.70      -158.00       34.70      -158.00   \n",
       "32  AG-683    SWC-30    5.0       36.57      -158.00       36.57      -158.00   \n",
       "33  AG-686    SWC-31   65.0       36.57      -158.00       36.57      -158.00   \n",
       "34  AG-345    SWC-06  180.0      -26.25      -103.96      -26.25      -103.96   \n",
       "\n",
       "    Depth(m)  CheckLat  CheckLon  \n",
       "0        NaN      True      True  \n",
       "1        NaN      True      True  \n",
       "2        NaN      True      True  \n",
       "3        NaN      True      True  \n",
       "4        NaN      True      True  \n",
       "5        NaN      True      True  \n",
       "6        NaN     False      True  \n",
       "7        NaN     False      True  \n",
       "8        NaN     False      True  \n",
       "9        NaN     False      True  \n",
       "10       NaN      True      True  \n",
       "11       NaN      True      True  \n",
       "12       NaN      True      True  \n",
       "13       NaN      True      True  \n",
       "14       NaN      True      True  \n",
       "15       NaN      True      True  \n",
       "16       NaN      True      True  \n",
       "17       NaN      True      True  \n",
       "18       NaN      True      True  \n",
       "19       NaN      True      True  \n",
       "20       NaN      True      True  \n",
       "21       NaN      True      True  \n",
       "22       NaN      True      True  \n",
       "23       NaN      True      True  \n",
       "24       NaN      True      True  \n",
       "25       NaN      True      True  \n",
       "26       NaN      True      True  \n",
       "27       NaN      True      True  \n",
       "28       NaN      True      True  \n",
       "29       NaN      True      True  \n",
       "30       NaN     False      True  \n",
       "31       NaN     False      True  \n",
       "32       NaN      True      True  \n",
       "33       NaN      True      True  \n",
       "34       NaN      True      True  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check to make sure that existing lat/lon values from IMG match the published ones from Berube et al. 2018\n",
    "# NOTE: it looks like for 6 biosamples the latitude is off by a couple of degrees. \n",
    "# However, it's small enough to make me think that it's just a typo\n",
    "# I'll go with the published Berube et al. values\n",
    "\n",
    "# down-select Berube SAGs from reference set\n",
    "sag_df = genome_df.loc[[name[:3] == 'AG-' for name in genome_df['GenomeName']], :]\n",
    "sag_df = sag_df[['Prefix', 'Latitude', 'Longitude', 'Depth(m)']].drop_duplicates().reset_index(drop=True)\n",
    "# read in collection metadata from Berube et al. 2018 paper\n",
    "collection_df = pd.read_excel('../data/berube_pro_syn_set/berube_collection_data.xls', sheet_name='Compiled', \n",
    "                              dtype={'Latitude':str, 'Longitude':str})\n",
    "# fix weird annoying character thing\n",
    "collection_df['Latitude'] = collection_df['Latitude'].str.replace('âˆ’', '-').astype(float)\n",
    "collection_df['Longitude'] = collection_df['Longitude'].str.replace('âˆ’', '-').astype(float)\n",
    "# join the two and check equality\n",
    "check_df = pd.merge(collection_df, sag_df, on=['Prefix'])\n",
    "check_df['CheckLat'] = check_df['Latitude_x'].eq(check_df['Latitude_y'])\n",
    "check_df['CheckLon'] = (check_df['Longitude_x'] == check_df['Longitude_y'])\n",
    "check_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Prefix</th>\n",
       "      <th>Depth(m)</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-311</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AG-311</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-311</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AG-311</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AG-311</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-20.08</td>\n",
       "      <td>-70.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>AG-686</td>\n",
       "      <td>65.0</td>\n",
       "      <td>36.57</td>\n",
       "      <td>-158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>AG-686</td>\n",
       "      <td>65.0</td>\n",
       "      <td>36.57</td>\n",
       "      <td>-158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>AG-686</td>\n",
       "      <td>65.0</td>\n",
       "      <td>36.57</td>\n",
       "      <td>-158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>AG-686</td>\n",
       "      <td>65.0</td>\n",
       "      <td>36.57</td>\n",
       "      <td>-158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537</th>\n",
       "      <td>AG-686</td>\n",
       "      <td>65.0</td>\n",
       "      <td>36.57</td>\n",
       "      <td>-158.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>538 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Prefix  Depth(m)  Latitude  Longitude\n",
       "0    AG-311      20.0    -20.08      -70.8\n",
       "1    AG-311      20.0    -20.08      -70.8\n",
       "2    AG-311      20.0    -20.08      -70.8\n",
       "3    AG-311      20.0    -20.08      -70.8\n",
       "4    AG-311      20.0    -20.08      -70.8\n",
       "..      ...       ...       ...        ...\n",
       "533  AG-686      65.0     36.57     -158.0\n",
       "534  AG-686      65.0     36.57     -158.0\n",
       "535  AG-686      65.0     36.57     -158.0\n",
       "536  AG-686      65.0     36.57     -158.0\n",
       "537  AG-686      65.0     36.57     -158.0\n",
       "\n",
       "[538 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Replace Berube et al. SAG isolation lat, lon, depth values from IMG with published values\n",
    "\n",
    "# Depth\n",
    "genome_df.loc[genome_df['Prefix'].notna(), 'Depth(m)'] = \\\n",
    "    genome_df.loc[genome_df['Prefix'].notna()].Prefix.map(collection_df.set_index('Prefix')['Depth'])\n",
    "# Latitude\n",
    "genome_df.loc[genome_df['Prefix'].notna(), 'Latitude'] = \\\n",
    "    genome_df.loc[genome_df['Prefix'].notna()].Prefix.map(collection_df.set_index('Prefix')['Latitude'])\n",
    "# Longitude\n",
    "genome_df.loc[genome_df['Prefix'].notna(), 'Longitude'] = \\\n",
    "    genome_df.loc[genome_df['Prefix'].notna()].Prefix.map(collection_df.set_index('Prefix')['Longitude'])\n",
    "\n",
    "genome_df.loc[genome_df['Prefix'].notna(), ['Prefix', 'Depth(m)', 'Latitude', 'Longitude']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BerubeProportalID</th>\n",
       "      <th>UpdatedProportalID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Ecotype</th>\n",
       "      <th>Clade</th>\n",
       "      <th>ReferenceType</th>\n",
       "      <th>IsolationLocation</th>\n",
       "      <th>Ecosystem</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Depth(m)</th>\n",
       "      <th>GenomeSize(bp)</th>\n",
       "      <th>GeneCount</th>\n",
       "      <th>Completeness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2716884681</td>\n",
       "      <td>2716884681</td>\n",
       "      <td>AG-311-D23</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1466304</td>\n",
       "      <td>1796</td>\n",
       "      <td>72.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2716884682</td>\n",
       "      <td>2716884682</td>\n",
       "      <td>AG-311-I02</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>20.0</td>\n",
       "      <td>195290</td>\n",
       "      <td>271</td>\n",
       "      <td>11.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2716884683</td>\n",
       "      <td>2716884683</td>\n",
       "      <td>AG-311-I09</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>20.0</td>\n",
       "      <td>697970</td>\n",
       "      <td>812</td>\n",
       "      <td>47.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2716884684</td>\n",
       "      <td>2716884684</td>\n",
       "      <td>AG-311-J05</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>20.0</td>\n",
       "      <td>623148</td>\n",
       "      <td>755</td>\n",
       "      <td>34.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2716884685</td>\n",
       "      <td>2716884685</td>\n",
       "      <td>AG-311-J23</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLI</td>\n",
       "      <td>SAG</td>\n",
       "      <td>South Pacific Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>-20.080000</td>\n",
       "      <td>-70.800000</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1427538</td>\n",
       "      <td>1677</td>\n",
       "      <td>77.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>676</th>\n",
       "      <td>2623620330</td>\n",
       "      <td>640427149</td>\n",
       "      <td>WH7803</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.1B-V</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Sargasso Sea, North Atlantic Ocean</td>\n",
       "      <td>Epipelagic</td>\n",
       "      <td>33.742300</td>\n",
       "      <td>-67.491300</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2366980</td>\n",
       "      <td>2591</td>\n",
       "      <td>98.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677</th>\n",
       "      <td>2623620868</td>\n",
       "      <td>638341215</td>\n",
       "      <td>WH7805</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.1B-VI</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>North Atlantic Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>33.750000</td>\n",
       "      <td>-67.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2620367</td>\n",
       "      <td>2937</td>\n",
       "      <td>99.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>678</th>\n",
       "      <td>2606217514</td>\n",
       "      <td>637000314</td>\n",
       "      <td>WH8102</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.1A-III</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Tropical Atlantic Ocean</td>\n",
       "      <td>Pelagic</td>\n",
       "      <td>22.495000</td>\n",
       "      <td>-65.600000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2434428</td>\n",
       "      <td>2588</td>\n",
       "      <td>99.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>679</th>\n",
       "      <td>2606217259</td>\n",
       "      <td>637000214</td>\n",
       "      <td>MED4</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>High light adapted (HL)</td>\n",
       "      <td>HLI</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Mediterranean Sea</td>\n",
       "      <td>Marginal Sea</td>\n",
       "      <td>43.200000</td>\n",
       "      <td>6.866667</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1657990</td>\n",
       "      <td>2069</td>\n",
       "      <td>99.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680</th>\n",
       "      <td>2623620733</td>\n",
       "      <td>637000213</td>\n",
       "      <td>SS120</td>\n",
       "      <td>Prochlorococcus</td>\n",
       "      <td>Low light adapted (LL)</td>\n",
       "      <td>LLII/III</td>\n",
       "      <td>ISOLATE</td>\n",
       "      <td>Sargasso Sea, North Atlantic Ocean</td>\n",
       "      <td>Epipelagic</td>\n",
       "      <td>28.983333</td>\n",
       "      <td>-64.350000</td>\n",
       "      <td>120.0</td>\n",
       "      <td>1751080</td>\n",
       "      <td>1933</td>\n",
       "      <td>99.86</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>681 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     BerubeProportalID  UpdatedProportalID  GenomeName            Genus  \\\n",
       "0           2716884681          2716884681  AG-311-D23  Prochlorococcus   \n",
       "1           2716884682          2716884682  AG-311-I02  Prochlorococcus   \n",
       "2           2716884683          2716884683  AG-311-I09  Prochlorococcus   \n",
       "3           2716884684          2716884684  AG-311-J05  Prochlorococcus   \n",
       "4           2716884685          2716884685  AG-311-J23  Prochlorococcus   \n",
       "..                 ...                 ...         ...              ...   \n",
       "676         2623620330           640427149      WH7803    Synechococcus   \n",
       "677         2623620868           638341215      WH7805    Synechococcus   \n",
       "678         2606217514           637000314      WH8102    Synechococcus   \n",
       "679         2606217259           637000214        MED4  Prochlorococcus   \n",
       "680         2623620733           637000213       SS120  Prochlorococcus   \n",
       "\n",
       "                     Ecotype     Clade ReferenceType  \\\n",
       "0     Low light adapted (LL)       LLI           SAG   \n",
       "1     Low light adapted (LL)       LLI           SAG   \n",
       "2    High light adapted (HL)       HLI           SAG   \n",
       "3    High light adapted (HL)       HLI           SAG   \n",
       "4     Low light adapted (LL)       LLI           SAG   \n",
       "..                       ...       ...           ...   \n",
       "676                      NaN    5.1B-V       ISOLATE   \n",
       "677                      NaN   5.1B-VI       ISOLATE   \n",
       "678                      NaN  5.1A-III       ISOLATE   \n",
       "679  High light adapted (HL)       HLI       ISOLATE   \n",
       "680   Low light adapted (LL)  LLII/III       ISOLATE   \n",
       "\n",
       "                      IsolationLocation     Ecosystem   Latitude  Longitude  \\\n",
       "0                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "1                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "2                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "3                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "4                   South Pacific Ocean       Pelagic -20.080000 -70.800000   \n",
       "..                                  ...           ...        ...        ...   \n",
       "676  Sargasso Sea, North Atlantic Ocean    Epipelagic  33.742300 -67.491300   \n",
       "677                North Atlantic Ocean       Pelagic  33.750000 -67.500000   \n",
       "678             Tropical Atlantic Ocean       Pelagic  22.495000 -65.600000   \n",
       "679                   Mediterranean Sea  Marginal Sea  43.200000   6.866667   \n",
       "680  Sargasso Sea, North Atlantic Ocean    Epipelagic  28.983333 -64.350000   \n",
       "\n",
       "     Depth(m)  GenomeSize(bp)  GeneCount  Completeness  \n",
       "0        20.0         1466304       1796         72.96  \n",
       "1        20.0          195290        271         11.16  \n",
       "2        20.0          697970        812         47.41  \n",
       "3        20.0          623148        755         34.64  \n",
       "4        20.0         1427538       1677         77.90  \n",
       "..        ...             ...        ...           ...  \n",
       "676      25.0         2366980       2591         98.91  \n",
       "677       NaN         2620367       2937         99.73  \n",
       "678       NaN         2434428       2588         99.46  \n",
       "679       5.0         1657990       2069         99.46  \n",
       "680     120.0         1751080       1933         99.86  \n",
       "\n",
       "[681 rows x 15 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# save cleaned up genome dataframe\n",
    "\n",
    "genome_df = genome_df.drop(columns='Prefix')\n",
    "genome_df.to_csv('../data/berube_pro_syn_set/genome_metadata.csv', index=False)\n",
    "genome_df.to_csv('../../../data/mappings/2021-02-18/collated/berube_pro_syn_set/genome_metadata.csv', index=False)\n",
    "genome_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean Ortholog Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MappingName</th>\n",
       "      <th>OrthologID</th>\n",
       "      <th>GenomeName</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WH8102_2607658325</td>\n",
       "      <td>60000001</td>\n",
       "      <td>WH8102</td>\n",
       "      <td>2607658325</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIT0917_2681971350</td>\n",
       "      <td>60000001</td>\n",
       "      <td>MIT0917</td>\n",
       "      <td>2681971350</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-424-P18_2717338506</td>\n",
       "      <td>60000001</td>\n",
       "      <td>AG-424-P18</td>\n",
       "      <td>2717338506</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scB245a_521A19_2655604637</td>\n",
       "      <td>60000001</td>\n",
       "      <td>scB245a_521A19</td>\n",
       "      <td>2655604637</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GFB01_2638208352</td>\n",
       "      <td>60000001</td>\n",
       "      <td>GFB01</td>\n",
       "      <td>2638208352</td>\n",
       "      <td>membrane protease FtsH catalytic subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964917</th>\n",
       "      <td>AG-363-C02_2667889608</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667889608</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964918</th>\n",
       "      <td>AG-363-C02_2667889615</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667889615</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964919</th>\n",
       "      <td>AG-363-C02_2667890048</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667890048</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964920</th>\n",
       "      <td>AG-363-C02_2667890054</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667890054</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964921</th>\n",
       "      <td>AG-363-C02_2667890060</td>\n",
       "      <td>60040295</td>\n",
       "      <td>AG-363-C02</td>\n",
       "      <td>2667890060</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>964922 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      MappingName  OrthologID      GenomeName      GeneID  \\\n",
       "0               WH8102_2607658325    60000001          WH8102  2607658325   \n",
       "1              MIT0917_2681971350    60000001         MIT0917  2681971350   \n",
       "2           AG-424-P18_2717338506    60000001      AG-424-P18  2717338506   \n",
       "3       scB245a_521A19_2655604637    60000001  scB245a_521A19  2655604637   \n",
       "4                GFB01_2638208352    60000001           GFB01  2638208352   \n",
       "...                           ...         ...             ...         ...   \n",
       "964917      AG-363-C02_2667889608    60040295      AG-363-C02  2667889608   \n",
       "964918      AG-363-C02_2667889615    60040295      AG-363-C02  2667889615   \n",
       "964919      AG-363-C02_2667890048    60040295      AG-363-C02  2667890048   \n",
       "964920      AG-363-C02_2667890054    60040295      AG-363-C02  2667890054   \n",
       "964921      AG-363-C02_2667890060    60040295      AG-363-C02  2667890060   \n",
       "\n",
       "                                      Annotation  \n",
       "0       membrane protease FtsH catalytic subunit  \n",
       "1       membrane protease FtsH catalytic subunit  \n",
       "2       membrane protease FtsH catalytic subunit  \n",
       "3       membrane protease FtsH catalytic subunit  \n",
       "4       membrane protease FtsH catalytic subunit  \n",
       "...                                          ...  \n",
       "964917                      hypothetical protein  \n",
       "964918                      hypothetical protein  \n",
       "964919                      hypothetical protein  \n",
       "964920                      hypothetical protein  \n",
       "964921                      hypothetical protein  \n",
       "\n",
       "[964922 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clean and join\n",
    "\n",
    "# read in cycog data\n",
    "filepath_cycogs = '../../../data/genomes/berube_et_al_2018/cycogs.tsv'\n",
    "cycog_df = pd.read_csv(filepath_cycogs, sep='\\t')\n",
    "cycog_df['OrthologID'] =[int(id_val[6:]) for id_val in cycog_df['cycog_iid']]\n",
    "# make dictionary of cycog genes\n",
    "cog_dict = {}\n",
    "for _, row in cycog_df.iterrows():\n",
    "    cog = int(row['cycog_iid'][6:])\n",
    "    for gene in row['cycog_genes'].split(','):\n",
    "        cog_dict[gene] = cog\n",
    "# make dataframe from cog dict       \n",
    "cogs_df = pd.DataFrame.from_dict(cog_dict, dtype=int, orient='index', columns=['cycog_iid'])\n",
    "# clean up data columns\n",
    "cogs_df.reset_index(inplace=True)\n",
    "cogs_df.rename(columns={'index':'MappingName', 'cycog_iid':'OrthologID'}, inplace=True)\n",
    "cogs_df['GenomeName'], cogs_df['GeneID'] = list(zip(*cogs_df['MappingName'].str.rsplit('_', 1)))\n",
    "# join in annotation information\n",
    "cogs_df['Annotation'] = cogs_df.OrthologID.map(cycog_df.set_index('OrthologID')['cycog_cns_product'])\n",
    "cogs_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save cleaned ortholog metadata\n",
    "\n",
    "cogs_df.to_csv('../data/berube_pro_syn_set/ortholog_metadata.csv', index=False)\n",
    "cogs_df.to_csv('../../../data/mappings/2021-02-18/collated/berube_pro_syn_set/ortholog_metadata.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
